error[E0597]: borrowed value does not live long enough
  --> $DIR/const-int-sign.rs:2:29
   |
LL |     let x: &'static bool = &(5_i32.is_negative()); //~ ERROR does not live long enough
   |                             ^^^^^^^^^^^^^^^^^^^^^ temporary value does not live long enough
LL |     let y: &'static bool = &(5_i32.is_positive()); //~ ERROR does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/const-int-sign.rs:3:29
   |
LL |     let y: &'static bool = &(5_i32.is_positive()); //~ ERROR does not live long enough
   |                             ^^^^^^^^^^^^^^^^^^^^^ temporary value does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
