error[E0005]: refutable pattern in local binding: `0u8..=1u8` not covered
  --> $DIR/const-pattern-irrefutable.rs:12:9
   |
LL |     let a = 4; //~ ERROR refutable pattern in local binding: `0u8..=1u8` not covered
   |         ^ interpreted as a constant pattern, not new variable

error[E0005]: refutable pattern in local binding: `0u8..=1u8` not covered
  --> $DIR/const-pattern-irrefutable.rs:13:9
   |
LL |     let c = 4; //~ ERROR refutable pattern in local binding: `0u8..=1u8` not covered
   |         ^ interpreted as a constant pattern, not new variable

error[E0005]: refutable pattern in local binding: `0u8..=1u8` not covered
  --> $DIR/const-pattern-irrefutable.rs:14:9
   |
LL |     let d = 4; //~ ERROR refutable pattern in local binding: `0u8..=1u8` not covered
   |         ^ interpreted as a constant pattern, not new variable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0005`.
