error[E0506]: cannot assign to `x` because it is borrowed
  --> $DIR/closure-use-spans.rs:7:5
   |
LL |     let y = &x;
   |             -- borrow of `x` occurs here
LL |     x = 0; //~ ERROR
   |     ^^^^^ assignment to borrowed `x` occurs here
LL |     || *y;
   |         - borrow later captured here by closure

error[E0506]: cannot assign to `x` because it is borrowed
  --> $DIR/closure-use-spans.rs:13:5
   |
LL |     let y = &mut x;
   |             ------ borrow of `x` occurs here
LL |     x = 0; //~ ERROR
   |     ^^^^^ assignment to borrowed `x` occurs here
LL |     || *y = 1;
   |         - borrow later captured here by closure

error[E0506]: cannot assign to `x` because it is borrowed
  --> $DIR/closure-use-spans.rs:19:5
   |
LL |     let y = &x;
   |             -- borrow of `x` occurs here
LL |     x = 0; //~ ERROR
   |     ^^^^^ assignment to borrowed `x` occurs here
LL |     move || *y;
   |              - borrow later captured here by closure

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0506`.
