error[E0597]: `s` does not live long enough
  --> $DIR/issue-55850.rs:28:16
   |
LL |         yield &s[..] //~ ERROR `s` does not live long enough [E0597]
   |                ^ borrowed value does not live long enough
LL |     })
   |     - `s` dropped here while still borrowed

error[E0626]: borrow may still be in use when generator yields
  --> $DIR/issue-55850.rs:28:16
   |
LL |         yield &s[..] //~ ERROR `s` does not live long enough [E0597]
   |         -------^---- possible yield occurs here

error: aborting due to 2 previous errors

Some errors occurred: E0597, E0626.
For more information about an error, try `rustc --explain E0597`.
