error[E0310]: the parameter type `U` may not live long enough
  --> $DIR/dont-infer-static.rs:10:5
   |
LL | struct Foo<U> {
   |            - help: consider adding an explicit lifetime bound `U: 'static`...
LL |     bar: Bar<U> //~ ERROR the parameter type `U` may not live long enough [E0310]
   |     ^^^^^^^^^^^
   |
note: ...so that the type `U` will meet its required lifetime bounds
  --> $DIR/dont-infer-static.rs:10:5
   |
LL |     bar: Bar<U> //~ ERROR the parameter type `U` may not live long enough [E0310]
   |     ^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0310`.
