error[E0384]: cannot assign twice to immutable variable `x` (Ast)
  --> $DIR/liveness-assign-imm-local-notes.rs:13:9
   |
LL |         x = 2;
   |         ----- first assignment to `x`
LL |         x = 3;      //~ ERROR (Ast) [E0384]
   |         ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Ast)
  --> $DIR/liveness-assign-imm-local-notes.rs:25:13
   |
LL |             x = 2;
   |             ----- first assignment to `x`
LL |             x = 3;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Ast)
  --> $DIR/liveness-assign-imm-local-notes.rs:35:13
   |
LL |             x = 1;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Ast)
  --> $DIR/liveness-assign-imm-local-notes.rs:38:13
   |
LL |             x = 1;      //~ ERROR (Ast) [E0384]
   |             ----- first assignment to `x`
...
LL |             x = 2;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Mir)
  --> $DIR/liveness-assign-imm-local-notes.rs:13:9
   |
LL |     let x;
   |         - help: make this binding mutable: `mut x`
...
LL |         x = 2;
   |         ----- first assignment to `x`
LL |         x = 3;      //~ ERROR (Ast) [E0384]
   |         ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Mir)
  --> $DIR/liveness-assign-imm-local-notes.rs:25:13
   |
LL |         let x;
   |             - help: make this binding mutable: `mut x`
...
LL |             x = 2;
   |             ----- first assignment to `x`
LL |             x = 3;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Mir)
  --> $DIR/liveness-assign-imm-local-notes.rs:35:13
   |
LL |     let x;
   |         - help: make this binding mutable: `mut x`
...
LL |             x = 1;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error[E0384]: cannot assign twice to immutable variable `x` (Mir)
  --> $DIR/liveness-assign-imm-local-notes.rs:38:13
   |
LL |     let x;
   |         - help: make this binding mutable: `mut x`
...
LL |             x = 1;      //~ ERROR (Ast) [E0384]
   |             ----- first assignment to `x`
...
LL |             x = 2;      //~ ERROR (Ast) [E0384]
   |             ^^^^^ cannot assign twice to immutable variable

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0384`.
