error: cannot find derive macro `rustfmt` in this scope
  --> $DIR/tool-attributes-misplaced-1.rs:6:10
   |
LL | #[derive(rustfmt)] //~ ERROR cannot find derive macro `rustfmt` in this scope
   |          ^^^^^^^

error: cannot find macro `rustfmt!` in this scope
  --> $DIR/tool-attributes-misplaced-1.rs:15:5
   |
LL |     rustfmt!(); //~ ERROR cannot find macro `rustfmt!` in this scope
   |     ^^^^^^^

error[E0573]: expected type, found tool module `rustfmt`
  --> $DIR/tool-attributes-misplaced-1.rs:3:10
   |
LL | type A = rustfmt; //~ ERROR expected type, found tool module `rustfmt`
   |          ^^^^^^^ not a type

error[E0573]: expected type, found tool attribute `rustfmt::skip`
  --> $DIR/tool-attributes-misplaced-1.rs:4:10
   |
LL | type B = rustfmt::skip; //~ ERROR expected type, found tool attribute `rustfmt::skip`
   |          ^^^^^^^^^^^^^ not a type

error[E0423]: expected value, found tool module `rustfmt`
  --> $DIR/tool-attributes-misplaced-1.rs:14:5
   |
LL |     rustfmt; //~ ERROR expected value, found tool module `rustfmt`
   |     ^^^^^^^ not a value

error[E0423]: expected value, found tool attribute `rustfmt::skip`
  --> $DIR/tool-attributes-misplaced-1.rs:17:5
   |
LL |     rustfmt::skip; //~ ERROR expected value, found tool attribute `rustfmt::skip`
   |     ^^^^^^^^^^^^^ not a value

error: aborting due to 6 previous errors

Some errors occurred: E0423, E0573.
For more information about an error, try `rustc --explain E0423`.
