error: can only call other `min_const_fn` within a `min_const_fn`
  --> $DIR/min_const_unsafe_fn_libstd_stability2.rs:15:32
   |
LL | const unsafe fn bar() -> u32 { foo() } //~ ERROR can only call other `min_const_fn`
   |                                ^^^^^

error: can only call other `min_const_fn` within a `min_const_fn`
  --> $DIR/min_const_unsafe_fn_libstd_stability2.rs:22:33
   |
LL | const unsafe fn bar2() -> u32 { foo2() } //~ ERROR can only call other `min_const_fn`
   |                                 ^^^^^^

error: can only call other `min_const_fn` within a `min_const_fn`
  --> $DIR/min_const_unsafe_fn_libstd_stability2.rs:30:39
   |
LL | const unsafe fn bar2_gated() -> u32 { foo2_gated() } //~ ERROR can only call other `min_const_fn`
   |                                       ^^^^^^^^^^^^

error: aborting due to 3 previous errors

