error: local variables in const fn are unstable
  --> $DIR/mutable_borrow.rs:2:9
   |
LL |     let mut a = 0; //~ ERROR local variables in const fn
   |         ^^^^^

error: local variables in const fn are unstable
  --> $DIR/mutable_borrow.rs:11:13
   |
LL |         let mut a = 0; //~ ERROR local variables in const fn
   |             ^^^^^

error: aborting due to 2 previous errors

