error[E0502]: cannot borrow `vector` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-for-loop-head-linkage.rs:7:9
   |
LL |     for &x in &vector {
   |               -------
   |               |
   |               immutable borrow occurs here
   |               immutable borrow used here, in later iteration of loop
LL |         let cap = vector.capacity();
LL |         vector.extend(repeat(0));      //~ ERROR cannot borrow
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here

error[E0502]: cannot borrow `vector` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-for-loop-head-linkage.rs:8:9
   |
LL |     for &x in &vector {
   |               -------
   |               |
   |               immutable borrow occurs here
   |               immutable borrow used here, in later iteration of loop
...
LL |         vector[1] = 5;   //~ ERROR cannot borrow
   |         ^^^^^^ mutable borrow occurs here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
