error[E0507]: cannot move out of borrowed content
  --> $DIR/borrowck-for-loop-correct-cmt-for-pattern.rs:12:9
   |
LL |     for &a in x.iter() {    //~ ERROR cannot move out
   |         ^-
   |         ||
   |         |hint: to prevent move, use `ref a` or `ref mut a`
   |         cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> $DIR/borrowck-for-loop-correct-cmt-for-pattern.rs:18:9
   |
LL |     for &a in &f.a {  //~ ERROR cannot move out
   |         ^-
   |         ||
   |         |hint: to prevent move, use `ref a` or `ref mut a`
   |         cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> $DIR/borrowck-for-loop-correct-cmt-for-pattern.rs:22:9
   |
LL |     for &a in x.iter() {    //~ ERROR cannot move out
   |         ^-
   |         ||
   |         |hint: to prevent move, use `ref a` or `ref mut a`
   |         cannot move out of borrowed content

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0507`.
