error[E0502]: cannot borrow `*x` as mutable because it is also borrowed as immutable
  --> $DIR/borrowck-object-lifetime.rs:20:13
   |
LL |     let y = x.borrowed();
   |             - immutable borrow occurs here
LL |     let z = x.mut_borrowed(); //~ ERROR cannot borrow
   |             ^ mutable borrow occurs here
LL |     y.use_ref();
LL | }
   | - immutable borrow ends here

error[E0502]: cannot borrow `x` as mutable because `*x` is also borrowed as immutable
  --> $DIR/borrowck-object-lifetime.rs:26:18
   |
LL |     let y = x.borrowed();
   |             - immutable borrow occurs here
LL |     let z = &mut x; //~ ERROR cannot borrow
   |                  ^ mutable borrow occurs here
LL |     y.use_ref();
LL | }
   | - immutable borrow ends here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
