error[E0502]: cannot borrow `a` as mutable because `*a` is also borrowed as immutable
  --> $DIR/borrowck-bad-nested-calls-free.rs:25:22
   |
LL |         &*a,
   |          -- immutable borrow occurs here
LL |         rewrite(&mut a)); //~ ERROR cannot borrow
   |                      ^ - immutable borrow ends here
   |                      |
   |                      mutable borrow occurs here

error[E0502]: cannot borrow `a` as mutable because `*a` is also borrowed as immutable
  --> $DIR/borrowck-bad-nested-calls-free.rs:32:22
   |
LL |         &*a,
   |          -- immutable borrow occurs here
LL |         rewrite(&mut a)); //~ ERROR cannot borrow
   |                      ^ - immutable borrow ends here
   |                      |
   |                      mutable borrow occurs here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
