error[E0502]: cannot borrow `*v` as immutable because `v` is also borrowed as mutable
  --> $DIR/borrowck-uniq-via-lend.rs:36:13
   |
LL |     let w = &mut v;
   |                  - mutable borrow occurs here
LL |     borrow(&*v); //~ ERROR cannot borrow `*v`
   |             ^^ immutable borrow occurs here
LL |     w.use_mut();
LL | }
   | - mutable borrow ends here

error[E0502]: cannot borrow `*v` as immutable because `v` is also borrowed as mutable
  --> $DIR/borrowck-uniq-via-lend.rs:53:13
   |
LL |     x = &mut v;
   |              - mutable borrow occurs here
LL |     borrow(&*v); //~ ERROR cannot borrow `*v`
   |             ^^ immutable borrow occurs here
LL |     x.use_mut();
LL | }
   | - mutable borrow ends here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0502`.
