warning[E0503]: cannot use `y` because it was mutably borrowed
  --> $DIR/borrowck-anon-fields-variant.rs:17:7
   |
LL |       Foo::Y(ref mut a, _) => a,
   |              --------- borrow of `y.0` occurs here
...
LL |       Foo::Y(_, ref mut b) => b,
   |       ^^^^^^^^^^^^^^^^^^^^ use of borrowed `y.0`
...
LL |     *a += 1;
   |     ------- borrow later used here
   |
   = warning: this error has been downgraded to a warning for backwards compatibility with previous releases
   = warning: this represents potential undefined behavior in your code and this warning will become a hard error in the future

error[E0503]: cannot use `y` because it was mutably borrowed
  --> $DIR/borrowck-anon-fields-variant.rs:34:7
   |
LL |       Foo::Y(ref mut a, _) => a,
   |              --------- borrow of `y.0` occurs here
...
LL |       Foo::Y(ref mut b, _) => b, //~ ERROR cannot borrow
   |       ^^^^^^^^^^^^^^^^^^^^ use of borrowed `y.0`
...
LL |     *a += 1;
   |     ------- borrow later used here

error[E0499]: cannot borrow `y.0` as mutable more than once at a time
  --> $DIR/borrowck-anon-fields-variant.rs:34:14
   |
LL |       Foo::Y(ref mut a, _) => a,
   |              --------- first mutable borrow occurs here
...
LL |       Foo::Y(ref mut b, _) => b, //~ ERROR cannot borrow
   |              ^^^^^^^^^ second mutable borrow occurs here
...
LL |     *a += 1;
   |     ------- first borrow later used here

error: aborting due to 2 previous errors

Some errors occurred: E0499, E0503.
For more information about an error, try `rustc --explain E0499`.
