error[E0499]: cannot borrow `void` as mutable more than once at a time
  --> $DIR/mut-borrow-outside-loop.rs:7:23
   |
LL |     let first = &mut void;
   |                      ---- first mutable borrow occurs here
LL |     let second = &mut void; //~ ERROR cannot borrow
   |                       ^^^^ second mutable borrow occurs here
...
LL | }
   | - first borrow ends here

error[E0499]: cannot borrow `inner_void` as mutable more than once at a time
  --> $DIR/mut-borrow-outside-loop.rs:15:33
   |
LL |         let inner_first = &mut inner_void;
   |                                ---------- first mutable borrow occurs here
LL |         let inner_second = &mut inner_void; //~ ERROR cannot borrow
   |                                 ^^^^^^^^^^ second mutable borrow occurs here
...
LL |     }
   |     - first borrow ends here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0499`.
