error[E0596]: cannot borrow immutable argument `b` as mutable
  --> $DIR/mut-borrow-of-mut-ref.rs:8:12
   |
LL |     g(&mut b) //~ ERROR cannot borrow
   |            ^ cannot borrow mutably
help: consider removing the `&mut`, as it is an immutable binding to a mutable reference
   |
LL |     g(b) //~ ERROR cannot borrow
   |       ^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
