error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:10:25
   |
LL |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
LL |         }
LL |     }
   |     - mutable borrow ends here

error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:16:25
   |
LL |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
LL |         }
LL |     }
   |     - mutable borrow ends here

error[E0499]: cannot borrow `*arg` as mutable more than once at a time
  --> $DIR/mut-borrow-in-loop.rs:23:25
   |
LL |             (self.func)(arg) //~ ERROR cannot borrow
   |                         ^^^ mutable borrow starts here in previous iteration of loop
LL |         }
LL |     }
   |     - mutable borrow ends here

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0499`.
