error[E0597]: borrowed value does not live long enough
  --> $DIR/promote-ref-mut-in-let-issue-46557.rs:5:9
   |
LL |     let ref mut x = 1234543; //~ ERROR
   |         ^^^^^^^^^ temporary value does not live long enough
LL |     x
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promote-ref-mut-in-let-issue-46557.rs:10:10
   |
LL |     let (ref mut x, ) = (1234543, ); //~ ERROR
   |          ^^^^^^^^^ borrowed value does not live long enough
LL |     x
LL | }
   | - borrowed value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promote-ref-mut-in-let-issue-46557.rs:16:9
   |
LL |         ref mut x => x //~ ERROR
   |         ^^^^^^^^^ temporary value does not live long enough
LL |     }
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promote-ref-mut-in-let-issue-46557.rs:22:10
   |
LL |         (ref mut x,) => x, //~ ERROR
   |          ^^^^^^^^^ borrowed value does not live long enough
LL |     }
LL | }
   | - borrowed value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/promote-ref-mut-in-let-issue-46557.rs:27:10
   |
LL |     &mut 1234543 //~ ERROR
   |          ^^^^^^^ temporary value does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0597`.
