error[E0594]: cannot assign to field `x.0` of immutable binding
  --> $DIR/reassignment_immutable_fields.rs:7:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `x.1` of immutable binding
  --> $DIR/reassignment_immutable_fields.rs:8:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
LL |     x.1 = 22; //~ ERROR
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error[E0381]: use of possibly uninitialized variable: `x.0`
  --> $DIR/reassignment_immutable_fields.rs:9:10
   |
LL |     drop(x.0); //~ ERROR
   |          ^^^ use of possibly uninitialized `x.0`

error[E0381]: use of possibly uninitialized variable: `x.1`
  --> $DIR/reassignment_immutable_fields.rs:10:10
   |
LL |     drop(x.1); //~ ERROR
   |          ^^^ use of possibly uninitialized `x.1`

error[E0594]: cannot assign to field `x.0` of immutable binding
  --> $DIR/reassignment_immutable_fields.rs:15:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `x.1` of immutable binding
  --> $DIR/reassignment_immutable_fields.rs:16:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
LL |     x.1 = 22; //~ ERROR
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error[E0381]: use of possibly uninitialized variable: `x`
  --> $DIR/reassignment_immutable_fields.rs:17:10
   |
LL |     drop(x); //~ ERROR
   |          ^ use of possibly uninitialized `x`

error: aborting due to 7 previous errors

Some errors occurred: E0381, E0594.
For more information about an error, try `rustc --explain E0381`.
