error[E0594]: cannot assign to field `x.0` of immutable binding
  --> $DIR/reassignment_immutable_fields_twice.rs:7:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x = (22, 44);
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `x.0` of immutable binding
  --> $DIR/reassignment_immutable_fields_twice.rs:12:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
   |     ^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `x.0` of immutable binding
  --> $DIR/reassignment_immutable_fields_twice.rs:13:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
LL |     x.0 = 1; //~ ERROR
LL |     x.0 = 22; //~ ERROR
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `x.1` of immutable binding
  --> $DIR/reassignment_immutable_fields_twice.rs:14:5
   |
LL |     let x: (u32, u32);
   |         - help: make this binding mutable: `mut x`
...
LL |     x.1 = 44; //~ ERROR
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0594`.
