error[E0594]: cannot assign to `*x` which is behind a `&` reference
  --> $DIR/enum.rs:19:5
   |
LL |     *x += 1; //~ ERROR cannot assign to immutable
   |     ^^^^^^^ cannot assign

error[E0594]: cannot assign to `*x` which is behind a `&` reference
  --> $DIR/enum.rs:23:9
   |
LL |         *x += 1; //~ ERROR cannot assign to immutable
   |         ^^^^^^^ cannot assign

error[E0594]: cannot assign to `*x` which is behind a `&` reference
  --> $DIR/enum.rs:29:9
   |
LL |         *x += 1; //~ ERROR cannot assign to immutable
   |         ^^^^^^^ cannot assign

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0594`.
