error[E0594]: cannot assign to immutable borrowed content `*x`
  --> $DIR/enum.rs:9:5
   |
LL |     *x += 1; //~ ERROR cannot assign to immutable
   |     ^^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to immutable borrowed content `*x`
  --> $DIR/enum.rs:13:9
   |
LL |         *x += 1; //~ ERROR cannot assign to immutable
   |         ^^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to immutable borrowed content `*x`
  --> $DIR/enum.rs:19:9
   |
LL |         *x += 1; //~ ERROR cannot assign to immutable
   |         ^^^^^^^ cannot borrow as mutable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0594`.
