error[E0594]: cannot assign to immutable borrowed content `*n`
  --> $DIR/explicit-mut.rs:17:13
   |
LL |         Some(n) => {
   |              - consider changing this to `n`
LL |             *n += 1; //~ ERROR cannot assign to immutable
   |             ^^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to immutable borrowed content `*n`
  --> $DIR/explicit-mut.rs:25:13
   |
LL |         Some(n) => {
   |              - consider changing this to `n`
LL |             *n += 1; //~ ERROR cannot assign to immutable
   |             ^^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to immutable borrowed content `*n`
  --> $DIR/explicit-mut.rs:33:13
   |
LL |         Some(n) => {
   |              - consider changing this to `n`
LL |             *n += 1; //~ ERROR cannot assign to immutable
   |             ^^^^^^^ cannot borrow as mutable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0594`.
