error: implementation of `Foo` is not general enough
  --> $DIR/hrtb-just-for-static.rs:24:5
   |
LL |     want_hrtb::<StaticInt>() //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: Due to a where-clause on `want_hrtb`,
   = note: `StaticInt` must implement `Foo<&'0 isize>` for any lifetime `'0`
   = note: but `StaticInt` only implements `Foo<&'1 isize>` for some lifetime `'1`

error: implementation of `Foo` is not general enough
  --> $DIR/hrtb-just-for-static.rs:30:5
   |
LL |     want_hrtb::<&'a u32>() //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = note: Due to a where-clause on `want_hrtb`,
   = note: `&'a u32` must implement `Foo<&'0 isize>` for any lifetime `'0`
   = note: but `&'1 u32` only implements `Foo<&'1 isize>` for the lifetime `'1`

error: aborting due to 2 previous errors

