// ts-simple-ast/dist/main.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/// <reference path="typings/index.d.ts" />
tslib_1.__exportStar(require("./codeBlockWriter"), exports);
tslib_1.__exportStar(require("./compiler"), exports);
tslib_1.__exportStar(require("./errors/classes"), exports);
var fileSystem_1 = require("./fileSystem");
exports.Directory = fileSystem_1.Directory;
exports.DirectoryEmitResult = fileSystem_1.DirectoryEmitResult;
tslib_1.__exportStar(require("./options"), exports);
var Project_1 = require("./Project");
exports.Project = Project_1.Project;
var Project_2 = require("./Project");
exports.default = Project_2.Project;
tslib_1.__exportStar(require("./typescript"), exports);
var createWrappedNode_1 = require("./utils/compiler/createWrappedNode");
exports.createWrappedNode = createWrappedNode_1.createWrappedNode;
var printNode_1 = require("./utils/compiler/printNode");
exports.printNode = printNode_1.printNode;
var getCompilerOptionsFromTsConfig_1 = require("./utils/tsconfig/getCompilerOptionsFromTsConfig");
exports.getCompilerOptionsFromTsConfig = getCompilerOptionsFromTsConfig_1.getCompilerOptionsFromTsConfig;
var TypeGuards_1 = require("./utils/TypeGuards");
exports.TypeGuards = TypeGuards_1.TypeGuards;
