error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/wf-outlives-ty-in-fn-or-trait.rs:9:5
   |
LL | impl<'a, T> Trait<'a, T> for usize {
   |          - help: consider adding an explicit lifetime bound `T: 'a`...
LL |     type Out = &'a fn(T); //~ ERROR `T` may not live long enough
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
note: ...so that the reference type `&'a fn(T)` does not outlive the data it points at
  --> $DIR/wf-outlives-ty-in-fn-or-trait.rs:9:5
   |
LL |     type Out = &'a fn(T); //~ ERROR `T` may not live long enough
   |     ^^^^^^^^^^^^^^^^^^^^^

error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/wf-outlives-ty-in-fn-or-trait.rs:19:5
   |
LL | impl<'a, T> Trait<'a, T> for u32 {
   |          - help: consider adding an explicit lifetime bound `T: 'a`...
LL |     type Out = &'a Baz<T>; //~ ERROR `T` may not live long enough
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
note: ...so that the reference type `&'a (dyn Baz<T> + 'a)` does not outlive the data it points at
  --> $DIR/wf-outlives-ty-in-fn-or-trait.rs:19:5
   |
LL |     type Out = &'a Baz<T>; //~ ERROR `T` may not live long enough
   |     ^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0309`.
