error[E0432]: unresolved import `libc`
  --> $DIR/issue-37887.rs:3:9
   |
LL |     use libc::*; //~ ERROR unresolved import
   |         ^^^^ maybe a missing `extern crate libc;`?

error[E0658]: use of unstable library feature 'rustc_private': this crate is being loaded from the sysroot, an unstable location; did you mean to load this crate from crates.io via `Cargo.toml` instead? (see issue #27812)
  --> $DIR/issue-37887.rs:2:5
   |
LL |     extern crate libc; //~ ERROR use of unstable
   |     ^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(rustc_private)] to the crate attributes to enable

error: aborting due to 2 previous errors

Some errors occurred: E0432, E0658.
For more information about an error, try `rustc --explain E0432`.
