error[E0308]: mismatched types
  --> $DIR/issue-12552.rs:6:5
   |
LL |     Some(k) => match k { //~ ERROR mismatched types
   |     ^^^^^^^ expected enum `std::result::Result`, found enum `std::option::Option`
   |
   = note: expected type `std::result::Result<_, {integer}>`
              found type `std::option::Option<_>`

error[E0308]: mismatched types
  --> $DIR/issue-12552.rs:9:5
   |
LL |     None => () //~ ERROR mismatched types
   |     ^^^^ expected enum `std::result::Result`, found enum `std::option::Option`
   |
   = note: expected type `std::result::Result<_, {integer}>`
              found type `std::option::Option<_>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
