error[E0412]: cannot find type `Foo` in this scope
  --> $DIR/issue-35075.rs:2:12
   |
LL |     inner: Foo<T> //~ ERROR cannot find type `Foo` in this scope
   |            ^^^ not found in this scope
help: there is an enum variant `Baz::Foo`; try using the variant's enum
   |
LL |     inner: Baz //~ ERROR cannot find type `Foo` in this scope
   |            ^^^

error[E0412]: cannot find type `Foo` in this scope
  --> $DIR/issue-35075.rs:6:9
   |
LL |     Foo(Foo<T>) //~ ERROR cannot find type `Foo` in this scope
   |         ^^^ not found in this scope
help: there is an enum variant `Baz::Foo`; try using the variant's enum
   |
LL |     Foo(Baz) //~ ERROR cannot find type `Foo` in this scope
   |         ^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0412`.
