error[E0393]: the type parameter `RHS` must be explicitly specified
  --> $DIR/issue-22560.rs:5:13
   |
LL | type Test = Add +
   |             ^^^ missing reference to `RHS`
   |
   = note: because of the default `Self` reference, type parameters must be specified on object types

error[E0393]: the type parameter `RHS` must be explicitly specified
  --> $DIR/issue-22560.rs:8:13
   |
LL |             Sub;
   |             ^^^ missing reference to `RHS`
   |
   = note: because of the default `Self` reference, type parameters must be specified on object types

error[E0225]: only auto traits can be used as additional traits in a trait object
  --> $DIR/issue-22560.rs:8:13
   |
LL |             Sub;
   |             ^^^ non-auto additional trait

error[E0191]: the value of the associated type `Output` (from the trait `std::ops::Add`) must be specified
  --> $DIR/issue-22560.rs:5:13
   |
LL |   type Test = Add +
   |  _____________^
LL | |             //~^ ERROR E0393
LL | |             //~| ERROR E0191
LL | |             Sub;
   | |_______________^ associated type `Output` must be specified

error: aborting due to 4 previous errors

Some errors occurred: E0191, E0225, E0393.
For more information about an error, try `rustc --explain E0191`.
