error[E0015]: calls in constants are limited to constant functions, tuple structs and tuple variants
  --> $DIR/issue-46843.rs:7:26
   |
LL | pub const Q: i32 = match non_const() { //~ ERROR E0015
   |                          ^^^^^^^^^^^

error[E0019]: constant contains unimplemented expression type
  --> $DIR/issue-46843.rs:8:5
   |
LL |     Thing::This => 1, //~ ERROR unimplemented expression type
   |     ^^^^^^^^^^^

error: aborting due to 2 previous errors

Some errors occurred: E0015, E0019.
For more information about an error, try `rustc --explain E0015`.
