error[E0401]: can't use type parameters from outer function
  --> $DIR/issue-3021-c.rs:4:24
   |
LL | fn siphash<T>() {
   |            - type variable from outer function
...
LL |         fn g(&self, x: T) -> T;  //~ ERROR can't use type parameters from outer function
   |            -           ^ use of type variable from outer function
   |            |
   |            help: try using a local type parameter instead: `g<T>`

error[E0401]: can't use type parameters from outer function
  --> $DIR/issue-3021-c.rs:4:30
   |
LL | fn siphash<T>() {
   |            - type variable from outer function
...
LL |         fn g(&self, x: T) -> T;  //~ ERROR can't use type parameters from outer function
   |            -                 ^ use of type variable from outer function
   |            |
   |            help: try using a local type parameter instead: `g<T>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0401`.
