error[E0658]: The attribute `derive_Clone` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/issue-32655.rs:6:11
   |
LL |         #[derive_Clone] //~ ERROR attribute `derive_Clone` is currently unknown
   |           ^^^^^^^^^^^^
...
LL | foo!();
   | ------- in this macro invocation
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error[E0658]: The attribute `derive_Clone` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/issue-32655.rs:18:7
   |
LL |     #[derive_Clone] //~ ERROR attribute `derive_Clone` is currently unknown
   |       ^^^^^^^^^^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
