error[E0425]: cannot find value `a` in this scope
  --> $DIR/issue-33525.rs:2:5
   |
LL |     a; //~ ERROR cannot find value `a`
   |     ^ not found in this scope

error[E0609]: no field `lorem` on type `&'static str`
  --> $DIR/issue-33525.rs:3:8
   |
LL |     "".lorem; //~ ERROR no field
   |        ^^^^^

error[E0609]: no field `ipsum` on type `&'static str`
  --> $DIR/issue-33525.rs:4:8
   |
LL |     "".ipsum; //~ ERROR no field
   |        ^^^^^

error: aborting due to 3 previous errors

Some errors occurred: E0425, E0609.
For more information about an error, try `rustc --explain E0425`.
