error[E0004]: non-exhaustive patterns: type `*const Bottom` is non-empty
  --> $DIR/issue-3096-2.rs:5:11
   |
LL |     match x { } //~ ERROR non-exhaustive patterns
   |           ^
   |
help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
  --> $DIR/issue-3096-2.rs:5:11
   |
LL |     match x { } //~ ERROR non-exhaustive patterns
   |           ^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0004`.
