error[E0435]: attempt to use a non-constant value in a constant
  --> $DIR/issue-42060.rs:3:23
   |
LL |     let other: typeof(thing) = thing; //~ ERROR attempt to use a non-constant value in a constant
   |                       ^^^^^ non-constant value

error[E0435]: attempt to use a non-constant value in a constant
  --> $DIR/issue-42060.rs:9:13
   |
LL |     <typeof(q)>::N //~ ERROR attempt to use a non-constant value in a constant
   |             ^ non-constant value

error[E0516]: `typeof` is a reserved keyword but unimplemented
  --> $DIR/issue-42060.rs:3:16
   |
LL |     let other: typeof(thing) = thing; //~ ERROR attempt to use a non-constant value in a constant
   |                ^^^^^^^^^^^^^ reserved keyword

error[E0516]: `typeof` is a reserved keyword but unimplemented
  --> $DIR/issue-42060.rs:9:6
   |
LL |     <typeof(q)>::N //~ ERROR attempt to use a non-constant value in a constant
   |      ^^^^^^^^^ reserved keyword

error: aborting due to 4 previous errors

Some errors occurred: E0435, E0516.
For more information about an error, try `rustc --explain E0435`.
