error[E0391]: cycle detected when processing `Y::A::{{constant}}`
  --> $DIR/issue-23302-2.rs:4:9
   |
LL |     A = Y::B as isize, //~ ERROR E0391
   |         ^^^^^^^^^^^^^
   |
   = note: ...which again requires processing `Y::A::{{constant}}`, completing the cycle
note: cycle used when const-evaluating `Y::A::{{constant}}`
  --> $DIR/issue-23302-2.rs:4:9
   |
LL |     A = Y::B as isize, //~ ERROR E0391
   |         ^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0391`.
