error: missing condition for `if` statemement
  --> $DIR/issue-13483.rs:3:14
   |
LL |     } else if { //~ ERROR missing condition
   |              ^ expected if condition here

error: missing condition for `if` statemement
  --> $DIR/issue-13483.rs:10:14
   |
LL |     } else if { //~ ERROR missing condition
   |              ^ expected if condition here

error: aborting due to 2 previous errors

