error[E0308]: mismatched types
  --> $DIR/issue-27042.rs:6:16
   |
LL |         loop { break }; //~ ERROR mismatched types
   |                ^^^^^ expected (), found i32
   |
   = note: expected type `()`
              found type `i32`

error[E0308]: mismatched types
  --> $DIR/issue-27042.rs:8:9
   |
LL | /         'b: //~ ERROR mismatched types
LL | |         while true { break }; // but here we cite the whole loop
   | |____________________________^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`

error[E0308]: mismatched types
  --> $DIR/issue-27042.rs:11:9
   |
LL | /         'c: //~ ERROR mismatched types
LL | |         for _ in None { break }; // but here we cite the whole loop
   | |_______________________________^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`

error[E0308]: mismatched types
  --> $DIR/issue-27042.rs:14:9
   |
LL | /         'd: //~ ERROR mismatched types
LL | |         while let Some(_) = None { break };
   | |__________________________________________^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
