error[E0252]: the name `Arc` is defined multiple times
  --> $DIR/issue-26886.rs:2:5
   |
LL | use std::sync::{self, Arc};
   |                       --- previous import of the type `Arc` here
LL | use std::sync::Arc; //~ ERROR the name `Arc` is defined multiple times
   |     ^^^^^^^^^^^^^^ `Arc` reimported here
   |
   = note: `Arc` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
   |
LL | use std::sync::Arc as OtherArc; //~ ERROR the name `Arc` is defined multiple times
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0252]: the name `sync` is defined multiple times
  --> $DIR/issue-26886.rs:4:5
   |
LL | use std::sync::{self, Arc};
   |                 ---- previous import of the module `sync` here
...
LL | use std::sync; //~ ERROR the name `sync` is defined multiple times
   |     ^^^^^^^^^ `sync` reimported here
   |
   = note: `sync` must be defined only once in the type namespace of this module
help: you can use `as` to change the binding name of the import
   |
LL | use std::sync as other_sync; //~ ERROR the name `sync` is defined multiple times
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0252`.
