error[E0004]: non-exhaustive patterns: type `()` is non-empty
  --> $DIR/issue-3096-1.rs:2:11
   |
LL |     match () { } //~ ERROR non-exhaustive
   |           ^^
   |
help: ensure that all possible cases are being handled, possibly by adding wildcards or more match arms
  --> $DIR/issue-3096-1.rs:2:11
   |
LL |     match () { } //~ ERROR non-exhaustive
   |           ^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0004`.
