error[E0277]: the trait bound `i32: std::convert::From<&str>` is not satisfied
  --> $DIR/try-block-bad-type.rs:7:9
   |
LL |         Err("")?; //~ ERROR the trait bound `i32: std::convert::From<&str>` is not satisfied
   |         ^^^^^^^^ the trait `std::convert::From<&str>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<u16>>
             <i32 as std::convert::From<u8>>
   = note: required by `std::convert::From::from`

error[E0271]: type mismatch resolving `<std::result::Result<i32, i32> as std::ops::Try>::Ok == &str`
  --> $DIR/try-block-bad-type.rs:12:9
   |
LL |         "" //~ ERROR type mismatch
   |         ^^ expected i32, found &str
   |
   = note: expected type `i32`
              found type `&str`

error[E0271]: type mismatch resolving `<std::result::Result<i32, i32> as std::ops::Try>::Ok == ()`
  --> $DIR/try-block-bad-type.rs:15:39
   |
LL |     let res: Result<i32, i32> = try { }; //~ ERROR type mismatch
   |                                       ^ expected i32, found ()
   |
   = note: expected type `i32`
              found type `()`

error[E0277]: the trait bound `(): std::ops::Try` is not satisfied
  --> $DIR/try-block-bad-type.rs:17:23
   |
LL |     let res: () = try { }; //~ the trait bound `(): std::ops::Try` is not satisfied
   |                       ^^^ the trait `std::ops::Try` is not implemented for `()`
   |
   = note: required by `std::ops::Try::from_ok`

error[E0277]: the trait bound `i32: std::ops::Try` is not satisfied
  --> $DIR/try-block-bad-type.rs:19:24
   |
LL |     let res: i32 = try { 5 }; //~ ERROR the trait bound `i32: std::ops::Try` is not satisfied
   |                        ^^^^^ the trait `std::ops::Try` is not implemented for `i32`
   |
   = note: required by `std::ops::Try::from_ok`

error: aborting due to 5 previous errors

Some errors occurred: E0271, E0277.
For more information about an error, try `rustc --explain E0271`.
