error: defining existential type use differs from previous
  --> $DIR/different_defining_uses_never_type.rs:12:1
   |
LL | / fn bar() -> Foo { //~ ERROR defining existential type use differs from previous
LL | |     panic!()
LL | | }
   | |_^
   |
note: previous use here
  --> $DIR/different_defining_uses_never_type.rs:8:1
   |
LL | / fn foo() -> Foo {
LL | |     ""
LL | | }
   | |_^

error: defining existential type use differs from previous
  --> $DIR/different_defining_uses_never_type.rs:16:1
   |
LL | / fn boo() -> Foo { //~ ERROR defining existential type use differs from previous
LL | |     loop {}
LL | | }
   | |_^
   |
note: previous use here
  --> $DIR/different_defining_uses_never_type.rs:8:1
   |
LL | / fn foo() -> Foo {
LL | |     ""
LL | | }
   | |_^

error: aborting due to 2 previous errors

