// tslint-no-circular-imports/noCircularImportsRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var path_1 = require("path");
var Lint = require("tslint");
var ts = require("typescript");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.applyWithProgram = function (sourceFile, program) {
        var resolvedFile = sourceFile.fileName;
        imports.delete(resolvedFile);
        var walker = new NoCircularImportsWalker(sourceFile, this.getOptions(), program);
        return this.applyWithWalker(walker);
    };
    Rule.FAILURE_STRING = 'circular import detected';
    Rule.metadata = {
        ruleName: 'no-circular-imports',
        description: 'Disallows circular imports.',
        rationale: (_a = ["\n        Circular dependencies cause hard-to-catch runtime exceptions."], _a.raw = ["\n        Circular dependencies cause hard-to-catch runtime exceptions."], Lint.Utils.dedent(_a)),
        optionsDescription: 'Not configurable.',
        options: null,
        optionExamples: ['true'],
        type: 'functionality',
        typescriptOnly: false
    };
    return Rule;
}(Lint.Rules.TypedRule));
exports.Rule = Rule;
// Graph of imports.
var imports = new Map();
// Keep a list of found circular dependencies to avoid showing them twice.
var found = new Set();
var nodeModulesRe = new RegExp("\\" + path_1.sep + "node_modules\\" + path_1.sep);
var NoCircularImportsWalker = /** @class */ (function (_super) {
    __extends(NoCircularImportsWalker, _super);
    function NoCircularImportsWalker(sourceFile, options, program) {
        var _this = _super.call(this, sourceFile, options) || this;
        _this.program = program;
        return _this;
    }
    NoCircularImportsWalker.prototype.visitSourceFile = function (sourceFile) {
        var _this = this;
        // Instead of visiting all children, this is faster. We know imports are statements anyway.
        sourceFile.statements.forEach(function (statement) {
            // export declarations seem to be missing from the current SyntaxWalker
            if (ts.isExportDeclaration(statement)) {
                _this.visitImportOrExportDeclaration(statement);
            }
            else if (ts.isImportDeclaration(statement)) {
                _this.visitImportOrExportDeclaration(statement);
            }
        });
        var fileName = sourceFile.fileName;
        // Check for cycles, remove any cycles that have been found already (otherwise we'll report
        // false positive on every files that import from the real cycles, and users will be driven
        // mad).
        // The checkCycle is many order of magnitude faster than getCycle, but does not keep a history
        // of the cycle itself. Only get the full cycle if we found one.
        if (this.checkCycle(fileName)) {
            var allCycles = this.getAllCycles(fileName);
            var _loop_1 = function (maybeCycle) {
                // Slice the array so we don't match this file twice.
                if (maybeCycle.slice(1, -1).some(function (fileName) { return found.has(fileName); })) {
                    return "continue";
                }
                maybeCycle.forEach(function (x) { return found.add(x); });
                var node = imports.get(fileName).get(maybeCycle[1]);
                var compilerOptions = this_1.program.getCompilerOptions();
                this_1.addFailureAt(node.getStart(), node.getWidth(), Rule.FAILURE_STRING + ": " + maybeCycle
                    .concat(fileName)
                    .map(function (x) { return path_1.relative(compilerOptions.rootDir || process.cwd(), x); })
                    .join(' -> '));
            };
            var this_1 = this;
            try {
                for (var allCycles_1 = __values(allCycles), allCycles_1_1 = allCycles_1.next(); !allCycles_1_1.done; allCycles_1_1 = allCycles_1.next()) {
                    var maybeCycle = allCycles_1_1.value;
                    _loop_1(maybeCycle);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (allCycles_1_1 && !allCycles_1_1.done && (_a = allCycles_1.return)) _a.call(allCycles_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        var e_1, _a;
    };
    NoCircularImportsWalker.prototype.visitImportOrExportDeclaration = function (node) {
        if (!node.parent || !ts.isSourceFile(node.parent)) {
            return;
        }
        if (!node.moduleSpecifier) {
            return;
        }
        var fileName = node.parent.fileName;
        if (!ts.isStringLiteral(node.moduleSpecifier)) {
            return;
        }
        var importFileName = node.moduleSpecifier.text;
        var compilerOptions = this.program.getCompilerOptions();
        var resolved = ts.resolveModuleName(importFileName, fileName, compilerOptions, ts.sys);
        if (!resolved || !resolved.resolvedModule) {
            return;
        }
        var resolvedImportFileName = resolved.resolvedModule.resolvedFileName;
        // Skip node modules entirely. We use this after resolution to support path mapping in the
        // tsconfig.json (which could override imports from/to node_modules).
        if (nodeModulesRe.test(resolvedImportFileName)) {
            return;
        }
        this.addToGraph(fileName, resolvedImportFileName, node);
    };
    NoCircularImportsWalker.prototype.addToGraph = function (thisFileName, importCanonicalName, node) {
        var i = imports.get(thisFileName);
        if (!i) {
            imports.set(thisFileName, i = new Map);
        }
        i.set(importCanonicalName, node);
    };
    NoCircularImportsWalker.prototype.checkCycle = function (moduleName) {
        var accumulator = new Set();
        var moduleImport = imports.get(moduleName);
        if (!moduleImport)
            return false;
        var toCheck = Array.from(moduleImport.keys());
        for (var i = 0; i < toCheck.length; i++) {
            var current = toCheck[i];
            if (current == moduleName) {
                return true;
            }
            accumulator.add(current);
            toCheck.push.apply(toCheck, __spread(Array.from((imports.get(current) || new Map).keys())
                .filter(function (i) { return !accumulator.has(i); })));
        }
        return false;
    };
    NoCircularImportsWalker.prototype.getAllCycles = function (moduleName, accumulator) {
        if (accumulator === void 0) { accumulator = []; }
        var moduleImport = imports.get(moduleName);
        if (!moduleImport)
            return [];
        if (accumulator.indexOf(moduleName) !== -1)
            return [accumulator];
        var all = [];
        try {
            for (var _a = __values(Array.from(moduleImport.keys())), _b = _a.next(); !_b.done; _b = _a.next()) {
                var imp = _b.value;
                var c = this.getAllCycles(imp, accumulator.concat(moduleName));
                if (c.length)
                    all.push.apply(all, __spread(c));
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return all;
        var e_2, _c;
    };
    return NoCircularImportsWalker;
}(Lint.RuleWalker));
var _a;
