// ts-morph/dist/compiler/ast/doc/base/JSDocPropertyLikeTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
function JSDocPropertyLikeTag(Base) {
    return /** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        // todo: more methods
        class_1.prototype.getTypeExpression = function () {
            return this._getNodeFromCompilerNodeIfExists(this.compilerNode.typeExpression);
        };
        class_1.prototype.getName = function () {
            return this.getNameNode().getText();
        };
        class_1.prototype.getNameNode = function () {
            return this._getNodeFromCompilerNode(this.compilerNode.name);
        };
        class_1.prototype.isBracketed = function () {
            return this.compilerNode.isBracketed;
        };
        return class_1;
    }(Base));
}
exports.JSDocPropertyLikeTag = JSDocPropertyLikeTag;
