error[E0532]: expected tuple struct/variant, found struct `Empty1`
  --> $DIR/empty-struct-braces-pat-2.rs:15:9
   |
LL |         Empty1() => () //~ ERROR expected tuple struct/variant, found struct `Empty1`
   |         ^^^^^^
   |         |
   |         did you mean `XEmpty6`?
   |         did you mean `Empty1 { /* fields */ }`?

error[E0532]: expected tuple struct/variant, found struct `XEmpty1`
  --> $DIR/empty-struct-braces-pat-2.rs:18:9
   |
LL |         XEmpty1() => () //~ ERROR expected tuple struct/variant, found struct `XEmpty1`
   |         ^^^^^^^
   |         |
   |         did you mean `XEmpty6`?
   |         did you mean `XEmpty1 { /* fields */ }`?

error[E0532]: expected tuple struct/variant, found struct `Empty1`
  --> $DIR/empty-struct-braces-pat-2.rs:21:9
   |
LL |         Empty1(..) => () //~ ERROR expected tuple struct/variant, found struct `Empty1`
   |         ^^^^^^
   |         |
   |         did you mean `XEmpty6`?
   |         did you mean `Empty1 { /* fields */ }`?

error[E0532]: expected tuple struct/variant, found struct `XEmpty1`
  --> $DIR/empty-struct-braces-pat-2.rs:24:9
   |
LL |         XEmpty1(..) => () //~ ERROR expected tuple struct/variant, found struct `XEmpty1`
   |         ^^^^^^^
   |         |
   |         did you mean `XEmpty6`?
   |         did you mean `XEmpty1 { /* fields */ }`?

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0532`.
