error: no path from `WillChange` to `TypeOfItem`
  --> $DIR/dep-graph-struct-signature.rs:27:5
   |
LL |     #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `AssociatedItems`
  --> $DIR/dep-graph-struct-signature.rs:28:5
   |
LL |     #[rustc_then_this_would_need(AssociatedItems)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `TraitDefOfItem`
  --> $DIR/dep-graph-struct-signature.rs:29:5
   |
LL |     #[rustc_then_this_would_need(TraitDefOfItem)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:35:5
   |
LL |     #[rustc_then_this_would_need(FnSignature)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:36:5
   |
LL |     #[rustc_then_this_would_need(TypeckTables)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:39:5
   |
LL |     #[rustc_then_this_would_need(FnSignature)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:40:5
   |
LL |     #[rustc_then_this_would_need(TypeckTables)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:45:5
   |
LL |     #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:52:5
   |
LL |     #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR OK
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:60:9
   |
LL |         #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:62:9
   |
LL |         #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `TypeOfItem`
  --> $DIR/dep-graph-struct-signature.rs:67:5
   |
LL |     #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `TypeOfItem`
  --> $DIR/dep-graph-struct-signature.rs:74:5
   |
LL |     #[rustc_then_this_would_need(TypeOfItem)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `FnSignature`
  --> $DIR/dep-graph-struct-signature.rs:80:5
   |
LL |     #[rustc_then_this_would_need(FnSignature)] //~ ERROR no path
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `FnSignature`
  --> $DIR/dep-graph-struct-signature.rs:83:5
   |
LL |     #[rustc_then_this_would_need(FnSignature)] //~ ERROR no path from `WillChange`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `TypeckTables`
  --> $DIR/dep-graph-struct-signature.rs:84:5
   |
LL |     #[rustc_then_this_would_need(TypeckTables)] //~ ERROR no path from `WillChange`
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:31:9
   |
LL |         #[rustc_then_this_would_need(FnSignature)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: no path from `WillChange` to `FnSignature`
  --> $DIR/dep-graph-struct-signature.rs:76:9
   |
LL |         #[rustc_then_this_would_need(FnSignature)] //~ ERROR no path
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:47:9
   |
LL |         #[rustc_then_this_would_need(FnSignature)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:48:9
   |
LL |         #[rustc_then_this_would_need(TypeckTables)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:54:9
   |
LL |         #[rustc_then_this_would_need(FnSignature)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: OK
  --> $DIR/dep-graph-struct-signature.rs:55:9
   |
LL |         #[rustc_then_this_would_need(TypeckTables)] //~ ERROR OK
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 22 previous errors

