error[E0659]: `Ambiguous` is ambiguous (name vs any other name during import resolution)
  --> $DIR/edition-imports-virtual-2015-ambiguity.rs:15:9
   |
LL |         edition_imports_2015::gen_ambiguous!(); //~ ERROR `Ambiguous` is ambiguous
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous name
   |
note: `Ambiguous` could refer to the struct defined here
  --> $DIR/edition-imports-virtual-2015-ambiguity.rs:9:1
   |
LL | pub struct Ambiguous {}
   | ^^^^^^^^^^^^^^^^^^^^^^^
   = help: use `crate::Ambiguous` to refer to this struct unambiguously
note: `Ambiguous` could also refer to the struct defined here
  --> $DIR/edition-imports-virtual-2015-ambiguity.rs:12:5
   |
LL |     pub struct Ambiguous {}
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   = help: use `self::Ambiguous` to refer to this struct unambiguously
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0659]: `edition_imports_2015` is ambiguous (name in the crate root vs extern crate during absolute path resolution)
  --> $DIR/edition-imports-virtual-2015-ambiguity.rs:15:9
   |
LL |         edition_imports_2015::gen_ambiguous!(); //~ ERROR `Ambiguous` is ambiguous
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous name
   |
   = note: `edition_imports_2015` could refer to an extern crate passed with `--extern`
note: `edition_imports_2015` could also refer to the module defined here
  --> $DIR/edition-imports-virtual-2015-ambiguity.rs:5:1
   |
LL | / mod edition_imports_2015 {
LL | |     pub struct Path;
LL | | }
   | |_^
   = help: use `crate::edition_imports_2015` to refer to this module unambiguously
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0659`.
