error[E0308]: mismatched types
  --> $DIR/consider-removing-last-semi.rs:11:18
   |
LL |   fn f() -> String {  //~ ERROR mismatched types
   |  __________________^
LL | |     0u8;
LL | |     "bla".to_string();
   | |                      - help: consider removing this semicolon
LL | | }
   | |_^ expected struct `std::string::String`, found ()
   |
   = note: expected type `std::string::String`
              found type `()`

error[E0308]: mismatched types
  --> $DIR/consider-removing-last-semi.rs:16:18
   |
LL |   fn g() -> String {  //~ ERROR mismatched types
   |  __________________^
LL | |     "this won't work".to_string();
LL | |     "removeme".to_string();
   | |                           - help: consider removing this semicolon
LL | | }
   | |_^ expected struct `std::string::String`, found ()
   |
   = note: expected type `std::string::String`
              found type `()`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
