error[E0282]: type annotations needed
  --> $DIR/pat-tuple-bad-type.rs:5:9
   |
LL |     let x;
   |         - consider giving `x` a type
...
LL |         (..) => {} //~ ERROR type annotations needed
   |         ^^^^ cannot infer type
   |
   = note: type must be known at this point

error[E0308]: mismatched types
  --> $DIR/pat-tuple-bad-type.rs:10:9
   |
LL |         (..) => {} //~ ERROR mismatched types
   |         ^^^^ expected u8, found ()
   |
   = note: expected type `u8`
              found type `()`

error: aborting due to 2 previous errors

Some errors occurred: E0282, E0308.
For more information about an error, try `rustc --explain E0282`.
