error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:9:9
   |
LL |         [84, 69, 83, 84] => (), //~ ERROR unreachable pattern
   |         ^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/slice-pattern-const.rs:1:9
   |
LL | #![deny(unreachable_patterns)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:15:9
   |
LL |         [84, 69, 83, 84] => (), //~ ERROR unreachable pattern
   |         ^^^^^^^^^^^^^^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:21:9
   |
LL |         MAGIC_TEST => (), //~ ERROR unreachable pattern
   |         ^^^^^^^^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:28:9
   |
LL |         FOO => (), //~ ERROR unreachable pattern
   |         ^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:35:9
   |
LL |         BAR => (), //~ ERROR unreachable pattern
   |         ^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:43:9
   |
LL |         BOO => (), //~ ERROR unreachable pattern
   |         ^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:44:9
   |
LL |         b"" => (), //~ ERROR unreachable pattern
   |         ^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const.rs:45:9
   |
LL |         _ => (), //~ ERROR unreachable pattern
   |         ^

error: aborting due to 8 previous errors

