error: unreachable pattern
  --> $DIR/slice-pattern-const-2.rs:9:9
   |
LL |         [4, 5, 6, 7] => (), //~ ERROR unreachable pattern
   |         ^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/slice-pattern-const-2.rs:1:9
   |
LL | #![deny(unreachable_patterns)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const-2.rs:15:9
   |
LL |         [4, 5, 6, 7] => (), //~ ERROR unreachable pattern
   |         ^^^^^^^^^^^^

error: unreachable pattern
  --> $DIR/slice-pattern-const-2.rs:28:9
   |
LL |         FOO => (), //~ ERROR unreachable pattern
   |         ^^^

error: aborting due to 3 previous errors

