// ts-simple-ast/dist/utils/tsconfig/getTsParseConfigHost.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var compiler_1 = require("../compiler");
function getTsParseConfigHost(fileSystemWrapper, options) {
    var directories = [];
    var useCaseSensitiveFileNames = false; // shouldn't this be true? (it was false like this in the compiler)
    var host = {
        useCaseSensitiveFileNames: useCaseSensitiveFileNames,
        readDirectory: function (rootDir, extensions, excludes, includes, depth) {
            var result = compiler_1.readDirectory(fileSystemWrapper, useCaseSensitiveFileNames, rootDir, extensions, excludes, includes, depth);
            directories.push.apply(directories, tslib_1.__spread(result.directories));
            return result.files;
        },
        fileExists: function (path) { return fileSystemWrapper.fileExistsSync(path); },
        readFile: function (path) { return fileSystemWrapper.readFileSync(path, options.encoding); },
        getDirectories: function () { return tslib_1.__spread(directories); },
        clearDirectories: function () { return directories.length = 0; }
    };
    return host;
}
exports.getTsParseConfigHost = getTsParseConfigHost;
