// ts-simple-ast/dist/options/SettingsContainer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../utils");
var SettingsContainer = /** @class */ (function () {
    /** @internal */
    function SettingsContainer(defaultSettings) {
        this.defaultSettings = utils_1.ObjectUtils.assign({}, defaultSettings);
        this.settings = defaultSettings;
    }
    /**
     * Resets the settings to the default.
     */
    SettingsContainer.prototype.reset = function () {
        this.settings = utils_1.ObjectUtils.assign({}, this.defaultSettings);
        this.fireModified();
    };
    /**
     * Gets a copy of the settings as an object.
     */
    SettingsContainer.prototype.get = function () {
        return utils_1.ObjectUtils.assign({}, this.settings);
    };
    /**
     * Sets one or all of the settings.
     * @param settings - Settings to set.
     */
    SettingsContainer.prototype.set = function (settings) {
        utils_1.ObjectUtils.assign(this.settings, settings);
        this.fireModified();
    };
    /**
     * Subscribe to modifications in the settings container.
     * @param action - Action to execute when the settings change.
     * @internal
     */
    SettingsContainer.prototype.onModified = function (action) {
        if (this._modifiedEventContainer == null)
            this._modifiedEventContainer = new utils_1.EventContainer();
        this._modifiedEventContainer.subscribe(action);
    };
    /** @internal */
    SettingsContainer.prototype.fireModified = function () {
        if (this._modifiedEventContainer != null)
            this._modifiedEventContainer.fire(undefined);
    };
    return SettingsContainer;
}());
exports.SettingsContainer = SettingsContainer;
