error[E0308]: mismatched types
  --> $DIR/coercion-missing-tail-expected-type.rs:3:24
   |
LL | fn plus_one(x: i32) -> i32 { //~ ERROR mismatched types
   |    --------            ^^^ expected i32, found ()
   |    |
   |    this function's body doesn't return
LL |     x + 1;
   |          - help: consider removing this semicolon
   |
   = note: expected type `i32`
              found type `()`

error[E0308]: mismatched types
  --> $DIR/coercion-missing-tail-expected-type.rs:7:13
   |
LL | fn foo() -> Result<u8, u64> { //~ ERROR mismatched types
   |    ---      ^^^^^^^^^^^^^^^ expected enum `std::result::Result`, found ()
   |    |
   |    this function's body doesn't return
LL |     Ok(1);
   |          - help: consider removing this semicolon
   |
   = note: expected type `std::result::Result<u8, u64>`
              found type `()`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
