error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:8:5
   |
LL | fn borrowed_receiver(x: &Foo) {
   |                         ---- use `&mut Foo` here to make mutable
LL |     x.borrowed();
LL |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error[E0596]: cannot borrow immutable `Box` content `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:18:5
   |
LL | fn owned_receiver(x: Box<Foo>) {
   |                   - help: make this binding mutable: `mut x`
LL |     x.borrowed();
LL |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0596`.
