error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:63:24
   |
LL | fn deref_mut_field1(x: Own<Point>) {
   |                     - consider changing this to `mut x`
LL |     let __isize = &mut x.y; //~ ERROR cannot borrow
   |                        ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:75:10
   |
LL | fn deref_extend_mut_field1(x: &Own<Point>) -> &mut isize {
   |                               ----------- use `&mut Own<Point>` here to make mutable
LL |     &mut x.y //~ ERROR cannot borrow
   |          ^ cannot borrow as mutable

error[E0499]: cannot borrow `*x` as mutable more than once at a time
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:88:19
   |
LL |     let _x = &mut x.x;
   |                   - first mutable borrow occurs here
LL |     let _y = &mut x.y; //~ ERROR cannot borrow
   |                   ^ second mutable borrow occurs here
LL | }
   | - first borrow ends here

error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:98:5
   |
LL | fn assign_field1<'a>(x: Own<Point>) {
   |                      - consider changing this to `mut x`
LL |     x.y = 3; //~ ERROR cannot borrow
   |     ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:102:5
   |
LL | fn assign_field2<'a>(x: &'a Own<Point>) {
   |                         -------------- use `&'a mut Own<Point>` here to make mutable
LL |     x.y = 3; //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error[E0499]: cannot borrow `*x` as mutable more than once at a time
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:111:5
   |
LL |     let _p: &mut Point = &mut **x;
   |                                -- first mutable borrow occurs here
LL |     x.y = 3; //~ ERROR cannot borrow
   |     ^ second mutable borrow occurs here
LL | }
   | - first borrow ends here

error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:119:5
   |
LL | fn deref_mut_method1(x: Own<Point>) {
   |                      - consider changing this to `mut x`
LL |     x.set(0, 0); //~ ERROR cannot borrow
   |     ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:131:5
   |
LL | fn deref_extend_mut_method1(x: &Own<Point>) -> &mut isize {
   |                                ----------- use `&mut Own<Point>` here to make mutable
LL |     x.y_mut() //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:139:6
   |
LL | fn assign_method1<'a>(x: Own<Point>) {
   |                       - consider changing this to `mut x`
LL |     *x.y_mut() = 3; //~ ERROR cannot borrow
   |      ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-auto-deref-mut.rs:143:6
   |
LL | fn assign_method2<'a>(x: &'a Own<Point>) {
   |                          -------------- use `&'a mut Own<Point>` here to make mutable
LL |     *x.y_mut() = 3; //~ ERROR cannot borrow
   |      ^ cannot borrow as mutable

error: aborting due to 10 previous errors

Some errors occurred: E0499, E0596.
For more information about an error, try `rustc --explain E0499`.
