error[E0596]: cannot borrow immutable item `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:19:5
   |
LL | fn borrowed_receiver(x: &Foo) {
   |                         ---- help: consider changing this to be a mutable reference: `&mut Foo`
LL |     x.borrowed();
LL |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ `x` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `*x` as mutable
  --> $DIR/borrowck-object-mutability.rs:29:5
   |
LL | fn owned_receiver(x: Box<Foo>) {
   |                   - help: consider changing this to be mutable: `mut x`
LL |     x.borrowed();
LL |     x.borrowed_mut(); //~ ERROR cannot borrow
   |     ^ cannot borrow as mutable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0596`.
