error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-deref-mut.rs:29:25
   |
LL | fn deref_mut1(x: Own<isize>) {
   |               - help: make this binding mutable: `mut x`
LL |     let __isize = &mut *x; //~ ERROR cannot borrow
   |                         ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-deref-mut.rs:41:11
   |
LL | fn deref_extend_mut1<'a>(x: &'a Own<isize>) -> &'a mut isize {
   |                             -------------- use `&'a mut Own<isize>` here to make mutable
LL |     &mut **x //~ ERROR cannot borrow
   |           ^^ cannot borrow as mutable

error[E0596]: cannot borrow immutable argument `x` as mutable
  --> $DIR/borrowck-borrow-overloaded-deref-mut.rs:49:6
   |
LL | fn assign1<'a>(x: Own<isize>) {
   |                - help: make this binding mutable: `mut x`
LL |     *x = 3; //~ ERROR cannot borrow
   |      ^ cannot borrow mutably

error[E0596]: cannot borrow immutable borrowed content `*x` as mutable
  --> $DIR/borrowck-borrow-overloaded-deref-mut.rs:53:6
   |
LL | fn assign2<'a>(x: &'a Own<isize>) {
   |                   -------------- use `&'a mut Own<isize>` here to make mutable
LL |     **x = 3; //~ ERROR cannot borrow
   |      ^^ cannot borrow as mutable

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0596`.
