error: expected one of `:` or `@`, found `<`
  --> $DIR/issue-34264.rs:1:14
   |
LL | fn foo(Option<i32>, String) {} //~ ERROR expected one of
   |              ^ expected one of `:` or `@` here

error: expected one of `:` or `@`, found `)`
  --> $DIR/issue-34264.rs:1:27
   |
LL | fn foo(Option<i32>, String) {} //~ ERROR expected one of
   |                           ^ expected one of `:` or `@` here

error: expected one of `:` or `@`, found `,`
  --> $DIR/issue-34264.rs:3:9
   |
LL | fn bar(x, y: usize) {} //~ ERROR expected one of
   |         ^ expected one of `:` or `@` here

error[E0061]: this function takes 2 parameters but 3 parameters were supplied
  --> $DIR/issue-34264.rs:7:5
   |
LL | fn foo(Option<i32>, String) {} //~ ERROR expected one of
   | --------------------------- defined here
...
LL |     foo(Some(42), 2, ""); //~ ERROR this function takes
   |     ^^^^^^^^^^^^^^^^^^^^ expected 2 parameters

error[E0308]: mismatched types
  --> $DIR/issue-34264.rs:8:13
   |
LL |     bar("", ""); //~ ERROR mismatched types
   |             ^^ expected usize, found reference
   |
   = note: expected type `usize`
              found type `&'static str`

error[E0061]: this function takes 2 parameters but 3 parameters were supplied
  --> $DIR/issue-34264.rs:10:5
   |
LL | fn bar(x, y: usize) {} //~ ERROR expected one of
   | ------------------- defined here
...
LL |     bar(1, 2, 3); //~ ERROR this function takes
   |     ^^^^^^^^^^^^ expected 2 parameters

error: aborting due to 6 previous errors

Some errors occurred: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
