warning: unused variable: `theOtherTwo`
  --> $DIR/issue-24690.rs:23:9
   |
LL |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^ help: consider using `_theOtherTwo` instead
   |
note: lint level defined here
  --> $DIR/issue-24690.rs:18:9
   |
LL | #![warn(unused)]
   |         ^^^^^^
   = note: #[warn(unused_variables)] implied by #[warn(unused)]

warning: variable `theTwo` should have a snake case name such as `the_two`
  --> $DIR/issue-24690.rs:22:9
   |
LL |     let theTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^
   |
   = note: #[warn(non_snake_case)] on by default

warning: variable `theOtherTwo` should have a snake case name such as `the_other_two`
  --> $DIR/issue-24690.rs:23:9
   |
LL |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^

error: compilation successful
  --> $DIR/issue-24690.rs:21:1
   |
LL | / fn main() { //~ ERROR compilation successful
LL | |     let theTwo = 2; //~ WARN should have a snake case name
LL | |     let theOtherTwo = 2; //~ WARN should have a snake case name
LL | |     //~^ WARN unused variable
LL | |     println!("{}", theTwo);
LL | | }
   | |_^

error: aborting due to previous error

