error[E0658]: The attribute `foo` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/issue-36530.rs:3:3
   |
LL | #[foo] //~ ERROR is currently unknown to the compiler
   |   ^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error[E0658]: non-builtin inner attributes are unstable (see issue #54726)
  --> $DIR/issue-36530.rs:5:5
   |
LL |     #![foo] //~ ERROR is currently unknown to the compiler
   |     ^^^^^^^
   |
   = help: add #![feature(custom_inner_attributes)] to the crate attributes to enable

error[E0658]: The attribute `foo` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/issue-36530.rs:5:8
   |
LL |     #![foo] //~ ERROR is currently unknown to the compiler
   |        ^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
