warning: unused variable: `theOtherTwo`
  --> $DIR/issue-24690.rs:13:9
   |
LL |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^ help: consider using `_theOtherTwo` instead
   |
note: lint level defined here
  --> $DIR/issue-24690.rs:8:9
   |
LL | #![warn(unused)]
   |         ^^^^^^
   = note: #[warn(unused_variables)] implied by #[warn(unused)]

warning: variable `theTwo` should have a snake case name such as `the_two`
  --> $DIR/issue-24690.rs:12:9
   |
LL |     let theTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^
   |
   = note: #[warn(non_snake_case)] on by default

warning: variable `theOtherTwo` should have a snake case name such as `the_other_two`
  --> $DIR/issue-24690.rs:13:9
   |
LL |     let theOtherTwo = 2; //~ WARN should have a snake case name
   |         ^^^^^^^^^^^

