error[E0061]: this function takes 1 parameter but 0 parameters were supplied
  --> $DIR/missing-unit-argument.rs:21:33
   |
LL |     let _: Result<(), String> = Ok(); //~ ERROR this function takes
   |                                 ^^^^
help: expected the unit value `()`; create it with empty parentheses
   |
LL |     let _: Result<(), String> = Ok(()); //~ ERROR this function takes
   |                                    ^^

error[E0061]: this function takes 2 parameters but 0 parameters were supplied
  --> $DIR/missing-unit-argument.rs:22:5
   |
LL | fn foo(():(), ():()) {}
   | -------------------- defined here
...
LL |     foo(); //~ ERROR this function takes
   |     ^^^^^ expected 2 parameters

error[E0061]: this function takes 2 parameters but 1 parameter was supplied
  --> $DIR/missing-unit-argument.rs:23:5
   |
LL | fn foo(():(), ():()) {}
   | -------------------- defined here
...
LL |     foo(()); //~ ERROR this function takes
   |     ^^^^^^^ expected 2 parameters

error[E0061]: this function takes 1 parameter but 0 parameters were supplied
  --> $DIR/missing-unit-argument.rs:24:5
   |
LL | fn bar(():()) {}
   | ------------- defined here
...
LL |     bar(); //~ ERROR this function takes
   |     ^^^^^
help: expected the unit value `()`; create it with empty parentheses
   |
LL |     bar(()); //~ ERROR this function takes
   |         ^^

error[E0061]: this function takes 1 parameter but 0 parameters were supplied
  --> $DIR/missing-unit-argument.rs:25:7
   |
LL |     fn baz(self, (): ()) { }
   |     -------------------- defined here
...
LL |     S.baz(); //~ ERROR this function takes
   |       ^^^
help: expected the unit value `()`; create it with empty parentheses
   |
LL |     S.baz(()); //~ ERROR this function takes
   |           ^^

error[E0061]: this function takes 1 parameter but 0 parameters were supplied
  --> $DIR/missing-unit-argument.rs:26:7
   |
LL |     fn generic<T>(self, _: T) { }
   |     ------------------------- defined here
...
LL |     S.generic::<()>(); //~ ERROR this function takes
   |       ^^^^^^^
help: expected the unit value `()`; create it with empty parentheses
   |
LL |     S.generic::<()>(()); //~ ERROR this function takes
   |                     ^^

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0061`.
