error[E0596]: cannot borrow immutable item `*a` as mutable
  --> $DIR/mut-arg-hint.rs:13:9
   |
LL |     fn foo(mut a: &String) {
   |                   ------- help: consider changing this to be a mutable reference: `&mut std::string::String`
LL |         a.push_str("bar"); //~ ERROR cannot borrow immutable borrowed content
   |         ^ `a` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `*a` as mutable
  --> $DIR/mut-arg-hint.rs:18:5
   |
LL | pub fn foo<'a>(mut a: &'a String) {
   |                       ---------- help: consider changing this to be a mutable reference: `&mut std::string::String`
LL |     a.push_str("foo"); //~ ERROR cannot borrow immutable borrowed content
   |     ^ `a` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `*a` as mutable
  --> $DIR/mut-arg-hint.rs:25:9
   |
LL |     pub fn foo(mut a: &String) {
   |                       ------- help: consider changing this to be a mutable reference: `&mut std::string::String`
LL |         a.push_str("foo"); //~ ERROR cannot borrow immutable borrowed content
   |         ^ `a` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0596`.
