error[E0597]: `z` does not live long enough
  --> $DIR/send-is-not-static-std-sync.rs:26:34
   |
LL |         *lock.lock().unwrap() = &z;
   |                                  ^ borrowed value does not live long enough
LL |     }
   |     - `z` dropped here while still borrowed
...
LL | }
   | - borrowed value needs to live until here

error[E0505]: cannot move out of `y` because it is borrowed
  --> $DIR/send-is-not-static-std-sync.rs:23:10
   |
LL |     *lock.lock().unwrap() = &*y;
   |                              -- borrow of `*y` occurs here
LL |     drop(y); //~ ERROR cannot move out
   |          ^ move out of `y` occurs here

error[E0597]: `z` does not live long enough
  --> $DIR/send-is-not-static-std-sync.rs:40:35
   |
LL |         *lock.write().unwrap() = &z;
   |                                   ^ borrowed value does not live long enough
LL |     }
   |     - `z` dropped here while still borrowed
...
LL | }
   | - borrowed value needs to live until here

error[E0505]: cannot move out of `y` because it is borrowed
  --> $DIR/send-is-not-static-std-sync.rs:37:10
   |
LL |     *lock.write().unwrap() = &*y;
   |                               -- borrow of `*y` occurs here
LL |     drop(y); //~ ERROR cannot move out
   |          ^ move out of `y` occurs here

error[E0597]: `z` does not live long enough
  --> $DIR/send-is-not-static-std-sync.rs:56:18
   |
LL |         tx.send(&z).unwrap();
   |                  ^ borrowed value does not live long enough
LL |     }
   |     - `z` dropped here while still borrowed
...
LL | }
   | - borrowed value needs to live until here

error[E0505]: cannot move out of `y` because it is borrowed
  --> $DIR/send-is-not-static-std-sync.rs:53:10
   |
LL |     tx.send(&*y);
   |              -- borrow of `*y` occurs here
LL |     drop(y); //~ ERROR cannot move out
   |          ^ move out of `y` occurs here

error: aborting due to 6 previous errors

Some errors occurred: E0505, E0597.
For more information about an error, try `rustc --explain E0505`.
