error: unexpected generic arguments in path
  --> $DIR/visibility-ty-params.rs:16:5
   |
LL | m!{ S<u8> } //~ ERROR unexpected generic arguments in path
   |     ^^^^^

error: unexpected generic arguments in path
  --> $DIR/visibility-ty-params.rs:19:9
   |
LL |     m!{ m<> } //~ ERROR unexpected generic arguments in path
   |         ^^^

error: aborting due to 2 previous errors

