error[E0416]: identifier `a` is bound more than once in the same pattern
  --> $DIR/shadowing-in-the-same-pattern.rs:3:10
   |
LL | fn f((a, a): (isize, isize)) {} //~ ERROR identifier `a` is bound more than once
   |          ^ used in a pattern more than once

error[E0416]: identifier `a` is bound more than once in the same pattern
  --> $DIR/shadowing-in-the-same-pattern.rs:6:13
   |
LL |     let (a, a) = (1, 1);    //~ ERROR identifier `a` is bound more than once
   |             ^ used in a pattern more than once

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0416`.
