error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:13:10
   |
LL |     foo!(a, 1, { //~ ERROR irrefutable if-let
   |          ^ irrefutable pattern

error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:16:10
   |
LL |     bar!(a, 1, { //~ ERROR irrefutable if-let
   |          ^ irrefutable pattern

error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:22:12
   |
LL |     if let a = 1 { //~ ERROR irrefutable if-let
   |            ^ irrefutable pattern

error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:26:12
   |
LL |     if let a = 1 { //~ ERROR irrefutable if-let
   |            ^ irrefutable pattern

error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:36:19
   |
LL |     } else if let a = 1 { //~ ERROR irrefutable if-let
   |                   ^ irrefutable pattern

error[E0162]: irrefutable if-let pattern
  --> $DIR/if-let.rs:42:19
   |
LL |     } else if let a = 1 { //~ ERROR irrefutable if-let
   |                   ^ irrefutable pattern

error: aborting due to 6 previous errors

For more information about this error, try `rustc --explain E0162`.
