error: must be of the form #[wasm_custom_section = "foo"]
  --> $DIR/malformed.rs:13:1
   |
LL | #[wasm_custom_section] //~ ERROR: must be of the form
   | ^^^^^^^^^^^^^^^^^^^^^^

error: must be of the form #[wasm_custom_section = "foo"]
  --> $DIR/malformed.rs:16:1
   |
LL | #[wasm_custom_section(foo)] //~ ERROR: must be of the form
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

