error: requires at least a format string argument
  --> $DIR/bad-format-args.rs:2:5
   |
LL |     format!(); //~ ERROR requires at least a format string argument
   |     ^^^^^^^^^^
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: expected token: `,`
  --> $DIR/bad-format-args.rs:3:16
   |
LL |     format!("" 1); //~ ERROR expected token: `,`
   |                ^

error: expected token: `,`
  --> $DIR/bad-format-args.rs:4:19
   |
LL |     format!("", 1 1); //~ ERROR expected token: `,`
   |                   ^

error: aborting due to 3 previous errors

