error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:1:22
   |
LL | extern crate core as bäz; //~ ERROR non-ascii idents
   |                      ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:3:5
   |
LL | use föö::bar; //~ ERROR non-ascii idents
   |     ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:5:5
   |
LL | mod föö { //~ ERROR non-ascii idents
   |     ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:9:4
   |
LL | fn bär( //~ ERROR non-ascii idents
   |    ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:10:5
   |
LL |     bäz: isize //~ ERROR non-ascii idents
   |     ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:12:9
   |
LL |     let _ö: isize; //~ ERROR non-ascii idents
   |         ^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:15:10
   |
LL |         (_ä, _) => {} //~ ERROR non-ascii idents
   |          ^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:19:8
   |
LL | struct Föö { //~ ERROR non-ascii idents
   |        ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:20:5
   |
LL |     föö: isize //~ ERROR non-ascii idents
   |     ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:23:6
   |
LL | enum Bär { //~ ERROR non-ascii idents
   |      ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:24:5
   |
LL |     Bäz { //~ ERROR non-ascii idents
   |     ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:25:9
   |
LL |         qüx: isize //~ ERROR non-ascii idents
   |         ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error[E0658]: non-ascii idents are not fully supported. (see issue #55467)
  --> $DIR/feature-gate-non_ascii_idents.rs:30:8
   |
LL |     fn qüx();  //~ ERROR non-ascii idents
   |        ^^^
   |
   = help: add #![feature(non_ascii_idents)] to the crate attributes to enable

error: aborting due to 13 previous errors

For more information about this error, try `rustc --explain E0658`.
