error[E0658]: The `!` type is experimental (see issue #35121)
  --> $DIR/feature-gate-never_type.rs:7:17
   |
LL | type Ma = (u32, !, i32); //~ ERROR type is experimental
   |                 ^
   |
   = help: add #![feature(never_type)] to the crate attributes to enable

error[E0658]: The `!` type is experimental (see issue #35121)
  --> $DIR/feature-gate-never_type.rs:8:20
   |
LL | type Meeshka = Vec<!>; //~ ERROR type is experimental
   |                    ^
   |
   = help: add #![feature(never_type)] to the crate attributes to enable

error[E0658]: The `!` type is experimental (see issue #35121)
  --> $DIR/feature-gate-never_type.rs:9:16
   |
LL | type Mow = &fn(!) -> !; //~ ERROR type is experimental
   |                ^
   |
   = help: add #![feature(never_type)] to the crate attributes to enable

error[E0658]: The `!` type is experimental (see issue #35121)
  --> $DIR/feature-gate-never_type.rs:10:19
   |
LL | type Skwoz = &mut !; //~ ERROR type is experimental
   |                   ^
   |
   = help: add #![feature(never_type)] to the crate attributes to enable

error[E0658]: The `!` type is experimental (see issue #35121)
  --> $DIR/feature-gate-never_type.rs:13:16
   |
LL |     type Wub = !; //~ ERROR type is experimental
   |                ^
   |
   = help: add #![feature(never_type)] to the crate attributes to enable

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0658`.
