error[E0004]: non-exhaustive patterns: `_` not covered
  --> $DIR/feature-gate-precise_pointer_size_matching.rs:6:11
   |
LL |     match 0usize { //~ERROR non-exhaustive patterns: `_` not covered
   |           ^^^^^^ pattern `_` not covered

error[E0004]: non-exhaustive patterns: `_` not covered
  --> $DIR/feature-gate-precise_pointer_size_matching.rs:10:11
   |
LL |     match 0isize { //~ERROR non-exhaustive patterns: `_` not covered
   |           ^^^^^^ pattern `_` not covered

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0004`.
