error[E0379]: trait fns cannot be declared const
  --> $DIR/feature-gate-const_fn.rs:6:5
   |
LL |     const fn foo() -> u32; //~ ERROR const fn is unstable
   |     ^^^^^ trait fns cannot be const

error[E0379]: trait fns cannot be declared const
  --> $DIR/feature-gate-const_fn.rs:8:5
   |
LL |     const fn bar() -> u32 { 0 } //~ ERROR const fn is unstable
   |     ^^^^^ trait fns cannot be const

error[E0379]: trait fns cannot be declared const
  --> $DIR/feature-gate-const_fn.rs:17:5
   |
LL |     const fn foo() -> u32 { 0 } //~ ERROR trait fns cannot be declared const
   |     ^^^^^ trait fns cannot be const

error[E0658]: const fn is unstable (see issue #24111)
  --> $DIR/feature-gate-const_fn.rs:6:5
   |
LL |     const fn foo() -> u32; //~ ERROR const fn is unstable
   |     ^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(const_fn)] to the crate attributes to enable

error[E0658]: const fn is unstable (see issue #24111)
  --> $DIR/feature-gate-const_fn.rs:8:5
   |
LL |     const fn bar() -> u32 { 0 } //~ ERROR const fn is unstable
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(const_fn)] to the crate attributes to enable

error: aborting due to 5 previous errors

Some errors occurred: E0379, E0658.
For more information about an error, try `rustc --explain E0379`.
