error[E0658]: `Ptr<Self>` cannot be used as the type of `self` without the `arbitrary_self_types` feature (see issue #44874)
  --> $DIR/feature-gate-arbitrary-self-types.rs:16:18
   |
LL |     fn foo(self: Ptr<Self>); //~ ERROR `Ptr<Self>` cannot be used as the type of `self` without
   |                  ^^^^^^^^^
   |
   = help: add #![feature(arbitrary_self_types)] to the crate attributes to enable
   = help: consider changing to `self`, `&self`, `&mut self`, or `self: Box<Self>`

error[E0658]: `Ptr<Bar>` cannot be used as the type of `self` without the `arbitrary_self_types` feature (see issue #44874)
  --> $DIR/feature-gate-arbitrary-self-types.rs:22:18
   |
LL |     fn foo(self: Ptr<Self>) {} //~ ERROR `Ptr<Bar>` cannot be used as the type of `self` without
   |                  ^^^^^^^^^
   |
   = help: add #![feature(arbitrary_self_types)] to the crate attributes to enable
   = help: consider changing to `self`, `&self`, `&mut self`, or `self: Box<Self>`

error[E0658]: `std::boxed::Box<Ptr<Bar>>` cannot be used as the type of `self` without the `arbitrary_self_types` feature (see issue #44874)
  --> $DIR/feature-gate-arbitrary-self-types.rs:26:18
   |
LL |     fn bar(self: Box<Ptr<Self>>) {} //~ ERROR `std::boxed::Box<Ptr<Bar>>` cannot be used as the
   |                  ^^^^^^^^^^^^^^
   |
   = help: add #![feature(arbitrary_self_types)] to the crate attributes to enable
   = help: consider changing to `self`, `&self`, `&mut self`, or `self: Box<Self>`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
