error[E0658]: unions with non-`Copy` fields are unstable (see issue #32836)
  --> $DIR/feature-gate-untagged_unions.rs:9:1
   |
LL | / union U3 { //~ ERROR unions with non-`Copy` fields are unstable
LL | |     a: String,
LL | | }
   | |_^
   |
   = help: add #![feature(untagged_unions)] to the crate attributes to enable

error[E0658]: unions with non-`Copy` fields are unstable (see issue #32836)
  --> $DIR/feature-gate-untagged_unions.rs:13:1
   |
LL | / union U4<T> { //~ ERROR unions with non-`Copy` fields are unstable
LL | |     a: T,
LL | | }
   | |_^
   |
   = help: add #![feature(untagged_unions)] to the crate attributes to enable

error[E0658]: unions with `Drop` implementations are unstable (see issue #32836)
  --> $DIR/feature-gate-untagged_unions.rs:17:1
   |
LL | / union U5 { //~ ERROR unions with `Drop` implementations are unstable
LL | |     a: u8,
LL | | }
   | |_^
   |
   = help: add #![feature(untagged_unions)] to the crate attributes to enable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
