error[E0658]: let bindings in constants are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:6:13
   |
LL |     let x = 42;
   |             ^^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: statements in constants are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:6:13
   |
LL |     let x = 42;
   |             ^^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: let bindings in constants are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:3:1
   |
LL | / const FOO: usize = {
LL | |     //~^ ERROR statements in constants are unstable
LL | |     //~| ERROR: let bindings in constants are unstable
LL | |     let x = 42;
...  |
LL | |     42
LL | | };
   | |__^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: statements in constants are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:3:1
   |
LL | / const FOO: usize = {
LL | |     //~^ ERROR statements in constants are unstable
LL | |     //~| ERROR: let bindings in constants are unstable
LL | |     let x = 42;
...  |
LL | |     42
LL | | };
   | |__^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: let bindings in statics are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:15:13
   |
LL |     let x = 42;
   |             ^^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: statements in statics are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:15:13
   |
LL |     let x = 42;
   |             ^^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: let bindings in statics are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:12:1
   |
LL | / static BAR: usize = {
LL | |     //~^ ERROR statements in statics are unstable
LL | |     //~| ERROR: let bindings in statics are unstable
LL | |     let x = 42;
...  |
LL | |     42
LL | | };
   | |__^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error[E0658]: statements in statics are unstable (see issue #48821)
  --> $DIR/feature-gate-const_let.rs:12:1
   |
LL | / static BAR: usize = {
LL | |     //~^ ERROR statements in statics are unstable
LL | |     //~| ERROR: let bindings in statics are unstable
LL | |     let x = 42;
...  |
LL | |     42
LL | | };
   | |__^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0658`.
