error[E0658]: intrinsics are subject to change
  --> $DIR/feature-gate-intrinsics.rs:1:1
   |
LL | / extern "rust-intrinsic" {   //~ ERROR intrinsics are subject to change
LL | |     fn bar();
LL | | }
   | |_^
   |
   = help: add #![feature(intrinsics)] to the crate attributes to enable

error[E0658]: intrinsics are subject to change
  --> $DIR/feature-gate-intrinsics.rs:5:1
   |
LL | / extern "rust-intrinsic" fn baz() {  //~ ERROR intrinsics are subject to change
LL | | }
   | |_^
   |
   = help: add #![feature(intrinsics)] to the crate attributes to enable

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
