error[E0658]: `cfg(target_vendor)` is experimental and subject to change (see issue #29718)
  --> $DIR/feature-gate-cfg-target-vendor.rs:2:12
   |
LL | #[cfg_attr(target_vendor = "x", x)] //~ ERROR `cfg(target_vendor)` is experimental
   |            ^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_vendor)] to the crate attributes to enable

error[E0658]: `cfg(target_vendor)` is experimental and subject to change (see issue #29718)
  --> $DIR/feature-gate-cfg-target-vendor.rs:1:7
   |
LL | #[cfg(target_vendor = "x")] //~ ERROR `cfg(target_vendor)` is experimental
   |       ^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_vendor)] to the crate attributes to enable

error[E0658]: `cfg(target_vendor)` is experimental and subject to change (see issue #29718)
  --> $DIR/feature-gate-cfg-target-vendor.rs:5:19
   |
LL | #[cfg(not(any(all(target_vendor = "x"))))] //~ ERROR `cfg(target_vendor)` is experimental
   |                   ^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_vendor)] to the crate attributes to enable

error[E0658]: `cfg(target_vendor)` is experimental and subject to change (see issue #29718)
  --> $DIR/feature-gate-cfg-target-vendor.rs:9:10
   |
LL |     cfg!(target_vendor = "x");
   |          ^^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(cfg_target_vendor)] to the crate attributes to enable

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0658`.
