error[E0658]: async fn is unstable (see issue #50547)
  --> $DIR/feature-gate-async-await.rs:5:1
   |
LL | async fn foo() {} //~ ERROR async fn is unstable
   | ^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(async_await)] to the crate attributes to enable

error[E0658]: async blocks are unstable (see issue #50547)
  --> $DIR/feature-gate-async-await.rs:8:13
   |
LL |     let _ = async {}; //~ ERROR async blocks are unstable
   |             ^^^^^^^^
   |
   = help: add #![feature(async_await)] to the crate attributes to enable

error[E0658]: async closures are unstable (see issue #50547)
  --> $DIR/feature-gate-async-await.rs:9:13
   |
LL |     let _ = async || {}; //~ ERROR async closures are unstable
   |             ^^^^^^^^^^^
   |
   = help: add #![feature(async_await)] to the crate attributes to enable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
