error[E0631]: type mismatch in closure arguments
  --> $DIR/closure-arg-type-mismatch.rs:3:14
   |
LL |     a.iter().map(|_: (u32, u32)| 45); //~ ERROR type mismatch
   |              ^^^ ------------------ found signature of `fn((u32, u32)) -> _`
   |              |
   |              expected signature of `fn(&(u32, u32)) -> _`

error[E0631]: type mismatch in closure arguments
  --> $DIR/closure-arg-type-mismatch.rs:4:14
   |
LL |     a.iter().map(|_: &(u16, u16)| 45); //~ ERROR type mismatch
   |              ^^^ ------------------- found signature of `for<'r> fn(&'r (u16, u16)) -> _`
   |              |
   |              expected signature of `fn(&(u32, u32)) -> _`

error[E0631]: type mismatch in closure arguments
  --> $DIR/closure-arg-type-mismatch.rs:5:14
   |
LL |     a.iter().map(|_: (u16, u16)| 45); //~ ERROR type mismatch
   |              ^^^ ------------------ found signature of `fn((u16, u16)) -> _`
   |              |
   |              expected signature of `fn(&(u32, u32)) -> _`

error[E0308]: mismatched types
  --> $DIR/closure-arg-type-mismatch.rs:10:5
   |
LL |     baz(f); //~ ERROR mismatched types
   |     ^^^ one type is more general than the other
   |
   = note: expected type `for<'r> std::ops::Fn<(*mut &'r u32,)>`
              found type `std::ops::Fn<(*mut &'a u32,)>`

error[E0308]: mismatched types
  --> $DIR/closure-arg-type-mismatch.rs:10:5
   |
LL |     baz(f); //~ ERROR mismatched types
   |     ^^^ one type is more general than the other
   |
   = note: expected type `std::ops::FnOnce<(*mut &u32,)>`
              found type `std::ops::FnOnce<(*mut &'a u32,)>`

error: aborting due to 5 previous errors

Some errors occurred: E0308, E0631.
For more information about an error, try `rustc --explain E0308`.
