error[E0379]: trait fns cannot be declared const
  --> $DIR/const-fn-in-trait.rs:7:5
   |
LL |     const fn g(); //~ ERROR cannot be declared const
   |     ^^^^^ trait fns cannot be const

error[E0379]: trait fns cannot be declared const
  --> $DIR/const-fn-in-trait.rs:11:5
   |
LL |     const fn f() -> u32 { 22 } //~ ERROR cannot be declared const
   |     ^^^^^ trait fns cannot be const

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0379`.
