error[E0308]: mismatched types
  --> $DIR/issue-38371.rs:4:8
   |
LL | fn foo(&foo: Foo) { //~ ERROR mismatched types
   |        ^^^^ expected struct `Foo`, found reference
   |
   = note: expected type `Foo`
              found type `&_`
   = help: did you mean `foo: &Foo`?

error[E0308]: mismatched types
  --> $DIR/issue-38371.rs:18:9
   |
LL | fn agh(&&bar: &u32) { //~ ERROR mismatched types
   |         ^^^^ expected u32, found reference
   |
   = note: expected type `u32`
              found type `&_`
   = help: did you mean `bar: &u32`?

error[E0308]: mismatched types
  --> $DIR/issue-38371.rs:21:8
   |
LL | fn bgh(&&bar: u32) { //~ ERROR mismatched types
   |        ^^^^^ expected u32, found reference
   |
   = note: expected type `u32`
              found type `&_`

error[E0529]: expected an array or slice, found `u32`
  --> $DIR/issue-38371.rs:24:9
   |
LL | fn ugh(&[bar]: &u32) { //~ ERROR expected an array or slice
   |         ^^^^^ pattern cannot match with input type `u32`

error: aborting due to 4 previous errors

Some errors occurred: E0308, E0529.
For more information about an error, try `rustc --explain E0308`.
