error[E0277]: the trait bound `Self: Tr<U>` is not satisfied
  --> $DIR/type-params-in-different-spaces-2.rs:10:9
   |
LL |         Tr::op(u)   //~ ERROR E0277
   |         ^^^^^^ the trait `Tr<U>` is not implemented for `Self`
   |
   = help: consider adding a `where Self: Tr<U>` bound
note: required by `Tr::op`
  --> $DIR/type-params-in-different-spaces-2.rs:5:5
   |
LL |     fn op(_: T) -> Self;
   |     ^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Self: Tr<U>` is not satisfied
  --> $DIR/type-params-in-different-spaces-2.rs:16:9
   |
LL |         Tr::op(u)   //~ ERROR E0277
   |         ^^^^^^ the trait `Tr<U>` is not implemented for `Self`
   |
   = help: consider adding a `where Self: Tr<U>` bound
note: required by `Tr::op`
  --> $DIR/type-params-in-different-spaces-2.rs:5:5
   |
LL |     fn op(_: T) -> Self;
   |     ^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
