error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:31:7
   |
LL |     &(S: &S); //~ ERROR mismatched types
   |       ^ expected &S, found struct `S`
   |
   = note: expected type `&S`
              found type `S`

error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:35:7
   |
LL |     *(S: Z); //~ ERROR mismatched types
   |       ^ expected struct `Z`, found struct `S`
   |
   = note: expected type `Z`
              found type `S`

error[E0614]: type `Z` cannot be dereferenced
  --> $DIR/type-ascription-precedence.rs:35:5
   |
LL |     *(S: Z); //~ ERROR mismatched types
   |     ^^^^^^^

error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:40:7
   |
LL |     -(S: Z); //~ ERROR mismatched types
   |       ^ expected struct `Z`, found struct `S`
   |
   = note: expected type `Z`
              found type `S`

error[E0600]: cannot apply unary operator `-` to type `Z`
  --> $DIR/type-ascription-precedence.rs:40:5
   |
LL |     -(S: Z); //~ ERROR mismatched types
   |     ^^^^^^^ cannot apply unary operator `-`
   |
   = note: an implementation of `std::ops::Neg` might be missing for `Z`

error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:45:5
   |
LL |     (S + Z): Z; //~ ERROR mismatched types
   |     ^^^^^^^ expected struct `Z`, found struct `S`
   |
   = note: expected type `Z`
              found type `S`

error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:49:5
   |
LL |     (S * Z): Z; //~ ERROR mismatched types
   |     ^^^^^^^ expected struct `Z`, found struct `S`
   |
   = note: expected type `Z`
              found type `S`

error[E0308]: mismatched types
  --> $DIR/type-ascription-precedence.rs:53:5
   |
LL |     (S .. S): S; //~ ERROR mismatched types
   |     ^^^^^^^^ expected struct `S`, found struct `std::ops::Range`
   |
   = note: expected type `S`
              found type `std::ops::Range<S>`

error: aborting due to 8 previous errors

Some errors occurred: E0308, E0600, E0614.
For more information about an error, try `rustc --explain E0308`.
