warning: private trait `m::PrivTr` in public interface (error E0445)
  --> $DIR/private-in-public-assoc-ty.rs:15:5
   |
LL | /     pub trait PubTr {
LL | |         //~^ WARN private trait `m::PrivTr` in public interface
LL | |         //~| WARN this was previously accepted
LL | |         //~| WARN private type `m::Priv` in public interface
...  |
LL | |         //~^ ERROR private type `m::Priv` in public interface
LL | |     }
   | |_____^
   |
   = note: #[warn(private_in_public)] on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #34537 <https://github.com/rust-lang/rust/issues/34537>

warning: private type `m::Priv` in public interface (error E0446)
  --> $DIR/private-in-public-assoc-ty.rs:15:5
   |
LL | /     pub trait PubTr {
LL | |         //~^ WARN private trait `m::PrivTr` in public interface
LL | |         //~| WARN this was previously accepted
LL | |         //~| WARN private type `m::Priv` in public interface
...  |
LL | |         //~^ ERROR private type `m::Priv` in public interface
LL | |     }
   | |_____^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #34537 <https://github.com/rust-lang/rust/issues/34537>

error[E0446]: private type `m::Priv` in public interface
  --> $DIR/private-in-public-assoc-ty.rs:24:9
   |
LL |     struct Priv;
   |     - `m::Priv` declared as private
...
LL |         type Alias4 = Priv;
   |         ^^^^^^^^^^^^^^^^^^^ can't leak private type

error[E0446]: private type `m::Priv` in public interface
  --> $DIR/private-in-public-assoc-ty.rs:28:9
   |
LL |     struct Priv;
   |     - `m::Priv` declared as private
...
LL |         type Alias1 = Priv;
   |         ^^^^^^^^^^^^^^^^^^^ can't leak private type

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0446`.
