error[E0423]: expected function, found trait `Bar`
  --> $DIR/privacy-ns1.rs:20:5
   |
LL |     Bar();  //~ ERROR expected function, found trait `Bar`
   |     ^^^ did you mean `Baz`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0573]: expected type, found function `Bar`
  --> $DIR/privacy-ns1.rs:35:17
   |
LL |     let _x: Box<Bar>;  //~ ERROR expected type, found function `Bar`
   |                 ^^^ did you mean `Baz`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0425]: cannot find function `Bar` in this scope
  --> $DIR/privacy-ns1.rs:50:5
   |
LL |     Bar();  //~ ERROR cannot find function `Bar` in this scope
   |     ^^^ did you mean `Baz`?
help: possible candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0412]: cannot find type `Bar` in this scope
  --> $DIR/privacy-ns1.rs:51:17
   |
LL |     let _x: Box<Bar>;  //~ ERROR cannot find type `Bar` in this scope
   |                 ^^^ did you mean `Baz`?
help: possible candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error: aborting due to 4 previous errors

Some errors occurred: E0412, E0423, E0425, E0573.
For more information about an error, try `rustc --explain E0412`.
