error[E0423]: expected function, found trait `Bar`
  --> $DIR/privacy-ns2.rs:20:5
   |
LL |     Bar(); //~ ERROR expected function, found trait `Bar`
   |     ^^^ not a function
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0423]: expected function, found trait `Bar`
  --> $DIR/privacy-ns2.rs:26:5
   |
LL |     Bar(); //~ ERROR expected function, found trait `Bar`
   |     ^^^ did you mean `Baz`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0573]: expected type, found function `Bar`
  --> $DIR/privacy-ns2.rs:41:18
   |
LL |     let _x : Box<Bar>; //~ ERROR expected type, found function `Bar`
   |                  ^^^ not a type
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0573]: expected type, found function `Bar`
  --> $DIR/privacy-ns2.rs:47:17
   |
LL |     let _x: Box<Bar>; //~ ERROR expected type, found function `Bar`
   |                 ^^^ did you mean `Baz`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use foo1::Bar;
   |
LL | use foo2::Bar;
   |
LL | use foo3::Bar;
   |

error[E0603]: trait `Bar` is private
  --> $DIR/privacy-ns2.rs:60:15
   |
LL |     use foo3::Bar;  //~ ERROR `Bar` is private
   |               ^^^

error[E0603]: trait `Bar` is private
  --> $DIR/privacy-ns2.rs:64:15
   |
LL |     use foo3::Bar;  //~ ERROR `Bar` is private
   |               ^^^

error[E0603]: trait `Bar` is private
  --> $DIR/privacy-ns2.rs:71:16
   |
LL |     use foo3::{Bar,Baz};  //~ ERROR `Bar` is private
   |                ^^^

error: aborting due to 7 previous errors

Some errors occurred: E0423, E0573, E0603.
For more information about an error, try `rustc --explain E0423`.
