error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:51:16
   |
LL |     let a = a::A(()); //~ ERROR tuple struct `A` is private
   |                ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:52:16
   |
LL |     let b = a::B(2); //~ ERROR tuple struct `B` is private
   |                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:53:16
   |
LL |     let c = a::C(2, 3); //~ ERROR tuple struct `C` is private
   |                ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:56:12
   |
LL |     let a::A(()) = a; //~ ERROR tuple struct `A` is private
   |            ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:57:12
   |
LL |     let a::A(_) = a; //~ ERROR tuple struct `A` is private
   |            ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:58:18
   |
LL |     match a { a::A(()) => {} } //~ ERROR tuple struct `A` is private
   |                  ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:59:18
   |
LL |     match a { a::A(_) => {} } //~ ERROR tuple struct `A` is private
   |                  ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:61:12
   |
LL |     let a::B(_) = b; //~ ERROR tuple struct `B` is private
   |            ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:62:12
   |
LL |     let a::B(_b) = b; //~ ERROR tuple struct `B` is private
   |            ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:63:18
   |
LL |     match b { a::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                  ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:64:18
   |
LL |     match b { a::B(_b) => {} } //~ ERROR tuple struct `B` is private
   |                  ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:65:18
   |
LL |     match b { a::B(1) => {} a::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                  ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:65:32
   |
LL |     match b { a::B(1) => {} a::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:68:12
   |
LL |     let a::C(_, _) = c; //~ ERROR tuple struct `C` is private
   |            ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:69:12
   |
LL |     let a::C(_a, _) = c; //~ ERROR tuple struct `C` is private
   |            ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:70:12
   |
LL |     let a::C(_, _b) = c; //~ ERROR tuple struct `C` is private
   |            ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:71:12
   |
LL |     let a::C(_a, _b) = c; //~ ERROR tuple struct `C` is private
   |            ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:72:18
   |
LL |     match c { a::C(_, _) => {} } //~ ERROR tuple struct `C` is private
   |                  ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:73:18
   |
LL |     match c { a::C(_a, _) => {} } //~ ERROR tuple struct `C` is private
   |                  ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:74:18
   |
LL |     match c { a::C(_, _b) => {} } //~ ERROR tuple struct `C` is private
   |                  ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:75:18
   |
LL |     match c { a::C(_a, _b) => {} } //~ ERROR tuple struct `C` is private
   |                  ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:83:17
   |
LL |     let a2 = a::A; //~ ERROR tuple struct `A` is private
   |                 ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:84:17
   |
LL |     let b2 = a::B; //~ ERROR tuple struct `B` is private
   |                 ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:85:17
   |
LL |     let c2 = a::C; //~ ERROR tuple struct `C` is private
   |                 ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:90:20
   |
LL |     let a = other::A(()); //~ ERROR tuple struct `A` is private
   |                    ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:91:20
   |
LL |     let b = other::B(2); //~ ERROR tuple struct `B` is private
   |                    ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:92:20
   |
LL |     let c = other::C(2, 3); //~ ERROR tuple struct `C` is private
   |                    ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:95:16
   |
LL |     let other::A(()) = a; //~ ERROR tuple struct `A` is private
   |                ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:96:16
   |
LL |     let other::A(_) = a; //~ ERROR tuple struct `A` is private
   |                ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:97:22
   |
LL |     match a { other::A(()) => {} } //~ ERROR tuple struct `A` is private
   |                      ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:98:22
   |
LL |     match a { other::A(_) => {} } //~ ERROR tuple struct `A` is private
   |                      ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:100:16
   |
LL |     let other::B(_) = b; //~ ERROR tuple struct `B` is private
   |                ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:101:16
   |
LL |     let other::B(_b) = b; //~ ERROR tuple struct `B` is private
   |                ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:102:22
   |
LL |     match b { other::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                      ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:103:22
   |
LL |     match b { other::B(_b) => {} } //~ ERROR tuple struct `B` is private
   |                      ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:104:22
   |
LL |     match b { other::B(1) => {} other::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                      ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:104:40
   |
LL |     match b { other::B(1) => {} other::B(_) => {} } //~ ERROR tuple struct `B` is private
   |                                        ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:107:16
   |
LL |     let other::C(_, _) = c; //~ ERROR tuple struct `C` is private
   |                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:108:16
   |
LL |     let other::C(_a, _) = c; //~ ERROR tuple struct `C` is private
   |                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:109:16
   |
LL |     let other::C(_, _b) = c; //~ ERROR tuple struct `C` is private
   |                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:110:16
   |
LL |     let other::C(_a, _b) = c; //~ ERROR tuple struct `C` is private
   |                ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:111:22
   |
LL |     match c { other::C(_, _) => {} } //~ ERROR tuple struct `C` is private
   |                      ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:112:22
   |
LL |     match c { other::C(_a, _) => {} } //~ ERROR tuple struct `C` is private
   |                      ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:113:22
   |
LL |     match c { other::C(_, _b) => {} } //~ ERROR tuple struct `C` is private
   |                      ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:114:22
   |
LL |     match c { other::C(_a, _b) => {} } //~ ERROR tuple struct `C` is private
   |                      ^

error[E0603]: tuple struct `A` is private
  --> $DIR/privacy5.rs:122:21
   |
LL |     let a2 = other::A; //~ ERROR tuple struct `A` is private
   |                     ^

error[E0603]: tuple struct `B` is private
  --> $DIR/privacy5.rs:123:21
   |
LL |     let b2 = other::B; //~ ERROR tuple struct `B` is private
   |                     ^

error[E0603]: tuple struct `C` is private
  --> $DIR/privacy5.rs:124:21
   |
LL |     let c2 = other::C; //~ ERROR tuple struct `C` is private
   |                     ^

error: aborting due to 48 previous errors

For more information about this error, try `rustc --explain E0603`.
