error: type `m::Priv` is private
  --> $DIR/private-inferred-type-2.rs:16:5
   |
LL |     m::Pub::get_priv; //~ ERROR type `m::Priv` is private
   |     ^^^^^^^^^^^^^^^^

error: type `m::Priv` is private
  --> $DIR/private-inferred-type-2.rs:17:5
   |
LL |     m::Pub::static_method; //~ ERROR type `m::Priv` is private
   |     ^^^^^^^^^^^^^^^^^^^^^

error: type `ext::Priv` is private
  --> $DIR/private-inferred-type-2.rs:18:5
   |
LL |     ext::Pub::static_method; //~ ERROR type `ext::Priv` is private
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

