error[E0451]: field `c` of union `m::U` is private
  --> $DIR/union-field-privacy-1.rs:12:20
   |
LL |     let u = m::U { c: 0 }; //~ ERROR field `c` of union `m::U` is private
   |                    ^^^^ field `c` is private

error[E0451]: field `c` of union `m::U` is private
  --> $DIR/union-field-privacy-1.rs:16:16
   |
LL |     let m::U { c } = u; //~ ERROR field `c` of union `m::U` is private
   |                ^ field `c` is private

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0451`.
