error: unknown byte escape: f
  --> $DIR/byte-literals.rs:6:21
   |
LL | static FOO: u8 = b'/f';  //~ ERROR unknown byte escape
   |                     ^

error: unknown byte escape: f
  --> $DIR/byte-literals.rs:9:8
   |
LL |     b'/f';  //~ ERROR unknown byte escape
   |        ^

error: invalid character in numeric character escape: Z
  --> $DIR/byte-literals.rs:10:10
   |
LL |     b'/x0Z';  //~ ERROR invalid character in numeric character escape: Z
   |          ^

error: byte constant must be escaped: /t
  --> $DIR/byte-literals.rs:11:7
   |
LL |     b'    ';  //~ ERROR byte constant must be escaped
   |       ^^^^

error: byte constant must be escaped: '
  --> $DIR/byte-literals.rs:12:7
   |
LL |     b''';  //~ ERROR byte constant must be escaped
   |       ^

error: byte constant must be ASCII. Use a /xHH escape for a non-ASCII byte
  --> $DIR/byte-literals.rs:13:7
   |
LL |     b'Ã©';  //~ ERROR byte constant must be ASCII
   |       ^

error: unterminated byte constant: b'a
  --> $DIR/byte-literals.rs:14:5
   |
LL |     b'a  //~ ERROR unterminated byte constant
   |     ^^^

error: aborting due to 7 previous errors

