error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:4:5
   |
LL |     0o1.0; //~ ERROR: octal float literal is not supported
   |     ^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:6:5
   |
LL |     0o3.0f32; //~ ERROR: octal float literal is not supported
   |     ^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:7:5
   |
LL |     0o4e4; //~ ERROR: octal float literal is not supported
   |     ^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:8:5
   |
LL |     0o5.0e5; //~ ERROR: octal float literal is not supported
   |     ^^^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:9:5
   |
LL |     0o6e6f32; //~ ERROR: octal float literal is not supported
   |     ^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:10:5
   |
LL |     0o7.0e7f64; //~ ERROR: octal float literal is not supported
   |     ^^^^^^^

error: hexadecimal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:11:5
   |
LL |     0x8.0e+9; //~ ERROR: hexadecimal float literal is not supported
   |     ^^^^^^^^

error: hexadecimal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:12:5
   |
LL |     0x9.0e-9; //~ ERROR: hexadecimal float literal is not supported
   |     ^^^^^^^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:13:5
   |
LL |     0o; //~ ERROR: no valid digits
   |     ^^

error: expected at least one digit in exponent
  --> $DIR/lex-bad-numeric-literals.rs:14:8
   |
LL |     1e+; //~ ERROR: expected at least one digit in exponent
   |        ^

error: hexadecimal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:15:5
   |
LL |     0x539.0; //~ ERROR: hexadecimal float literal is not supported
   |     ^^^^^^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:18:5
   |
LL |     0x; //~ ERROR: no valid digits
   |     ^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:19:5
   |
LL |     0xu32; //~ ERROR: no valid digits
   |     ^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:20:5
   |
LL |     0ou32; //~ ERROR: no valid digits
   |     ^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:21:5
   |
LL |     0bu32; //~ ERROR: no valid digits
   |     ^^

error: no valid digits found for number
  --> $DIR/lex-bad-numeric-literals.rs:22:5
   |
LL |     0b; //~ ERROR: no valid digits
   |     ^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:24:5
   |
LL |     0o123.456; //~ ERROR: octal float literal is not supported
   |     ^^^^^^^^^

error: binary float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:26:5
   |
LL |     0b111.101; //~ ERROR: binary float literal is not supported
   |     ^^^^^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:5:5
   |
LL |     0o2f32; //~ ERROR: octal float literal is not supported
   |     ^^^^^^

error: int literal is too large
  --> $DIR/lex-bad-numeric-literals.rs:16:5
   |
LL |     9900000000000000000000000000999999999999999999999999999999; //~ ERROR: int literal is too large
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: int literal is too large
  --> $DIR/lex-bad-numeric-literals.rs:17:5
   |
LL |     9900000000000000000000000000999999999999999999999999999999; //~ ERROR: int literal is too large
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: octal float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:23:5
   |
LL |     0o123f64; //~ ERROR: octal float literal is not supported
   |     ^^^^^^^^

error: binary float literal is not supported
  --> $DIR/lex-bad-numeric-literals.rs:25:5
   |
LL |     0b101f64; //~ ERROR: binary float literal is not supported
   |     ^^^^^^^^

error: aborting due to 23 previous errors

