error: unexpected token: `...`
  --> $DIR/range_inclusive_dotdotdot.rs:8:12
   |
LL |     return ...1; //~ERROR unexpected token: `...`
   |            ^^^
help: use `..` for an exclusive range
   |
LL |     return ..1; //~ERROR unexpected token: `...`
   |            ^^
help: or `..=` for an inclusive range
   |
LL |     return ..=1; //~ERROR unexpected token: `...`
   |            ^^^

error: unexpected token: `...`
  --> $DIR/range_inclusive_dotdotdot.rs:14:13
   |
LL |     let x = ...0;    //~ERROR unexpected token: `...`
   |             ^^^
help: use `..` for an exclusive range
   |
LL |     let x = ..0;    //~ERROR unexpected token: `...`
   |             ^^
help: or `..=` for an inclusive range
   |
LL |     let x = ..=0;    //~ERROR unexpected token: `...`
   |             ^^^

error: unexpected token: `...`
  --> $DIR/range_inclusive_dotdotdot.rs:18:14
   |
LL |     let x = 5...5;   //~ERROR unexpected token: `...`
   |              ^^^
help: use `..` for an exclusive range
   |
LL |     let x = 5..5;   //~ERROR unexpected token: `...`
   |              ^^
help: or `..=` for an inclusive range
   |
LL |     let x = 5..=5;   //~ERROR unexpected token: `...`
   |              ^^^

error: unexpected token: `...`
  --> $DIR/range_inclusive_dotdotdot.rs:22:15
   |
LL |     for _ in 0...1 {} //~ERROR unexpected token: `...`
   |               ^^^
help: use `..` for an exclusive range
   |
LL |     for _ in 0..1 {} //~ERROR unexpected token: `...`
   |               ^^
help: or `..=` for an inclusive range
   |
LL |     for _ in 0..=1 {} //~ERROR unexpected token: `...`
   |               ^^^

error: aborting due to 4 previous errors

