error: expected one of `=>`, `@`, `if`, or `|`, found `:`
  --> $DIR/struct-literal-in-match-discriminant.rs:9:10
   |
LL |         x: 3    //~ ERROR expected one of `=>`, `@`, `if`, or `|`, found `:`
   |          ^ expected one of `=>`, `@`, `if`, or `|` here

error: expected one of `.`, `;`, `?`, `}`, or an operator, found `=>`
  --> $DIR/struct-literal-in-match-discriminant.rs:13:11
   |
LL |         } => {} //~ ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `=>`
   |           ^^ expected one of `.`, `;`, `?`, `}`, or an operator here

error: aborting due to 2 previous errors

