error: parenthesized lifetime bounds are not supported
  --> $DIR/trait-object-lifetime-parens.rs:3:19
   |
LL | fn f<T: Copy + ('a)>() {} //~ ERROR parenthesized lifetime bounds are not supported
   |                   ^

error: parenthesized lifetime bounds are not supported
  --> $DIR/trait-object-lifetime-parens.rs:6:26
   |
LL |     let _: Box<Copy + ('a)>; //~ ERROR parenthesized lifetime bounds are not supported
   |                          ^

error: expected type, found `'a`
  --> $DIR/trait-object-lifetime-parens.rs:7:17
   |
LL |     let _: Box<('a) + Copy>; //~ ERROR expected type, found `'a`
   |         -       ^^
   |         |
   |         while parsing the type for `_`

error: aborting due to 3 previous errors

