error: expected identifier, found keyword `as`
  --> $DIR/use-as-where-use-ends-with-mod-sep.rs:3:16
   |
LL | use std::any:: as foo; //~ ERROR expected identifier, found keyword `as`
   |                ^^ expected identifier, found keyword
help: you can escape reserved keywords to use them as identifiers
   |
LL | use std::any:: r#as foo; //~ ERROR expected identifier, found keyword `as`
   |                ^^^^

error: expected one of `::`, `;`, or `as`, found `foo`
  --> $DIR/use-as-where-use-ends-with-mod-sep.rs:3:19
   |
LL | use std::any:: as foo; //~ ERROR expected identifier, found keyword `as`
   |                   ^^^ expected one of `::`, `;`, or `as` here

error: aborting due to 2 previous errors

