error[E0600]: cannot apply unary operator `!` to type `!`
  --> $DIR/expr_unary.rs:8:16
   |
LL |     let x: ! = ! { return; }; //~ ERROR unreachable
   |                ^^^^^^^^^^^^^ cannot apply unary operator `!`

error: unreachable expression
  --> $DIR/expr_unary.rs:8:16
   |
LL |     let x: ! = ! { return; }; //~ ERROR unreachable
   |                ^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/expr_unary.rs:5:9
   |
LL | #![deny(unreachable_code)]
   |         ^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0600`.
