error: unreachable expression
  --> $DIR/expr_call.rs:13:17
   |
LL |     foo(return, 22); //~ ERROR unreachable
   |                 ^^
   |
note: lint level defined here
  --> $DIR/expr_call.rs:5:9
   |
LL | #![deny(unreachable_code)]
   |         ^^^^^^^^^^^^^^^^

error: unreachable expression
  --> $DIR/expr_call.rs:18:5
   |
LL |     bar(return); //~ ERROR unreachable
   |     ^^^^^^^^^^^

error: aborting due to 2 previous errors

