error[E0308]: mismatched types
  --> $DIR/associated-type-projection-from-supertrait.rs:33:23
   |
LL | fn b() { dent(ModelT, Blue); } //~ ERROR mismatched types
   |                       ^^^^ expected struct `Black`, found struct `Blue`
   |
   = note: expected type `Black`
              found type `Blue`

error[E0308]: mismatched types
  --> $DIR/associated-type-projection-from-supertrait.rs:34:23
   |
LL | fn c() { dent(ModelU, Black); } //~ ERROR mismatched types
   |                       ^^^^^ expected struct `Blue`, found struct `Black`
   |
   = note: expected type `Blue`
              found type `Black`

error[E0308]: mismatched types
  --> $DIR/associated-type-projection-from-supertrait.rs:40:28
   |
LL | fn f() { ModelT.chip_paint(Blue); } //~ ERROR mismatched types
   |                            ^^^^ expected struct `Black`, found struct `Blue`
   |
   = note: expected type `Black`
              found type `Blue`

error[E0308]: mismatched types
  --> $DIR/associated-type-projection-from-supertrait.rs:41:28
   |
LL | fn g() { ModelU.chip_paint(Black); } //~ ERROR mismatched types
   |                            ^^^^^ expected struct `Blue`, found struct `Black`
   |
   = note: expected type `Blue`
              found type `Black`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
