// ts-simple-ast/dist/compiler/ast/base/ReadonlyableNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var errors = require("../../../errors");
var typescript_1 = require("../../../typescript");
var callBaseSet_1 = require("../callBaseSet");
var callBaseGetStructure_1 = require("../callBaseGetStructure");
function ReadonlyableNode(Base) {
    return /** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        class_1.prototype.isReadonly = function () {
            return this.getReadonlyKeyword() != null;
        };
        class_1.prototype.getReadonlyKeyword = function () {
            return this.getFirstModifierByKind(typescript_1.SyntaxKind.ReadonlyKeyword);
        };
        class_1.prototype.getReadonlyKeywordOrThrow = function () {
            return errors.throwIfNullOrUndefined(this.getReadonlyKeyword(), "Expected to find a readonly keyword.");
        };
        class_1.prototype.setIsReadonly = function (value) {
            this.toggleModifier("readonly", value);
            return this;
        };
        class_1.prototype.set = function (structure) {
            callBaseSet_1.callBaseSet(Base.prototype, this, structure);
            if (structure.isReadonly != null)
                this.setIsReadonly(structure.isReadonly);
            return this;
        };
        class_1.prototype.getStructure = function () {
            return callBaseGetStructure_1.callBaseGetStructure(Base.prototype, this, {
                isReadonly: this.isReadonly()
            });
        };
        return class_1;
    }(Base));
}
exports.ReadonlyableNode = ReadonlyableNode;
