error[E0252]: the name `foo` is defined multiple times
  --> $DIR/duplicate.rs:25:9
   |
LL |     use a::foo;
   |         ------ previous import of the value `foo` here
LL |     use a::foo; //~ ERROR the name `foo` is defined multiple times
   |         ^^^^^^ `foo` reimported here
   |
   = note: `foo` must be defined only once in the value namespace of this module
help: You can use `as` to change the binding name of the import
   |
LL |     use a::foo as other_foo; //~ ERROR the name `foo` is defined multiple times
   |         ^^^^^^^^^^^^^^^^^^^

error[E0659]: `foo` is ambiguous
  --> $DIR/duplicate.rs:56:9
   |
LL |     use self::foo::bar; //~ ERROR `foo` is ambiguous
   |         ^^^^^^^^^^^^^^
   |
note: `foo` could refer to the name imported here
  --> $DIR/duplicate.rs:53:9
   |
LL |     use self::m1::*;
   |         ^^^^^^^^^^^
note: `foo` could also refer to the name imported here
  --> $DIR/duplicate.rs:54:9
   |
LL |     use self::m2::*;
   |         ^^^^^^^^^^^
   = note: consider adding an explicit import of `foo` to disambiguate

error[E0659]: `foo` is ambiguous
  --> $DIR/duplicate.rs:45:5
   |
LL |     f::foo(); //~ ERROR `foo` is ambiguous
   |     ^^^^^^
   |
note: `foo` could refer to the name imported here
  --> $DIR/duplicate.rs:34:13
   |
LL |     pub use a::*;
   |             ^^^^
note: `foo` could also refer to the name imported here
  --> $DIR/duplicate.rs:35:13
   |
LL |     pub use b::*;
   |             ^^^^
   = note: consider adding an explicit import of `foo` to disambiguate

error[E0659]: `foo` is ambiguous
  --> $DIR/duplicate.rs:46:5
   |
LL |     g::foo(); //~ ERROR `foo` is ambiguous
   |     ^^^^^^
   |
note: `foo` could refer to the name imported here
  --> $DIR/duplicate.rs:39:13
   |
LL |     pub use a::*;
   |             ^^^^
note: `foo` could also refer to the name imported here
  --> $DIR/duplicate.rs:40:13
   |
LL |     pub use f::*;
   |             ^^^^
   = note: consider adding an explicit import of `foo` to disambiguate

error[E0659]: `foo` is ambiguous
  --> $DIR/duplicate.rs:59:9
   |
LL |         foo::bar(); //~ ERROR `foo` is ambiguous
   |         ^^^^^^^^
   |
note: `foo` could refer to the name imported here
  --> $DIR/duplicate.rs:53:9
   |
LL |     use self::m1::*;
   |         ^^^^^^^^^^^
note: `foo` could also refer to the name imported here
  --> $DIR/duplicate.rs:54:9
   |
LL |     use self::m2::*;
   |         ^^^^^^^^^^^
   = note: consider adding an explicit import of `foo` to disambiguate

error: aborting due to 5 previous errors

Some errors occurred: E0252, E0659.
For more information about an error, try `rustc --explain E0252`.
