error[E0425]: cannot find function `f` in module `glob_conflict`
  --> $DIR/glob-conflict-cross-crate.rs:6:20
   |
LL |     glob_conflict::f(); //~ ERROR cannot find function `f` in module `glob_conflict`
   |                    ^ not found in `glob_conflict`

error[E0425]: cannot find function `f` in module `glob_conflict::glob`
  --> $DIR/glob-conflict-cross-crate.rs:7:26
   |
LL |     glob_conflict::glob::f(); //~ ERROR cannot find function `f` in module `glob_conflict::glob`
   |                          ^ not found in `glob_conflict::glob`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0425`.
