error[E0432]: unresolved import `nonexistent_module`
  --> $DIR/issue-53269.rs:6:9
   |
LL |     use nonexistent_module::mac; //~ ERROR unresolved import `nonexistent_module`
   |         ^^^^^^^^^^^^^^^^^^ maybe a missing `extern crate nonexistent_module;`?

error[E0659]: `mac` is ambiguous (`macro_rules` vs non-`macro_rules` from other module)
  --> $DIR/issue-53269.rs:8:5
   |
LL |     mac!(); //~ ERROR `mac` is ambiguous
   |     ^^^ ambiguous name
   |
note: `mac` could refer to the macro defined here
  --> $DIR/issue-53269.rs:3:1
   |
LL | macro_rules! mac { () => () }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: `mac` could also refer to the unresolved item imported here
  --> $DIR/issue-53269.rs:6:9
   |
LL |     use nonexistent_module::mac; //~ ERROR unresolved import `nonexistent_module`
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   = help: use `self::mac` to refer to this unresolved item unambiguously

error: aborting due to 2 previous errors

Some errors occurred: E0432, E0659.
For more information about an error, try `rustc --explain E0432`.
