error: `panic` is already in scope
  --> $DIR/shadow_builtin_macros.rs:42:9
   |
LL |         macro_rules! panic { () => {} } //~ ERROR `panic` is already in scope
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
LL |     } }
LL |     m!();
   |     ----- in this macro invocation
   |
   = note: macro-expanded `macro_rules!`s may not shadow existing macros (see RFC 1560)

error[E0659]: `panic` is ambiguous
  --> $DIR/shadow_builtin_macros.rs:27:14
   |
LL |     fn f() { panic!(); } //~ ERROR ambiguous
   |              ^^^^^
   |
note: `panic` could refer to the name imported here
  --> $DIR/shadow_builtin_macros.rs:26:9
   |
LL |     use foo::*;
   |         ^^^^^^
   = note: `panic` is also a builtin macro
   = note: consider adding an explicit import of `panic` to disambiguate

error[E0659]: `panic` is ambiguous
  --> $DIR/shadow_builtin_macros.rs:32:14
   |
LL |     fn f() { panic!(); } //~ ERROR ambiguous
   |              ^^^^^
   |
note: `panic` could refer to the name imported here
  --> $DIR/shadow_builtin_macros.rs:31:26
   |
LL |     ::two_macros::m!(use foo::panic;);
   |                          ^^^^^^^^^^
   = note: `panic` is also a builtin macro
   = note: macro-expanded macro imports do not shadow

error[E0659]: `n` is ambiguous
  --> $DIR/shadow_builtin_macros.rs:61:5
   |
LL |     n!(); //~ ERROR ambiguous
   |     ^
   |
note: `n` could refer to the name imported here
  --> $DIR/shadow_builtin_macros.rs:60:9
   |
LL |     use bar::*;
   |         ^^^^^^
note: `n` could also refer to the name imported here
  --> $DIR/shadow_builtin_macros.rs:48:13
   |
LL | #[macro_use(n)]
   |             ^
   = note: consider adding an explicit import of `n` to disambiguate

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0659`.
