error[E0425]: cannot find function `f` in this scope
  --> $DIR/globs.rs:32:9
   |
LL |         f(); //~ ERROR cannot find function `f` in this scope
   |         ^ not found in this scope
help: possible candidates are found in other modules, you can import them into scope
   |
LL | use foo::f;
   |
LL | use foo::f;
   |
LL | use foo::f;
   |

error[E0425]: cannot find function `g` in this scope
  --> $DIR/globs.rs:25:5
   |
LL |       g(); //~ ERROR cannot find function `g` in this scope
   |       ^ not found in this scope
...
LL | /     m! {
LL | |         use bar::*;
LL | |         g();
LL | |         f(); //~ ERROR cannot find function `f` in this scope
LL | |     }
   | |_____- in this macro invocation
help: possible candidates are found in other modules, you can import them into scope
   |
LL | use bar::g;
   |
LL | use foo::test2::test::g;
   |
LL | use foo::test::g;
   |
LL | use foo::test::g;
   |

error[E0425]: cannot find function `f` in this scope
  --> $DIR/globs.rs:64:17
   |
LL | n!(f);
   | ------ in this macro invocation
...
LL |                 f //~ ERROR cannot find function `f` in this scope
   |                 ^ not found in this scope

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0425`.
