error[E0433]: failed to resolve: use of undeclared type or module `Vec`
  --> $DIR/no_implicit_prelude.rs:11:9
   |
LL |     fn f() { ::bar::m!(); }
   |              ------------ in this macro invocation
...
LL |         Vec::new(); //~ ERROR failed to resolve
   |         ^^^ use of undeclared type or module `Vec`

error[E0599]: no method named `clone` found for type `()` in the current scope
  --> $DIR/no_implicit_prelude.rs:12:12
   |
LL |     fn f() { ::bar::m!(); }
   |              ------------ in this macro invocation
...
LL |         ().clone() //~ ERROR no method named `clone` found
   |            ^^^^^
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           `use std::clone::Clone;`

error: aborting due to 2 previous errors

Some errors occurred: E0433, E0599.
For more information about an error, try `rustc --explain E0433`.
