error[E0432]: unresolved import `ImportError`
  --> $DIR/issue-5035.rs:15:5
   |
LL | use ImportError; //~ ERROR unresolved import `ImportError` [E0432]
   |     ^^^^^^^^^^^ no `ImportError` in the root

error[E0404]: expected trait, found type alias `K`
  --> $DIR/issue-5035.rs:13:6
   |
LL | impl K for isize {} //~ ERROR expected trait, found type alias `K`
   |      ^
   |      |
   |      did you mean `I`?
   |      type aliases cannot be used for traits

error: aborting due to 2 previous errors

Some errors occurred: E0404, E0432.
For more information about an error, try `rustc --explain E0404`.
