error: expected `}`, found `,`
  --> $DIR/issue-54379.rs:9:22
   |
LL |         MyStruct { .., Some(_) } => {}, //~ ERROR pattern does not mention field `s1`
   |                    --^
   |                    | |
   |                    | expected `}`
   |                    `..` must be at the end and cannot have a trailing comma

error: expected `,`
  --> $DIR/issue-54379.rs:9:24
   |
LL |         MyStruct { .., Some(_) } => {}, //~ ERROR pattern does not mention field `s1`
   |                        ^^^^

error[E0027]: pattern does not mention field `s1`
  --> $DIR/issue-54379.rs:9:9
   |
LL |         MyStruct { .., Some(_) } => {}, //~ ERROR pattern does not mention field `s1`
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ missing field `s1`

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0027`.
