error[E0391]: cycle detected when processing `<impl at $DIR/resolve-self-in-impl.rs:14:1: 14:20>`
  --> $DIR/resolve-self-in-impl.rs:14:13
   |
LL | impl Tr for Self {} //~ ERROR cycle detected
   |             ^^^^
   |
   = note: ...which again requires processing `<impl at $DIR/resolve-self-in-impl.rs:14:1: 14:20>`, completing the cycle

error[E0391]: cycle detected when processing `<impl at $DIR/resolve-self-in-impl.rs:15:1: 15:23>`
  --> $DIR/resolve-self-in-impl.rs:15:15
   |
LL | impl Tr for S<Self> {} //~ ERROR cycle detected
   |               ^^^^
   |
   = note: ...which again requires processing `<impl at $DIR/resolve-self-in-impl.rs:15:1: 15:23>`, completing the cycle

error[E0391]: cycle detected when processing `<impl at $DIR/resolve-self-in-impl.rs:16:1: 16:13>`
  --> $DIR/resolve-self-in-impl.rs:16:6
   |
LL | impl Self {} //~ ERROR cycle detected
   |      ^^^^
   |
   = note: ...which again requires processing `<impl at $DIR/resolve-self-in-impl.rs:16:1: 16:13>`, completing the cycle

error[E0391]: cycle detected when processing `<impl at $DIR/resolve-self-in-impl.rs:17:1: 17:16>`
  --> $DIR/resolve-self-in-impl.rs:17:8
   |
LL | impl S<Self> {} //~ ERROR cycle detected
   |        ^^^^
   |
   = note: ...which again requires processing `<impl at $DIR/resolve-self-in-impl.rs:17:1: 17:16>`, completing the cycle

error[E0391]: cycle detected when processing `<impl at $DIR/resolve-self-in-impl.rs:18:1: 18:26>`
  --> $DIR/resolve-self-in-impl.rs:18:1
   |
LL | impl Tr<Self::A> for S {} //~ ERROR cycle detected
   | ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: ...which again requires processing `<impl at $DIR/resolve-self-in-impl.rs:18:1: 18:26>`, completing the cycle

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0391`.
