error[E0423]: expected function, found type alias `A`
  --> $DIR/tuple-struct-alias.rs:5:13
   |
LL |     let s = A(0, 1); //~ ERROR expected function
   |             ^ did you mean `S`?
   |
   = note: can't use a type alias as a constructor

error[E0532]: expected tuple struct/variant, found type alias `A`
  --> $DIR/tuple-struct-alias.rs:7:9
   |
LL |         A(..) => {} //~ ERROR expected tuple struct/variant
   |         ^ did you mean `S`?
   |
   = note: can't use a type alias as a constructor

error: aborting due to 2 previous errors

Some errors occurred: E0423, E0532.
For more information about an error, try `rustc --explain E0423`.
