error[E0426]: use of undeclared label `'fo`
  --> $DIR/suggest-labels.rs:4:15
   |
LL |         break 'fo; //~ ERROR use of undeclared label
   |               ^^^ did you mean `'foo`?

error[E0426]: use of undeclared label `'bor`
  --> $DIR/suggest-labels.rs:8:18
   |
LL |         continue 'bor; //~ ERROR use of undeclared label
   |                  ^^^^ did you mean `'bar`?

error[E0426]: use of undeclared label `'longlable`
  --> $DIR/suggest-labels.rs:13:19
   |
LL |             break 'longlable; //~ ERROR use of undeclared label
   |                   ^^^^^^^^^^ did you mean `'longlabel1`?

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0426`.
