error[E0512]: transmute called with types of different sizes
  --> $DIR/main.rs:26:5
   |
LL |     transmute(x) //~ ERROR transmute called with types of different sizes
   |     ^^^^^^^^^
   |
   = note: source type: <C as TypeConstructor<'a>>::T (size can vary because of <C as TypeConstructor>::T)
   = note: target type: <C as TypeConstructor<'b>>::T (size can vary because of <C as TypeConstructor>::T)

error[E0512]: transmute called with types of different sizes
  --> $DIR/main.rs:30:17
   |
LL |     let x: u8 = transmute(10u16); //~ ERROR transmute called with types of different sizes
   |                 ^^^^^^^^^
   |
   = note: source type: u16 (16 bits)
   = note: target type: u8 (8 bits)

error[E0512]: transmute called with types of different sizes
  --> $DIR/main.rs:34:17
   |
LL |     let x: u8 = transmute("test"); //~ ERROR transmute called with types of different sizes
   |                 ^^^^^^^^^
   |
   = note: source type: &str ($STR bits)
   = note: target type: u8 (8 bits)

error[E0512]: transmute called with types of different sizes
  --> $DIR/main.rs:39:18
   |
LL |     let x: Foo = transmute(10); //~ ERROR transmute called with types of different sizes
   |                  ^^^^^^^^^
   |
   = note: source type: i32 (32 bits)
   = note: target type: Foo (0 bits)

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0512`.
