error: lifetime parameter `'a` only used once
  --> $DIR/one-use-in-fn-argument.rs:8:6
   |
LL | fn a<'a>(x: &'a u32) { //~ ERROR `'a` only used once
   |      ^^      -- ...is used only here
   |      |
   |      this lifetime...
   |
note: lint level defined here
  --> $DIR/one-use-in-fn-argument.rs:1:9
   |
LL | #![deny(single_use_lifetimes)]
   |         ^^^^^^^^^^^^^^^^^^^^
help: elide the single-use lifetime
   |
LL | fn a(x: &u32) { //~ ERROR `'a` only used once
   |    --   --

error: aborting due to previous error

