error[E0603]: extern crate `core` is private
  --> $DIR/extern-crate-visibility.rs:6:10
   |
LL | use foo::core::cell; //~ ERROR extern crate `core` is private
   |          ^^^^

error[E0603]: extern crate `core` is private
  --> $DIR/extern-crate-visibility.rs:9:10
   |
LL |     foo::core::cell::Cell::new(0); //~ ERROR extern crate `core` is private
   |          ^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0603`.
