warning: not reporting region error due to nll
  --> $DIR/ex1-return-one-existing-name-self-is-anon.rs:18:30
   |
LL |         if true { x } else { self } //~ ERROR lifetime mismatch
   |                              ^^^^

error[E0623]: lifetime mismatch
  --> $DIR/ex1-return-one-existing-name-self-is-anon.rs:18:9
   |
LL |     fn foo<'a>(&self, x: &'a Foo) -> &'a Foo {
   |                -----                 -------
   |                |
   |                this parameter and the return type are declared with different lifetimes...
LL | 
LL |         if true { x } else { self } //~ ERROR lifetime mismatch
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...but data from `self` is returned here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0623`.
