warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-3.rs:12:13
   |
LL |     z.push((x,y)); //~ ERROR lifetime mismatch
   |             ^

warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-3.rs:12:15
   |
LL |     z.push((x,y)); //~ ERROR lifetime mismatch
   |               ^

error[E0623]: lifetime mismatch
  --> $DIR/ex3-both-anon-regions-3.rs:11:33
   |
LL | fn foo(z: &mut Vec<(&u8,&u8)>, (x, y): (&u8, &u8)) {
   |                     ---         ^       --- these two types are declared with different lifetimes...
   |                                 |
   |                                 ...but data flows into `z` here

error[E0623]: lifetime mismatch
  --> $DIR/ex3-both-anon-regions-3.rs:11:33
   |
LL | fn foo(z: &mut Vec<(&u8,&u8)>, (x, y): (&u8, &u8)) {
   |                         ---     ^            --- these two types are declared with different lifetimes...
   |                                 |
   |                                 ...but data flows into `z` here

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0623`.
