warning: not reporting region error due to nll
  --> $DIR/ex3-both-anon-regions-both-are-structs-3.rs:16:11
   |
LL |     x.a = x.b; //~ ERROR lifetime mismatch
   |           ^^^

error[E0623]: lifetime mismatch
  --> $DIR/ex3-both-anon-regions-both-are-structs-3.rs:16:5
   |
LL | fn foo(mut x: Ref) {
   |               ---
   |               |
   |               this type is declared with multiple lifetimes...
LL |     x.a = x.b; //~ ERROR lifetime mismatch
   |     ^^^^^^^^^ ...but data with one lifetime flows into the other here

error: aborting due to previous error

For more information about this error, try `rustc --explain E0623`.
