error[E0106]: missing lifetime specifier
  --> $DIR/dyn-trait-underscore-in-struct.rs:9:24
   |
LL |     x: Box<dyn Debug + '_>, //~ ERROR missing lifetime specifier
   |                        ^^ expected lifetime parameter

error[E0228]: the lifetime bound for this object type cannot be deduced from context; please supply an explicit bound
  --> $DIR/dyn-trait-underscore-in-struct.rs:9:12
   |
LL |     x: Box<dyn Debug + '_>, //~ ERROR missing lifetime specifier
   |            ^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

Some errors occurred: E0106, E0228.
For more information about an error, try `rustc --explain E0106`.
