error[E0637]: `'_` cannot be used here
  --> $DIR/underscore-lifetime-binders.rs:4:8
   |
LL | fn foo<'_> //~ ERROR cannot be used here
   |        ^^ `'_` is a reserved lifetime name

error[E0637]: `'_` cannot be used here
  --> $DIR/underscore-lifetime-binders.rs:10:21
   |
LL | fn meh() -> Box<for<'_> Meh<'_>> //~ ERROR cannot be used here
   |                     ^^ `'_` is a reserved lifetime name

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:2:17
   |
LL | struct Baz<'a>(&'_ &'a u8); //~ ERROR missing lifetime specifier
   |                 ^^ expected lifetime parameter

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:10:29
   |
LL | fn meh() -> Box<for<'_> Meh<'_>> //~ ERROR cannot be used here
   |                             ^^ help: consider giving it a 'static lifetime: `'static`
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from

error[E0106]: missing lifetime specifier
  --> $DIR/underscore-lifetime-binders.rs:16:35
   |
LL | fn foo2(_: &'_ u8, y: &'_ u8) -> &'_ u8 { y } //~ ERROR missing lifetime specifier
   |                                   ^^ expected lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but the signature does not say whether it is borrowed from `_` or `y`

error: aborting due to 5 previous errors

Some errors occurred: E0106, E0637.
For more information about an error, try `rustc --explain E0106`.
