error[E0369]: binary operation `==` cannot be applied to type `Error`
  --> $DIR/derives-span-PartialEq-enum.rs:9:6
   |
LL |      Error //~ ERROR
   |      ^^^^^
   |
   = note: an implementation of `std::cmp::PartialEq` might be missing for `Error`

error[E0369]: binary operation `!=` cannot be applied to type `Error`
  --> $DIR/derives-span-PartialEq-enum.rs:9:6
   |
LL |      Error //~ ERROR
   |      ^^^^^
   |
   = note: an implementation of `std::cmp::PartialEq` might be missing for `Error`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0369`.
