error[E0369]: binary operation `==` cannot be applied to type `Error`
  --> $DIR/derives-span-PartialEq-struct.rs:8:5
   |
LL |     x: Error //~ ERROR
   |     ^^^^^^^^
   |
   = note: an implementation of `std::cmp::PartialEq` might be missing for `Error`

error[E0369]: binary operation `!=` cannot be applied to type `Error`
  --> $DIR/derives-span-PartialEq-struct.rs:8:5
   |
LL |     x: Error //~ ERROR
   |     ^^^^^^^^
   |
   = note: an implementation of `std::cmp::PartialEq` might be missing for `Error`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0369`.
