warning: empty trait list in `derive`
  --> $DIR/deriving-meta-empty-trait-list.rs:5:1
   |
LL | #[derive]   //~ WARNING empty trait list in `derive`
   | ^^^^^^^^^

warning: empty trait list in `derive`
  --> $DIR/deriving-meta-empty-trait-list.rs:8:1
   |
LL | #[derive()] //~ WARNING empty trait list in `derive`
   | ^^^^^^^^^^^

