error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2015-ques-rep.rs:6:10
   |
LL |     ($(a)?) => {} //~ERROR expected `*` or `+`
   |          ^
   |
   = note: `?` is not a macro repetition operator

error: expected `*` or `+`
  --> $DIR/macro-at-most-once-rep-2015-ques-rep.rs:10:11
   |
LL |     ($(a),?) => {} //~ERROR expected `*` or `+`
   |           ^
   |
   = note: `?` is not a macro repetition operator

error: aborting due to 2 previous errors

