error[E0665]: `Default` cannot be derived for enums, only structs
  --> $DIR/macros-nonfatal-errors.rs:9:10
   |
LL | #[derive(Default)] //~ ERROR
   |          ^^^^^^^

error: inline assembly must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:13:10
   |
LL |     asm!(invalid); //~ ERROR
   |          ^^^^^^^

error: concat_idents! requires ident args.
  --> $DIR/macros-nonfatal-errors.rs:15:5
   |
LL |     concat_idents!("not", "idents"); //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: argument must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:17:17
   |
LL |     option_env!(invalid); //~ ERROR
   |                 ^^^^^^^

error: expected string literal
  --> $DIR/macros-nonfatal-errors.rs:18:10
   |
LL |     env!(invalid); //~ ERROR
   |          ^^^^^^^

error: expected string literal
  --> $DIR/macros-nonfatal-errors.rs:19:10
   |
LL |     env!(foo, abr, baz); //~ ERROR
   |          ^^^

error: environment variable `RUST_HOPEFULLY_THIS_DOESNT_EXIST` not defined
  --> $DIR/macros-nonfatal-errors.rs:20:5
   |
LL |     env!("RUST_HOPEFULLY_THIS_DOESNT_EXIST"); //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: format argument must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:22:13
   |
LL |     format!(invalid); //~ ERROR
   |             ^^^^^^^
help: you might be missing a string literal to format with
   |
LL |     format!("{}", invalid); //~ ERROR
   |             ^^^^^

error: argument must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:24:14
   |
LL |     include!(invalid); //~ ERROR
   |              ^^^^^^^

error: argument must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:26:18
   |
LL |     include_str!(invalid); //~ ERROR
   |                  ^^^^^^^

error: couldn't read $DIR/i'd be quite surprised if a file with this name existed: $FILE_NOT_FOUND_MSG (os error 2)
  --> $DIR/macros-nonfatal-errors.rs:27:5
   |
LL |     include_str!("i'd be quite surprised if a file with this name existed"); //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: argument must be a string literal
  --> $DIR/macros-nonfatal-errors.rs:28:20
   |
LL |     include_bytes!(invalid); //~ ERROR
   |                    ^^^^^^^

error: couldn't read $DIR/i'd be quite surprised if a file with this name existed: $FILE_NOT_FOUND_MSG (os error 2)
  --> $DIR/macros-nonfatal-errors.rs:29:5
   |
LL |     include_bytes!("i'd be quite surprised if a file with this name existed"); //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/macros-nonfatal-errors.rs:31:5
   |
LL |     trace_macros!(invalid); //~ ERROR
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 14 previous errors

For more information about this error, try `rustc --explain E0665`.
