error: the `?` macro repetition operator does not take a separator
  --> $DIR/macro-at-most-once-rep-2018.rs:10:10
   |
LL |     ($(a),?) => {}; //~ERROR the `?` macro repetition operator
   |          ^

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:24:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:25:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?a); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:26:11
   |
LL | macro_rules! foo {
   | ---------------- when calling this macro
...
LL |     foo!(a?a?a); //~ ERROR no rules expected the token `?`
   |           ^ no rules expected this token in macro call

error: unexpected end of macro invocation
  --> $DIR/macro-at-most-once-rep-2018.rs:28:5
   |
LL | macro_rules! barplus {
   | -------------------- when calling this macro
...
LL |     barplus!(); //~ERROR unexpected end of macro invocation
   |     ^^^^^^^^^^^ missing tokens in macro arguments

error: unexpected end of macro invocation
  --> $DIR/macro-at-most-once-rep-2018.rs:29:15
   |
LL | macro_rules! barplus {
   | -------------------- when calling this macro
...
LL |     barplus!(a); //~ERROR unexpected end of macro invocation
   |               ^ missing tokens in macro arguments

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:30:15
   |
LL | macro_rules! barplus {
   | -------------------- when calling this macro
...
LL |     barplus!(a?); //~ ERROR no rules expected the token `?`
   |               ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:31:15
   |
LL | macro_rules! barplus {
   | -------------------- when calling this macro
...
LL |     barplus!(a?a); //~ ERROR no rules expected the token `?`
   |               ^ no rules expected this token in macro call

error: unexpected end of macro invocation
  --> $DIR/macro-at-most-once-rep-2018.rs:35:5
   |
LL | macro_rules! barstar {
   | -------------------- when calling this macro
...
LL |     barstar!(); //~ERROR unexpected end of macro invocation
   |     ^^^^^^^^^^^ missing tokens in macro arguments

error: unexpected end of macro invocation
  --> $DIR/macro-at-most-once-rep-2018.rs:36:15
   |
LL | macro_rules! barstar {
   | -------------------- when calling this macro
...
LL |     barstar!(a); //~ERROR unexpected end of macro invocation
   |               ^ missing tokens in macro arguments

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:37:15
   |
LL | macro_rules! barstar {
   | -------------------- when calling this macro
...
LL |     barstar!(a?); //~ ERROR no rules expected the token `?`
   |               ^ no rules expected this token in macro call

error: no rules expected the token `?`
  --> $DIR/macro-at-most-once-rep-2018.rs:38:15
   |
LL | macro_rules! barstar {
   | -------------------- when calling this macro
...
LL |     barstar!(a?a); //~ ERROR no rules expected the token `?`
   |               ^ no rules expected this token in macro call

error: aborting due to 12 previous errors

