warning: using `?` as a separator is deprecated and will be a hard error in an upcoming edition
  --> $DIR/macro-at-most-once-rep-2015-ques-sep.rs:10:10
   |
LL |     ($(a)?*) => {} //~WARN using `?` as a separator
   |          ^
   |
note: lint level defined here
  --> $DIR/macro-at-most-once-rep-2015-ques-sep.rs:7:9
   |
LL | #![warn(rust_2018_compatibility)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^
   = note: #[warn(question_mark_macro_sep)] implied by #[warn(rust_2018_compatibility)]
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in the 2018 edition!
   = note: for more information, see issue #48075 <https://github.com/rust-lang/rust/issues/48075>

warning: using `?` as a separator is deprecated and will be a hard error in an upcoming edition
  --> $DIR/macro-at-most-once-rep-2015-ques-sep.rs:15:10
   |
LL |     ($(a)?+) => {} //~WARN using `?` as a separator
   |          ^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in the 2018 edition!
   = note: for more information, see issue #48075 <https://github.com/rust-lang/rust/issues/48075>

