error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:1
   |
LL | auto trait Magic: Copy {} //~ ERROR E0568
   | ^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `NoClone: std::marker::Copy` is not satisfied
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:15:18
   |
LL |     let (a, b) = copy(NoClone); //~ ERROR
   |                  ^^^^ the trait `std::marker::Copy` is not implemented for `NoClone`
   |
   = note: required because of the requirements on the impl of `Magic` for `NoClone`
note: required by `copy`
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:9:1
   |
LL | fn copy<T: Magic>(x: T) -> (T, T) { (x, x) }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

Some errors occurred: E0277, E0568.
For more information about an error, try `rustc --explain E0277`.
