error[E0404]: expected trait, found struct `Foo`
  --> $DIR/trait-bounds-not-on-struct.rs:5:16
   |
LL | fn foo(_x: Box<Foo + Send>) { } //~ ERROR expected trait, found struct `Foo`
   |                ^^^ not a trait

error[E0404]: expected trait, found struct `Vec`
  --> $DIR/trait-bounds-not-on-struct.rs:7:21
   |
LL | type A<T> = Box<dyn Vec<T>>; //~ ERROR expected trait, found struct `Vec`
   |                     ^^^^^^ not a trait

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0404`.
