error[E0038]: the trait `Tr` cannot be made into an object
  --> $DIR/trait-object-safety.rs:15:18
   |
LL |     let _: &Tr = &St; //~ ERROR E0038
   |                  ^^^ the trait `Tr` cannot be made into an object
   |
   = note: method `foo` has no receiver
   = note: required because of the requirements on the impl of `std::ops::CoerceUnsized<&dyn Tr>` for `&St`

error[E0038]: the trait `Tr` cannot be made into an object
  --> $DIR/trait-object-safety.rs:15:12
   |
LL |     let _: &Tr = &St; //~ ERROR E0038
   |            ^^^ the trait `Tr` cannot be made into an object
   |
   = note: method `foo` has no receiver

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0038`.
