error[E0200]: the trait `UnsafeTrait` requires an `unsafe impl` declaration
  --> $DIR/trait-safety-trait-impl.rs:14:1
   |
LL | impl UnsafeTrait for u16 { } //~ ERROR requires an `unsafe impl` declaration
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0199]: implementing the trait `SafeTrait` is not unsafe
  --> $DIR/trait-safety-trait-impl.rs:16:1
   |
LL | unsafe impl SafeTrait for u32 { } //~ ERROR the trait `SafeTrait` is not unsafe
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

Some errors occurred: E0199, E0200.
For more information about an error, try `rustc --explain E0199`.
