error[E0507]: cannot move out of borrowed content
  --> $DIR/issue-40402-2.rs:5:18
   |
LL |     let (a, b) = x[0]; //~ ERROR cannot move out of indexed content
   |          -  -    ^^^^
   |          |  |    |
   |          |  |    cannot move out of borrowed content
   |          |  |    help: consider borrowing here: `&x[0]`
   |          |  ...and here
   |          data moved here
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/issue-40402-2.rs:5:10
   |
LL |     let (a, b) = x[0]; //~ ERROR cannot move out of indexed content
   |          ^  ^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0507`.
