error: expected `{`, found `=>`
  --> $DIR/missing-block-hint.rs:3:18
   |
LL |         if (foo) => {} //~ ERROR expected `{`, found `=>`
   |         --       ^^ expected `{`
   |         |
   |         this `if` statement has a condition, but no block

error: expected `{`, found `bar`
  --> $DIR/missing-block-hint.rs:7:13
   |
LL |         if (foo)
   |         -- this `if` statement has a condition, but no block
LL |             bar; //~ ERROR expected `{`, found `bar`
   |             ^^^-
   |             |
   |             expected `{`
   |             help: try placing this code inside a block: `{ bar; }`

error: aborting due to 2 previous errors

