error: unused attribute
  --> $DIR/unused-attr.rs:7:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^
   |
note: lint level defined here
  --> $DIR/unused-attr.rs:1:9
   |
LL | #![deny(unused_attributes)]
   |         ^^^^^^^^^^^^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:10:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:15:5
   |
LL |     #[foo] //~ ERROR unused attribute
   |     ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:13:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:23:9
   |
LL |         #[foo] //~ ERROR unused attribute
   |         ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:21:5
   |
LL |     #[foo] //~ ERROR unused attribute
   |     ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:19:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:31:9
   |
LL |         #[foo] //~ ERROR unused attribute
   |         ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:28:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:38:5
   |
LL |     #[foo] //~ ERROR unused attribute
   |     ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:36:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:44:5
   |
LL |     #[foo] //~ ERROR unused attribute
   |     ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:46:5
   |
LL |     #[foo] //~ ERROR unused attribute
   |     ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:42:1
   |
LL | #[foo] //~ ERROR unused attribute
   | ^^^^^^

error: unused attribute
  --> $DIR/unused-attr.rs:5:1
   |
LL | #![foo] //~ ERROR unused attribute
   | ^^^^^^^

error: aborting due to 15 previous errors

