error: unused `MustUse` that must be used
  --> $DIR/unused-result.rs:21:5
   |
LL |     foo::<MustUse>(); //~ ERROR: unused `MustUse` that must be used
   |     ^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/unused-result.rs:2:25
   |
LL | #![deny(unused_results, unused_must_use)]
   |                         ^^^^^^^^^^^^^^^

error: unused `MustUseMsg` that must be used
  --> $DIR/unused-result.rs:22:5
   |
LL |     foo::<MustUseMsg>(); //~ ERROR: unused `MustUseMsg` that must be used
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = note: some message

error: unused result
  --> $DIR/unused-result.rs:34:5
   |
LL |     foo::<isize>(); //~ ERROR: unused result
   |     ^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/unused-result.rs:2:9
   |
LL | #![deny(unused_results, unused_must_use)]
   |         ^^^^^^^^^^^^^^

error: unused `MustUse` that must be used
  --> $DIR/unused-result.rs:35:5
   |
LL |     foo::<MustUse>(); //~ ERROR: unused `MustUse` that must be used
   |     ^^^^^^^^^^^^^^^^^

error: unused `MustUseMsg` that must be used
  --> $DIR/unused-result.rs:36:5
   |
LL |     foo::<MustUseMsg>(); //~ ERROR: unused `MustUseMsg` that must be used
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = note: some message

error: aborting due to 5 previous errors

