error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:14:1
   |
LL | / const VALS_I8: (i8,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      i8::MAX + 1,
   | |      ----------- attempt to add with overflow
LL | |      );
   | |_______^
   |
note: lint level defined here
  --> $DIR/const-eval-overflow2b.rs:8:9
   |
LL | #![deny(const_err)]
   |         ^^^^^^^^^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:19:1
   |
LL | / const VALS_I16: (i16,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      i16::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:24:1
   |
LL | / const VALS_I32: (i32,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      i32::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:29:1
   |
LL | / const VALS_I64: (i64,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      i64::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:34:1
   |
LL | / const VALS_U8: (u8,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      u8::MAX + 1,
   | |      ----------- attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:39:1
   |
LL | / const VALS_U16: (u16,) = ( //~ ERROR any use of this value will cause an error
LL | |      u16::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:43:1
   |
LL | / const VALS_U32: (u32,) = ( //~ ERROR any use of this value will cause an error
LL | |      u32::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: any use of this value will cause an error
  --> $DIR/const-eval-overflow2b.rs:47:1
   |
LL | / const VALS_U64: (u64,) = //~ ERROR any use of this value will cause an error
LL | |     (
LL | |      u64::MAX + 1,
   | |      ------------ attempt to add with overflow
LL | |      );
   | |_______^

error: aborting due to 8 previous errors

