error: `foo` is not yet stable as a const fn
  --> $DIR/dont_promote_unstable_const_fn.rs:15:25
   |
LL | const fn bar() -> u32 { foo() } //~ ERROR `foo` is not yet stable as a const fn
   |                         ^^^^^
   |
   = help: add `#![feature(foo)]` to the crate attributes to enable

error[E0716]: temporary value dropped while borrowed
  --> $DIR/dont_promote_unstable_const_fn.rs:18:28
   |
LL |     let _: &'static u32 = &foo(); //~ ERROR does not live long enough
   |            ------------    ^^^^^ creates a temporary which is freed while still in use
   |            |
   |            type annotation requires that borrow lasts for `'static`
LL | }
   | - temporary value is freed at the end of this statement

error[E0716]: temporary value dropped while borrowed
  --> $DIR/dont_promote_unstable_const_fn.rs:22:28
   |
LL |     let _: &'static u32 = &meh(); //~ ERROR does not live long enough
   |            ------------    ^^^^^ creates a temporary which is freed while still in use
   |            |
   |            type annotation requires that borrow lasts for `'static`
...
LL | }
   | - temporary value is freed at the end of this statement

error[E0716]: temporary value dropped while borrowed
  --> $DIR/dont_promote_unstable_const_fn.rs:23:26
   |
LL |     let x: &'static _ = &std::time::Duration::from_millis(42).subsec_millis();
   |            ----------    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary which is freed while still in use
   |            |
   |            type annotation requires that borrow lasts for `'static`
LL |     //~^ ERROR does not live long enough
LL | }
   | - temporary value is freed at the end of this statement

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0716`.
