error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_unstable_const_fn_cross_crate.rs:8:28
   |
LL |     let _: &'static u32 = &foo(); //~ ERROR does not live long enough
   |                            ^^^^^ temporary value does not live long enough
LL |     let _x: &'static u32 = &foo(); //~ ERROR does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough
  --> $DIR/dont_promote_unstable_const_fn_cross_crate.rs:9:29
   |
LL |     let _x: &'static u32 = &foo(); //~ ERROR does not live long enough
   |                             ^^^^^ temporary value does not live long enough
LL | }
   | - temporary value only lives until here
   |
   = note: borrowed value must be valid for the static lifetime...

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0597`.
