error[E0019]: constant contains unimplemented expression type
  --> $DIR/issue-52442.rs:2:14
   |
LL |     [();  { &loop { break } as *const _ as usize } ]; //~ ERROR unimplemented expression type
   |              ^^^^^^^^^^^^^^

error[E0080]: it is undefined behavior to use this value
  --> $DIR/issue-52442.rs:2:11
   |
LL |     [();  { &loop { break } as *const _ as usize } ]; //~ ERROR unimplemented expression type
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered a pointer, but expected initialized plain (non-pointer) bytes
   |
   = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rust compiler repository if you believe it should not be considered undefined behavior

error: aborting due to 2 previous errors

Some errors occurred: E0019, E0080.
For more information about an error, try `rustc --explain E0019`.
