// ts-morph/dist/compiler/symbols/Signature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Signature = /** @class */ (function () {
    /**
     * Initializes a new instance of Signature.
     * @private
     * @param context - Project context.
     * @param signature - Compiler signature.
     */
    function Signature(context, signature) {
        this._context = context;
        this._compilerSignature = signature;
    }
    Object.defineProperty(Signature.prototype, "compilerSignature", {
        /**
         * Gets the underlying compiler signature.
         */
        get: function () {
            return this._compilerSignature;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Gets the type parameters.
     */
    Signature.prototype.getTypeParameters = function () {
        var _this = this;
        var typeParameters = this.compilerSignature.typeParameters || [];
        return typeParameters.map(function (t) { return _this._context.compilerFactory.getTypeParameter(t); });
    };
    /**
     * Gets the parameters.
     */
    Signature.prototype.getParameters = function () {
        var _this = this;
        return this.compilerSignature.parameters.map(function (p) { return _this._context.compilerFactory.getSymbol(p); });
    };
    /**
     * Gets the signature return type.
     */
    Signature.prototype.getReturnType = function () {
        return this._context.compilerFactory.getType(this.compilerSignature.getReturnType());
    };
    /**
     * Get the documentation comments.
     */
    Signature.prototype.getDocumentationComments = function () {
        var _this = this;
        var docs = this.compilerSignature.getDocumentationComment(this._context.typeChecker.compilerObject);
        return docs.map(function (d) { return _this._context.compilerFactory.getSymbolDisplayPart(d); });
    };
    /**
     * Gets the JS doc tags.
     */
    Signature.prototype.getJsDocTags = function () {
        var _this = this;
        var tags = this.compilerSignature.getJsDocTags();
        return tags.map(function (t) { return _this._context.compilerFactory.getJSDocTagInfo(t); });
    };
    /**
     * Gets the signature's declaration.
     */
    Signature.prototype.getDeclaration = function () {
        var compilerFactory = this._context.compilerFactory;
        // the compiler says this is non-nullable, but it can return undefined for an unknown signature
        // returned by calling `TypeChecker#getResolvedType()`; however, we're returning undefined in that scenario
        // and so this should never be null (hopefully)
        var compilerSignatureDeclaration = this.compilerSignature.getDeclaration();
        return compilerFactory.getNodeFromCompilerNode(compilerSignatureDeclaration, compilerFactory.getSourceFileForNode(compilerSignatureDeclaration));
    };
    return Signature;
}());
exports.Signature = Signature;
