error[E0392]: parameter `'a` is never used
  --> $DIR/variance-regions-unused-direct.rs:5:18
   |
LL | struct Bivariant<'a>; //~ ERROR parameter `'a` is never used
   |                  ^^ unused type parameter
   |
   = help: consider removing `'a` or using a marker such as `std::marker::PhantomData`

error[E0392]: parameter `'d` is never used
  --> $DIR/variance-regions-unused-direct.rs:7:19
   |
LL | struct Struct<'a, 'd> { //~ ERROR parameter `'d` is never used
   |                   ^^ unused type parameter
   |
   = help: consider removing `'d` or using a marker such as `std::marker::PhantomData`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0392`.
