error[E0392]: parameter `'a` is never used
  --> $DIR/variance-unused-region-param.rs:3:19
   |
LL | struct SomeStruct<'a> { x: u32 } //~ ERROR parameter `'a` is never used
   |                   ^^ unused type parameter
   |
   = help: consider removing `'a` or using a marker such as `std::marker::PhantomData`

error[E0392]: parameter `'a` is never used
  --> $DIR/variance-unused-region-param.rs:4:15
   |
LL | enum SomeEnum<'a> { Nothing } //~ ERROR parameter `'a` is never used
   |               ^^ unused type parameter
   |
   = help: consider removing `'a` or using a marker such as `std::marker::PhantomData`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0392`.
