error[E0208]: [+, -, o, *]
  --> $DIR/variance-regions-indirect.rs:8:1
   |
LL | / enum Base<'a, 'b, 'c:'b, 'd> { //~ ERROR [+, -, o, *]
LL | |     Test8A(extern "Rust" fn(&'a isize)),
LL | |     Test8B(&'b [isize]),
LL | |     Test8C(&'b mut &'c str),
LL | | }
   | |_^

error[E0208]: [*, o, -, +]
  --> $DIR/variance-regions-indirect.rs:15:1
   |
LL | / struct Derived1<'w, 'x:'y, 'y, 'z> { //~ ERROR [*, o, -, +]
LL | |     f: Base<'z, 'y, 'x, 'w>
LL | | }
   | |_^

error[E0208]: [o, o, *]
  --> $DIR/variance-regions-indirect.rs:20:1
   |
LL | / struct Derived2<'a, 'b:'a, 'c> { //~ ERROR [o, o, *]
LL | |     f: Base<'a, 'a, 'b, 'c>
LL | | }
   | |_^

error[E0208]: [o, -, *]
  --> $DIR/variance-regions-indirect.rs:25:1
   |
LL | / struct Derived3<'a:'b, 'b, 'c> { //~ ERROR [o, -, *]
LL | |     f: Base<'a, 'b, 'a, 'c>
LL | | }
   | |_^

error[E0208]: [+, -, o]
  --> $DIR/variance-regions-indirect.rs:30:1
   |
LL | / struct Derived4<'a, 'b, 'c:'b> { //~ ERROR [+, -, o]
LL | |     f: Base<'a, 'b, 'c, 'a>
LL | | }
   | |_^

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0208`.
