error[E0026]: variant `X::Y` does not have a field named `number`
  --> $DIR/issue-41314.rs:7:16
   |
LL |         X::Y { number } => {} //~ ERROR does not have a field named `number`
   |                ^^^^^^ variant `X::Y` does not have this field

error[E0027]: pattern does not mention field `0`
  --> $DIR/issue-41314.rs:7:9
   |
LL |         X::Y { number } => {} //~ ERROR does not have a field named `number`
   |         ^^^^^^^^^^^^^^^ missing field `0`
   |
   = note: trying to match a tuple variant with a struct variant pattern

error: aborting due to 2 previous errors

Some errors occurred: E0026, E0027.
For more information about an error, try `rustc --explain E0026`.
