error[E0070]: invalid left-hand side expression
  --> $DIR/E0070.rs:16:5
   |
LL |     SOME_CONST = 14; //~ ERROR E0070
   |     ^^^^^^^^^^^^^^^ left-hand of expression not valid

error[E0070]: invalid left-hand side expression
  --> $DIR/E0070.rs:17:5
   |
LL |     1 = 3; //~ ERROR E0070
   |     ^^^^^ left-hand of expression not valid

error[E0308]: mismatched types
  --> $DIR/E0070.rs:18:25
   |
LL |     some_other_func() = 4; //~ ERROR E0070
   |                         ^ expected (), found integral variable
   |
   = note: expected type `()`
              found type `{integer}`

error[E0070]: invalid left-hand side expression
  --> $DIR/E0070.rs:18:5
   |
LL |     some_other_func() = 4; //~ ERROR E0070
   |     ^^^^^^^^^^^^^^^^^^^^^ left-hand of expression not valid

error: aborting due to 4 previous errors

Some errors occurred: E0070, E0308.
For more information about an error, try `rustc --explain E0070`.
