error: attempt to shift left with overflow
  --> $DIR/E0080.rs:12:9
   |
LL |     X = (1 << 500), //~ ERROR E0080
   |         ^^^^^^^^^^
   |
   = note: #[deny(exceeding_bitshifts)] on by default

error[E0080]: could not evaluate enum discriminant
  --> $DIR/E0080.rs:12:9
   |
LL |     X = (1 << 500), //~ ERROR E0080
   |         ^^^^^^^^^^ attempt to shift left with overflow

error: attempt to divide by zero
  --> $DIR/E0080.rs:14:9
   |
LL |     Y = (1 / 0) //~ ERROR E0080
   |         ^^^^^^^
   |
   = note: #[deny(const_err)] on by default

error: this expression will panic at runtime
  --> $DIR/E0080.rs:14:9
   |
LL |     Y = (1 / 0) //~ ERROR E0080
   |         ^^^^^^^ attempt to divide by zero

error[E0080]: could not evaluate enum discriminant
  --> $DIR/E0080.rs:14:9
   |
LL |     Y = (1 / 0) //~ ERROR E0080
   |         ^^^^^^^ attempt to divide by zero

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0080`.
