error[E0120]: the Drop trait may only be implemented on structures
  --> $DIR/E0117.rs:1:15
   |
LL | impl Drop for u32 {} //~ ERROR E0117
   |               ^^^ implementing Drop requires a struct

error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
  --> $DIR/E0117.rs:1:1
   |
LL | impl Drop for u32 {} //~ ERROR E0117
   | ^^^^^^^^^^^^^^^^^ impl doesn't use types inside crate
   |
   = note: the impl does not reference any types defined in this crate
   = note: define and implement a trait or new type instead

error: aborting due to 2 previous errors

Some errors occurred: E0117, E0120.
For more information about an error, try `rustc --explain E0117`.
