error[E0121]: the type placeholder `_` is not allowed within types on item signatures
  --> $DIR/E0121.rs:1:13
   |
LL | fn foo() -> _ { 5 } //~ ERROR E0121
   |             ^ not allowed in type signatures

error[E0121]: the type placeholder `_` is not allowed within types on item signatures
  --> $DIR/E0121.rs:3:13
   |
LL | static BAR: _ = "test"; //~ ERROR E0121
   |             ^ not allowed in type signatures

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0121`.
