error[E0201]: duplicate definitions with name `bar`:
  --> $DIR/E0201.rs:15:5
   |
LL |     fn bar(&self) -> bool { self.0 > 5 }
   |     ------------------------------------ previous definition of `bar` here
LL |     fn bar() {} //~ ERROR E0201
   |     ^^^^^^^^^^^ duplicate definition

error[E0201]: duplicate definitions with name `baz`:
  --> $DIR/E0201.rs:27:5
   |
LL |     fn baz(&self) -> bool { true }
   |     ------------------------------ previous definition of `baz` here
LL |     fn baz(&self) -> bool { self.0 > 5 } //~ ERROR E0201
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ duplicate definition

error[E0201]: duplicate definitions with name `Quux`:
  --> $DIR/E0201.rs:28:5
   |
LL |     type Quux = u32;
   |     ---------------- previous definition of `Quux` here
...
LL |     type Quux = u32; //~ ERROR E0201
   |     ^^^^^^^^^^^^^^^^ duplicate definition

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0201`.
