error[E0017]: references in constants may only refer to immutable values
  --> $DIR/E0388.rs:14:30
   |
LL | const CR: &'static mut i32 = &mut C; //~ ERROR E0017
   |                              ^^^^^^ constants require immutable values

error[E0017]: references in statics may only refer to immutable values
  --> $DIR/E0388.rs:15:39
   |
LL | static STATIC_REF: &'static mut i32 = &mut X; //~ ERROR E0017
   |                                       ^^^^^^ statics require immutable values

error[E0017]: references in statics may only refer to immutable values
  --> $DIR/E0388.rs:17:38
   |
LL | static CONST_REF: &'static mut i32 = &mut C; //~ ERROR E0017
   |                                      ^^^^^^ statics require immutable values

error[E0596]: cannot borrow immutable item `X` as mutable
  --> $DIR/E0388.rs:15:39
   |
LL | static STATIC_REF: &'static mut i32 = &mut X; //~ ERROR E0017
   |                                       ^^^^^^ cannot borrow as mutable

error: aborting due to 4 previous errors

Some errors occurred: E0017, E0596.
For more information about an error, try `rustc --explain E0017`.
