error[E0637]: `'_` cannot be used here
  --> $DIR/E0637.rs:1:16
   |
LL | struct Foo<'a: '_>(&'a u8); //~ ERROR cannot be used here
   |                ^^ `'_` is a reserved lifetime name

error[E0637]: `'_` cannot be used here
  --> $DIR/E0637.rs:2:12
   |
LL | fn foo<'a: '_>(_: &'a u8) {} //~ ERROR cannot be used here
   |            ^^ `'_` is a reserved lifetime name

error[E0637]: `'_` cannot be used here
  --> $DIR/E0637.rs:5:10
   |
LL | impl<'a: '_> Bar<'a> { //~ ERROR cannot be used here
   |          ^^ `'_` is a reserved lifetime name

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0637`.
