error[E0423]: expected value, found type alias `m1::S`
  --> $DIR/namespace-mix.rs:34:11
   |
LL |     check(m1::S); //~ ERROR expected value, found type alias `m1::S`
   |           ^^^^-
   |               |
   |               did you mean `TS`?
   |
   = note: can't use a type alias as a constructor
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use m2::S;
   |
LL | use namespace_mix::xm2::S;
   |

error[E0423]: expected value, found type alias `xm1::S`
  --> $DIR/namespace-mix.rs:40:11
   |
LL |     check(xm1::S); //~ ERROR expected value, found type alias `xm1::S`
   |           ^^^^^-
   |                |
   |                did you mean `TS`?
   |
   = note: can't use a type alias as a constructor
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use m2::S;
   |
LL | use namespace_mix::xm2::S;
   |

error[E0423]: expected value, found struct variant `m7::V`
  --> $DIR/namespace-mix.rs:100:11
   |
LL |     check(m7::V); //~ ERROR expected value, found struct variant `m7::V`
   |           ^^^^-
   |           |   |
   |           |   did you mean `TV`?
   |           did you mean `m7::V { /* fields */ }`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use m8::V;
   |
LL | use namespace_mix::xm8::V;
   |

error[E0423]: expected value, found struct variant `xm7::V`
  --> $DIR/namespace-mix.rs:106:11
   |
LL |     check(xm7::V); //~ ERROR expected value, found struct variant `xm7::V`
   |           ^^^^^-
   |           |    |
   |           |    did you mean `TV`?
   |           did you mean `xm7::V { /* fields */ }`?
help: possible better candidates are found in other modules, you can import them into scope
   |
LL | use m8::V;
   |
LL | use namespace_mix::xm8::V;
   |

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:33:5
   |
LL |     check(m1::S{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::S: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:35:5
   |
LL |     check(m2::S{}); //~ ERROR c::S
   |     ^^^^^ the trait `Impossible` is not implemented for `c::S`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:36:5
   |
LL |     check(m2::S); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:39:5
   |
LL |     check(xm1::S{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::S: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:41:5
   |
LL |     check(xm2::S{}); //~ ERROR c::S
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::S`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:42:5
   |
LL |     check(xm2::S); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:55:5
   |
LL |     check(m3::TS{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn() -> c::TS {c::TS}: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:56:5
   |
LL |     check(m3::TS); //~ ERROR c::TS
   |     ^^^^^ the trait `Impossible` is not implemented for `fn() -> c::TS {c::TS}`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::TS: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:57:5
   |
LL |     check(m4::TS{}); //~ ERROR c::TS
   |     ^^^^^ the trait `Impossible` is not implemented for `c::TS`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:58:5
   |
LL |     check(m4::TS); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:61:5
   |
LL |     check(xm3::TS{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn() -> namespace_mix::c::TS {namespace_mix::c::TS}: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:62:5
   |
LL |     check(xm3::TS); //~ ERROR c::TS
   |     ^^^^^ the trait `Impossible` is not implemented for `fn() -> namespace_mix::c::TS {namespace_mix::c::TS}`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::TS: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:63:5
   |
LL |     check(xm4::TS{}); //~ ERROR c::TS
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::TS`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:64:5
   |
LL |     check(xm4::TS); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:77:5
   |
LL |     check(m5::US{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::US: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:78:5
   |
LL |     check(m5::US); //~ ERROR c::US
   |     ^^^^^ the trait `Impossible` is not implemented for `c::US`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::US: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:79:5
   |
LL |     check(m6::US{}); //~ ERROR c::US
   |     ^^^^^ the trait `Impossible` is not implemented for `c::US`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:80:5
   |
LL |     check(m6::US); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:83:5
   |
LL |     check(xm5::US{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::US: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:84:5
   |
LL |     check(xm5::US); //~ ERROR c::US
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::US`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::US: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:85:5
   |
LL |     check(xm6::US{}); //~ ERROR c::US
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::US`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:86:5
   |
LL |     check(xm6::US); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:99:5
   |
LL |     check(m7::V{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:101:5
   |
LL |     check(m8::V{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:102:5
   |
LL |     check(m8::V); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:105:5
   |
LL |     check(xm7::V{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:107:5
   |
LL |     check(xm8::V{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:108:5
   |
LL |     check(xm8::V); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:121:5
   |
LL |     check(m9::TV{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn() -> c::E {c::E::TV}: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:122:5
   |
LL |     check(m9::TV); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `fn() -> c::E {c::E::TV}`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:123:5
   |
LL |     check(mA::TV{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:124:5
   |
LL |     check(mA::TV); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:127:5
   |
LL |     check(xm9::TV{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `fn() -> namespace_mix::c::E {namespace_mix::c::E::TV}: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:128:5
   |
LL |     check(xm9::TV); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `fn() -> namespace_mix::c::E {namespace_mix::c::E::TV}`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:129:5
   |
LL |     check(xmA::TV{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:130:5
   |
LL |     check(xmA::TV); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:143:5
   |
LL |     check(mB::UV{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:144:5
   |
LL |     check(mB::UV); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:145:5
   |
LL |     check(mC::UV{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:146:5
   |
LL |     check(mC::UV); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:149:5
   |
LL |     check(xmB::UV{}); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:150:5
   |
LL |     check(xmB::UV); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::E: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:151:5
   |
LL |     check(xmC::UV{}); //~ ERROR c::E
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::E`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `namespace_mix::c::Item: Impossible` is not satisfied
  --> $DIR/namespace-mix.rs:152:5
   |
LL |     check(xmC::UV); //~ ERROR c::Item
   |     ^^^^^ the trait `Impossible` is not implemented for `namespace_mix::c::Item`
   |
note: required by `check`
  --> $DIR/namespace-mix.rs:21:1
   |
LL | fn check<T: Impossible>(_: T) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 48 previous errors

Some errors occurred: E0277, E0423.
For more information about an error, try `rustc --explain E0277`.
