error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:41:27
   |
LL |     let &(X(_t), X(_u)) = &(x.clone(), x.clone());
   |         ---------------   ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |         |   |      |
   |         |   |      ...and here
   |         |   data moved here
   |         help: consider removing the `&`: `(X(_t), X(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:41:13
   |
LL |     let &(X(_t), X(_u)) = &(x.clone(), x.clone());
   |             ^^     ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:45:50
   |
LL |     if let &(Either::One(_t), Either::Two(_u)) = &(e.clone(), e.clone()) { }
   |            -----------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |            |             |                |
   |            |             |                ...and here
   |            |             data moved here
   |            help: consider removing the `&`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:45:26
   |
LL |     if let &(Either::One(_t), Either::Two(_u)) = &(e.clone(), e.clone()) { }
   |                          ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:49:53
   |
LL |     while let &(Either::One(_t), Either::Two(_u)) = &(e.clone(), e.clone()) { }
   |               -----------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |               |             |                |
   |               |             |                ...and here
   |               |             data moved here
   |               help: consider removing the `&`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:49:29
   |
LL |     while let &(Either::One(_t), Either::Two(_u)) = &(e.clone(), e.clone()) { }
   |                             ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:53:11
   |
LL |     match &(e.clone(), e.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |                       --               -- ...and here
   |                       |
   |                       data moved here
...
LL |         &(Either::Two(_t), Either::One(_u)) => (),
   |                       -- ...and here   -- ...and here
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:55:23
   |
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |                       ^^               ^^
...
LL |         &(Either::Two(_t), Either::One(_u)) => (),
   |                       ^^               ^^
help: consider removing the `&`
   |
LL |         (Either::One(_t), Either::Two(_u)) => (),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider removing the `&`
   |
LL |         (Either::Two(_t), Either::One(_u)) => (),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:63:11
   |
LL |     match &(e.clone(), e.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &(Either::One(_t), Either::Two(_u))
   |         -----------------------------------
   |         |             |                |
   |         |             |                ...and here
   |         |             data moved here
   |         help: consider removing the `&`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:65:23
   |
LL |         &(Either::One(_t), Either::Two(_u))
   |                       ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:72:11
   |
LL |     match &(e.clone(), e.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |         -----------------------------------
   |         |             |                |
   |         |             |                ...and here
   |         |             data moved here
   |         help: consider removing the `&`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:74:23
   |
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |                       ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:80:11
   |
LL |     match &(e.clone(), e.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |         -----------------------------------
   |         |             |                |
   |         |             |                ...and here
   |         |             data moved here
   |         help: consider removing the `&`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:82:23
   |
LL |         &(Either::One(_t), Either::Two(_u)) => (),
   |                       ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:93:31
   |
LL |     let &mut (X(_t), X(_u)) = &mut (xm.clone(), xm.clone());
   |         -------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |         |       |      |
   |         |       |      ...and here
   |         |       data moved here
   |         help: consider removing the `&mut`: `(X(_t), X(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:93:17
   |
LL |     let &mut (X(_t), X(_u)) = &mut (xm.clone(), xm.clone());
   |                 ^^     ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:97:54
   |
LL |     if let &mut (Either::One(_t), Either::Two(_u)) = &mut (em.clone(), em.clone()) { }
   |            ---------------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |            |                 |                |
   |            |                 |                ...and here
   |            |                 data moved here
   |            help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:97:30
   |
LL |     if let &mut (Either::One(_t), Either::Two(_u)) = &mut (em.clone(), em.clone()) { }
   |                              ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:101:57
   |
LL |     while let &mut (Either::One(_t), Either::Two(_u)) = &mut (em.clone(), em.clone()) { }
   |               ---------------------------------------   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
   |               |                 |                |
   |               |                 |                ...and here
   |               |                 data moved here
   |               help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:101:33
   |
LL |     while let &mut (Either::One(_t), Either::Two(_u)) = &mut (em.clone(), em.clone()) { }
   |                                 ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:105:11
   |
LL |     match &mut (em.clone(), em.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |                           --               -- ...and here
   |                           |
   |                           data moved here
...
LL |         &mut (Either::Two(_t), Either::One(_u)) => (),
   |                           -- ...and here   -- ...and here
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:107:27
   |
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |                           ^^               ^^
...
LL |         &mut (Either::Two(_t), Either::One(_u)) => (),
   |                           ^^               ^^
help: consider removing the `&mut`
   |
LL |         (Either::One(_t), Either::Two(_u)) => (),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: consider removing the `&mut`
   |
LL |         (Either::Two(_t), Either::One(_u)) => (),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:115:11
   |
LL |     match &mut (em.clone(), em.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &mut (Either::One(_t), Either::Two(_u))
   |         ---------------------------------------
   |         |                 |                |
   |         |                 |                ...and here
   |         |                 data moved here
   |         help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:117:27
   |
LL |         &mut (Either::One(_t), Either::Two(_u))
   |                           ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:124:11
   |
LL |     match &mut (em.clone(), em.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |         ---------------------------------------
   |         |                 |                |
   |         |                 |                ...and here
   |         |                 data moved here
   |         help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:126:27
   |
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |                           ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:132:11
   |
LL |     match &mut (em.clone(), em.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |         ---------------------------------------
   |         |                 |                |
   |         |                 |                ...and here
   |         |                 data moved here
   |         help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:134:27
   |
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |                           ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:140:11
   |
LL |     match &mut (em.clone(), em.clone()) {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot move out of borrowed content
LL |         //~^ ERROR cannot move
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |         ---------------------------------------
   |         |                 |                |
   |         |                 |                ...and here
   |         |                 data moved here
   |         help: consider removing the `&mut`: `(Either::One(_t), Either::Two(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:142:27
   |
LL |         &mut (Either::One(_t), Either::Two(_u)) => (),
   |                           ^^               ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:88:11
   |
LL |     fn f5(&(X(_t), X(_u)): &(X, X)) { }
   |           ^^^^--^^^^^--^^
   |           |   |      |
   |           |   |      ...and here
   |           |   data moved here
   |           cannot move out of borrowed content
   |           help: consider removing the `&`: `(X(_t), X(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:88:15
   |
LL |     fn f5(&(X(_t), X(_u)): &(X, X)) { }
   |               ^^     ^^

error[E0507]: cannot move out of borrowed content
  --> $DIR/duplicate-suggestions.rs:148:11
   |
LL |     fn f6(&mut (X(_t), X(_u)): &mut (X, X)) { }
   |           ^^^^^^^^--^^^^^--^^
   |           |       |      |
   |           |       |      ...and here
   |           |       data moved here
   |           cannot move out of borrowed content
   |           help: consider removing the `&mut`: `(X(_t), X(_u))`
   |
note: move occurs because these variables have types that don't implement the `Copy` trait
  --> $DIR/duplicate-suggestions.rs:148:19
   |
LL |     fn f6(&mut (X(_t), X(_u)): &mut (X, X)) { }
   |                   ^^     ^^

error: aborting due to 17 previous errors

For more information about this error, try `rustc --explain E0507`.
