error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:9:13
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
LL |     move || x = 1; //~ ERROR cannot assign
   |             ^^^^^ cannot assign

error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:10:17
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
LL |     move || x = 1; //~ ERROR cannot assign
LL |     move || set(&mut x); //~ ERROR cannot borrow
   |                 ^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:11:13
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     move || x = 1; //~ ERROR cannot assign
   |             ^^^^^ cannot assign

error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:12:17
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     move || set(&mut x); //~ ERROR cannot borrow
   |                 ^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:13:8
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     || x = 1; //~ ERROR cannot assign
   |        ^^^^^ cannot assign

error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:15:12
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     || set(&mut x); //~ ERROR cannot assign
   |            ^^^^^^ cannot borrow as mutable

error[E0594]: cannot assign to `x`, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:16:8
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     || x = 1; //~ ERROR cannot assign
   |        ^^^^^ cannot assign

error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:18:12
   |
LL |     let x = 0;
   |         - help: consider changing this to be mutable: `mut x`
...
LL |     || set(&mut x); //~ ERROR cannot assign
   |            ^^^^^^ cannot borrow as mutable

error: aborting due to 8 previous errors

Some errors occurred: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
