error[E0595]: closure cannot assign to immutable local variable `x`
  --> $DIR/unboxed-closure-immutable-capture.rs:13:5
   |
LL |     let x = 0;
   |         - help: make this binding mutable: `mut x`
...
LL |     || x = 1; //~ ERROR cannot assign
   |     ^^ cannot borrow mutably

error[E0595]: closure cannot assign to immutable local variable `x`
  --> $DIR/unboxed-closure-immutable-capture.rs:15:5
   |
LL |     let x = 0;
   |         - help: make this binding mutable: `mut x`
...
LL |     || set(&mut x); //~ ERROR cannot assign
   |     ^^ cannot borrow mutably

error[E0595]: closure cannot assign to immutable local variable `x`
  --> $DIR/unboxed-closure-immutable-capture.rs:16:5
   |
LL |     let x = 0;
   |         - help: make this binding mutable: `mut x`
...
LL |     || x = 1; //~ ERROR cannot assign
   |     ^^ cannot borrow mutably

error[E0595]: closure cannot assign to immutable local variable `x`
  --> $DIR/unboxed-closure-immutable-capture.rs:18:5
   |
LL |     let x = 0;
   |         - help: make this binding mutable: `mut x`
...
LL |     || set(&mut x); //~ ERROR cannot assign
   |     ^^ cannot borrow mutably

error[E0594]: cannot assign to captured outer variable in an `FnMut` closure
  --> $DIR/unboxed-closure-immutable-capture.rs:9:13
   |
LL |     let x = 0;
   |         - help: consider making `x` mutable: `mut x`
LL |     move || x = 1; //~ ERROR cannot assign
   |             ^^^^^

error[E0596]: cannot borrow captured outer variable in an `FnMut` closure as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:10:22
   |
LL |     move || set(&mut x); //~ ERROR cannot borrow
   |                      ^

error[E0594]: cannot assign to captured outer variable in an `FnMut` closure
  --> $DIR/unboxed-closure-immutable-capture.rs:11:13
   |
LL |     let x = 0;
   |         - help: consider making `x` mutable: `mut x`
...
LL |     move || x = 1; //~ ERROR cannot assign
   |             ^^^^^

error[E0596]: cannot borrow captured outer variable in an `FnMut` closure as mutable
  --> $DIR/unboxed-closure-immutable-capture.rs:12:22
   |
LL |     move || set(&mut x); //~ ERROR cannot borrow
   |                      ^

error: aborting due to 8 previous errors

Some errors occurred: E0594, E0595, E0596.
For more information about an error, try `rustc --explain E0594`.
