error: rustc_peek: bit not set
  --> $DIR/uninits-1.rs:19:14
   |
LL |     unsafe { rustc_peek(&x) }; //~ ERROR rustc_peek: bit not set
   |              ^^^^^^^^^^^^^^

error: rustc_peek: bit not set
  --> $DIR/uninits-1.rs:20:14
   |
LL |     unsafe { rustc_peek(&y) }; //~ ERROR rustc_peek: bit not set
   |              ^^^^^^^^^^^^^^

error: rustc_peek: bit not set
  --> $DIR/uninits-1.rs:21:14
   |
LL |     unsafe { rustc_peek(&z) }; //~ ERROR rustc_peek: bit not set
   |              ^^^^^^^^^^^^^^

error: rustc_peek: bit not set
  --> $DIR/uninits-1.rs:37:14
   |
LL |     unsafe { rustc_peek(&x); } //~ ERROR rustc_peek: bit not set
   |              ^^^^^^^^^^^^^^

error: rustc_peek: bit not set
  --> $DIR/uninits-1.rs:45:14
   |
LL |     unsafe { rustc_peek(&ret); } //~ ERROR rustc_peek: bit not set
   |              ^^^^^^^^^^^^^^^^

error: stop_after_dataflow ended compilation

error: aborting due to 6 previous errors

