error: incorrect close delimiter: `}`
  --> $DIR/issue-10636-2.rs:18:1
   |
LL | } //~ ERROR: incorrect close delimiter
   | ^
   |
note: unclosed delimiter
  --> $DIR/issue-10636-2.rs:15:15
   |
LL |     option.map(|some| 42;
   |               ^

error: expected one of `)`, `,`, `.`, `?`, or an operator, found `;`
  --> $DIR/issue-10636-2.rs:15:25
   |
LL |     option.map(|some| 42;
   |                         ^ expected one of `)`, `,`, `.`, `?`, or an operator here

error: expected expression, found `)`
  --> $DIR/issue-10636-2.rs:18:1
   |
LL | } //~ ERROR: incorrect close delimiter
   | ^ expected expression

error: aborting due to 3 previous errors

