error: macro expansion ignores token `,` and any following
  --> $DIR/macro-incomplete-parse.rs:17:9
   |
LL |         , //~ ERROR macro expansion ignores token `,`
   |         ^
   |
note: caused by the macro expansion here; the usage of `ignored_item!` is likely invalid in item context
  --> $DIR/macro-incomplete-parse.rs:31:1
   |
LL | ignored_item!();
   | ^^^^^^^^^^^^^^^^

error: expected one of `.`, `;`, `?`, `}`, or an operator, found `,`
  --> $DIR/macro-incomplete-parse.rs:22:14
   |
LL |     () => ( 1,  //~ ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `,`
   |              ^ expected one of `.`, `;`, `?`, `}`, or an operator here
...
LL |     ignored_expr!();
   |     ---------------- in this macro invocation

error: macro expansion ignores token `,` and any following
  --> $DIR/macro-incomplete-parse.rs:28:14
   |
LL |     () => ( 1, 2 ) //~ ERROR macro expansion ignores token `,`
   |              ^
   |
note: caused by the macro expansion here; the usage of `ignored_pat!` is likely invalid in pattern context
  --> $DIR/macro-incomplete-parse.rs:36:9
   |
LL |         ignored_pat!() => (),
   |         ^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

