error[E0060]: this function takes at least 2 parameters but 0 parameters were supplied
  --> $DIR/variadic-ffi-3.rs:11:9
   |
LL |     fn foo(f: isize, x: u8, ...);
   |     ----------------------------- defined here
...
LL |         foo(); //~ ERROR: this function takes at least 2 parameters but 0 parameters were supplied
   |         ^^^^^ expected at least 2 parameters

error[E0060]: this function takes at least 2 parameters but 1 parameter was supplied
  --> $DIR/variadic-ffi-3.rs:12:9
   |
LL |     fn foo(f: isize, x: u8, ...);
   |     ----------------------------- defined here
...
LL |         foo(1); //~ ERROR: this function takes at least 2 parameters but 1 parameter was supplied
   |         ^^^^^^ expected at least 2 parameters

error[E0308]: mismatched types
  --> $DIR/variadic-ffi-3.rs:14:56
   |
LL |         let x: unsafe extern "C" fn(f: isize, x: u8) = foo;
   |                                                        ^^^ expected non-variadic fn, found variadic function
   |
   = note: expected type `unsafe extern "C" fn(isize, u8)`
              found type `unsafe extern "C" fn(isize, u8, ...) {foo}`

error[E0308]: mismatched types
  --> $DIR/variadic-ffi-3.rs:19:54
   |
LL |         let y: extern "C" fn(f: isize, x: u8, ...) = bar;
   |                                                      ^^^ expected variadic fn, found non-variadic function
   |
   = note: expected type `extern "C" fn(isize, u8, ...)`
              found type `extern "C" fn(isize, u8) {bar}`

error[E0617]: can't pass `f32` to variadic function
  --> $DIR/variadic-ffi-3.rs:24:19
   |
LL |         foo(1, 2, 3f32); //~ ERROR can't pass `f32` to variadic function
   |                   ^^^^ help: cast the value to `c_double`: `3f32 as c_double`

error[E0617]: can't pass `bool` to variadic function
  --> $DIR/variadic-ffi-3.rs:25:19
   |
LL |         foo(1, 2, true); //~ ERROR can't pass `bool` to variadic function
   |                   ^^^^ help: cast the value to `c_int`: `true as c_int`

error[E0617]: can't pass `i8` to variadic function
  --> $DIR/variadic-ffi-3.rs:26:19
   |
LL |         foo(1, 2, 1i8); //~ ERROR can't pass `i8` to variadic function
   |                   ^^^ help: cast the value to `c_int`: `1i8 as c_int`

error[E0617]: can't pass `u8` to variadic function
  --> $DIR/variadic-ffi-3.rs:27:19
   |
LL |         foo(1, 2, 1u8); //~ ERROR can't pass `u8` to variadic function
   |                   ^^^ help: cast the value to `c_uint`: `1u8 as c_uint`

error[E0617]: can't pass `i16` to variadic function
  --> $DIR/variadic-ffi-3.rs:28:19
   |
LL |         foo(1, 2, 1i16); //~ ERROR can't pass `i16` to variadic function
   |                   ^^^^ help: cast the value to `c_int`: `1i16 as c_int`

error[E0617]: can't pass `u16` to variadic function
  --> $DIR/variadic-ffi-3.rs:29:19
   |
LL |         foo(1, 2, 1u16); //~ ERROR can't pass `u16` to variadic function
   |                   ^^^^ help: cast the value to `c_uint`: `1u16 as c_uint`

error: aborting due to 10 previous errors

Some errors occurred: E0060, E0308, E0617.
For more information about an error, try `rustc --explain E0060`.
