error: imports cannot refer to type parameters
  --> $DIR/future-proofing-locals.rs:14:9
   |
LL |     use T as _; //~ ERROR imports cannot refer to type parameters
   |         ^

error: imports cannot refer to type parameters
  --> $DIR/future-proofing-locals.rs:15:9
   |
LL |     use T::U; //~ ERROR imports cannot refer to type parameters
   |         ^

error: imports cannot refer to type parameters
  --> $DIR/future-proofing-locals.rs:16:9
   |
LL |     use T::*; //~ ERROR imports cannot refer to type parameters
   |         ^

error: imports cannot refer to local variables
  --> $DIR/future-proofing-locals.rs:26:9
   |
LL |     use x as _; //~ ERROR imports cannot refer to local variables
   |         ^

error: imports cannot refer to local variables
  --> $DIR/future-proofing-locals.rs:32:9
   |
LL |     use x; //~ ERROR imports cannot refer to local variables
   |         ^

error: imports cannot refer to local variables
  --> $DIR/future-proofing-locals.rs:38:17
   |
LL |             use x; //~ ERROR imports cannot refer to local variables
   |                 ^

error: imports cannot refer to type parameters
  --> $DIR/future-proofing-locals.rs:46:10
   |
LL |     use {T as _, x}; //~ ERROR imports cannot refer to type parameters
   |          ^

error: imports cannot refer to local variables
  --> $DIR/future-proofing-locals.rs:46:18
   |
LL |     use {T as _, x}; //~ ERROR imports cannot refer to type parameters
   |                  ^

error: aborting due to 8 previous errors

