error[E0432]: unresolved import `foo`
  --> $DIR/local-path-suggestions-2018.rs:12:9
   |
LL |     use foo::Bar; //~ ERROR unresolved import `foo`
   |         ^^^ did you mean `crate::foo`?
   |
   = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>

error[E0432]: unresolved import `foobar`
  --> $DIR/local-path-suggestions-2018.rs:21:5
   |
LL | use foobar::Baz; //~ ERROR unresolved import `foobar`
   |     ^^^^^^ did you mean `baz::foobar`?

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
