error[E0599]: no method named `foobar` found for type `u32` in the current scope
  --> $DIR/trait-import-suggestions.rs:22:11
   |
LL |         x.foobar(); //~ ERROR no method named `foobar`
   |           ^^^^^^
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           `use crate::foo::foobar::Foobar;`

error[E0599]: no method named `bar` found for type `u32` in the current scope
  --> $DIR/trait-import-suggestions.rs:28:7
   |
LL |     x.bar(); //~ ERROR no method named `bar`
   |       ^^^
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
   |
LL | use crate::foo::Bar;
   |

error[E0599]: no method named `baz` found for type `u32` in the current scope
  --> $DIR/trait-import-suggestions.rs:29:7
   |
LL |     x.baz(); //~ ERROR no method named `baz`
   |       ^^^

error[E0599]: no function or associated item named `from_str` found for type `u32` in the current scope
  --> $DIR/trait-import-suggestions.rs:30:18
   |
LL |     let y = u32::from_str("33"); //~ ERROR no function or associated item named `from_str`
   |             -----^^^^^^^^
   |             |
   |             function or associated item not found in `u32`
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
   |
LL | use std::str::FromStr;
   |

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
