warning: not reporting region error due to nll
  --> $DIR/propagate-approximated-fail-no-postdom.rs:55:21
   |
LL |             let p = x.get();
   |                     ^^^^^^^

error: free region `ReFree(DefId(0/1:20 ~ propagate_approximated_fail_no_postdom[317d]::supply[0]::{{closure}}[0]), BrAnon(1))` does not outlive free region `ReFree(DefId(0/1:20 ~ propagate_approximated_fail_no_postdom[317d]::supply[0]::{{closure}}[0]), BrAnon(2))`
  --> $DIR/propagate-approximated-fail-no-postdom.rs:55:17
   |
LL |             let p = x.get();
   |                 ^

note: No external requirements
  --> $DIR/propagate-approximated-fail-no-postdom.rs:53:9
   |
LL | /         |_outlives1, _outlives2, _outlives3, x, y| {
LL | |             // Only works if 'x: 'y:
LL | |             let p = x.get();
LL | |             //~^ WARN not reporting region error due to nll
LL | |             //~| ERROR does not outlive free region
LL | |             demand_y(x, y, p)
LL | |         },
   | |_________^
   |
   = note: defining type: DefId(0/1:20 ~ propagate_approximated_fail_no_postdom[317d]::supply[0]::{{closure}}[0]) with closure substs [
               i16,
               for<'r, 's> extern "rust-call" fn((std::cell::Cell<&'_#1r &ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 'r)) u32>, std::cell::Cell<&'_#2r &ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 'r)) u32>, std::cell::Cell<&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 's)) &'_#3r u32>, std::cell::Cell<&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 'r)) u32>, std::cell::Cell<&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 's)) u32>))
           ]

note: No external requirements
  --> $DIR/propagate-approximated-fail-no-postdom.rs:48:1
   |
LL | / fn supply<'a, 'b, 'c>(cell_a: Cell<&'a u32>, cell_b: Cell<&'b u32>, cell_c: Cell<&'c u32>) {
LL | |     establish_relationships(
LL | |         cell_a,
LL | |         cell_b,
...  |
LL | |     );
LL | | }
   | |_^
   |
   = note: defining type: DefId(0/0:6 ~ propagate_approximated_fail_no_postdom[317d]::supply[0]) with substs []

error: aborting due to previous error

