warning: not reporting region error due to nll
  --> $DIR/propagate-approximated-shorter-to-static-no-bound.rs:49:9
   |
LL |         demand_y(x, y, x.get()) //~ WARNING not reporting region error due to nll
   |         ^^^^^^^^^^^^^^^^^^^^^^^

note: External requirements
  --> $DIR/propagate-approximated-shorter-to-static-no-bound.rs:45:47
   |
LL |       establish_relationships(&cell_a, &cell_b, |_outlives, x, y| {
   |  _______________________________________________^
LL | |         //~^ ERROR does not outlive free region
LL | |
LL | |         // Only works if 'x: 'y:
LL | |         demand_y(x, y, x.get()) //~ WARNING not reporting region error due to nll
LL | |     });
   | |_____^
   |
   = note: defining type: DefId(0/1:18 ~ propagate_approximated_shorter_to_static_no_bound[317d]::supply[0]::{{closure}}[0]) with closure substs [
               i16,
               for<'r, 's, 't0, 't1, 't2> extern "rust-call" fn((&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 'r)) std::cell::Cell<&'_#1r &ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 's)) u32>, &ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 't0)) std::cell::Cell<&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 's)) u32>, &ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 't1)) std::cell::Cell<&ReLateBound(DebruijnIndex(0), BrNamed(crate0:DefIndex(0:0), 't2)) u32>))
           ]
   = note: number of external vids: 2
   = note: where '_#1r: '_#0r

error: free region `ReFree(DefId(0/0:6 ~ propagate_approximated_shorter_to_static_no_bound[317d]::supply[0]), BrNamed(crate0:DefIndex(1:16), 'a))` does not outlive free region `ReStatic`
  --> $DIR/propagate-approximated-shorter-to-static-no-bound.rs:45:5
   |
LL | /     establish_relationships(&cell_a, &cell_b, |_outlives, x, y| {
LL | |         //~^ ERROR does not outlive free region
LL | |
LL | |         // Only works if 'x: 'y:
LL | |         demand_y(x, y, x.get()) //~ WARNING not reporting region error due to nll
LL | |     });
   | |______^

note: No external requirements
  --> $DIR/propagate-approximated-shorter-to-static-no-bound.rs:44:1
   |
LL | / fn supply<'a, 'b>(cell_a: Cell<&'a u32>, cell_b: Cell<&'b u32>) {
LL | |     establish_relationships(&cell_a, &cell_b, |_outlives, x, y| {
LL | |         //~^ ERROR does not outlive free region
LL | |
...  |
LL | |     });
LL | | }
   | |_^
   |
   = note: defining type: DefId(0/0:6 ~ propagate_approximated_shorter_to_static_no_bound[317d]::supply[0]) with substs []

error: aborting due to previous error

