error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:6:22
   |
LL | fn option(i: i32) -> impl Sized { //~ ERROR
   |                      ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `std::option::Option<(impl Sized, i32)>`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:14:15
   |
LL | fn tuple() -> impl Sized { //~ ERROR
   |               ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `(impl Sized,)`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:18:15
   |
LL | fn array() -> impl Sized { //~ ERROR
   |               ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[impl Sized; 1]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:22:13
   |
LL | fn ptr() -> impl Sized { //~ ERROR
   |             ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `*const impl Sized`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:26:16
   |
LL | fn fn_ptr() -> impl Sized { //~ ERROR
   |                ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `fn() -> impl Sized`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:30:25
   |
LL | fn closure_capture() -> impl Sized { //~ ERROR
   |                         ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[closure@$DIR/recursive-impl-trait-type.rs:32:5: 32:19 x:impl Sized]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:35:29
   |
LL | fn closure_ref_capture() -> impl Sized { //~ ERROR
   |                             ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[closure@$DIR/recursive-impl-trait-type.rs:37:5: 37:20 x:impl Sized]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:40:21
   |
LL | fn closure_sig() -> impl Sized { //~ ERROR
   |                     ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[closure@$DIR/recursive-impl-trait-type.rs:41:5: 41:21]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:44:23
   |
LL | fn generator_sig() -> impl Sized { //~ ERROR
   |                       ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[closure@$DIR/recursive-impl-trait-type.rs:45:5: 45:23]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:48:27
   |
LL | fn generator_capture() -> impl Sized { //~ ERROR
   |                           ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[generator@$DIR/recursive-impl-trait-type.rs:50:5: 50:26 x:impl Sized {()}]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:53:26
   |
LL | fn substs_change<T>() -> impl Sized { //~ ERROR
   |                          ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `(impl Sized,)`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:57:24
   |
LL | fn generator_hold() -> impl Sized { //~ ERROR
   |                        ^^^^^^^^^^ expands to self-referential type
   |
   = note: expanded type is `[generator@$DIR/recursive-impl-trait-type.rs:58:5: 62:6 {impl Sized, ()}]`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:65:40
   |
LL | async fn recursive_async_function() -> () { //~ ERROR
   |                                        ^^ expands to self-referential type
   |
   = note: expanded type is `std::future::GenFuture<[static generator@$DIR/recursive-impl-trait-type.rs:65:43: 67:2 {impl std::future::Future, ()}]>`

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:73:26
   |
LL | fn mutual_recursion() -> impl Sync { //~ ERROR
   |                          ^^^^^^^^^ expands to self-referential type
   |
   = note: type resolves to itself

error[E0720]: opaque type expands to a recursive type
  --> $DIR/recursive-impl-trait-type.rs:77:28
   |
LL | fn mutual_recursion_b() -> impl Sized { //~ ERROR
   |                            ^^^^^^^^^^ expands to self-referential type
   |
   = note: type resolves to itself

error: aborting due to 15 previous errors

For more information about this error, try `rustc --explain E0720`.
