error[E0405]: cannot find trait `Debug` in this scope
  --> $DIR/universal_wrong_bounds.rs:9:24
   |
LL | fn wants_debug(g: impl Debug) { } //~ ERROR cannot find
   |                        ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
LL | use std::fmt::Debug;
   |

error[E0405]: cannot find trait `Debug` in this scope
  --> $DIR/universal_wrong_bounds.rs:10:26
   |
LL | fn wants_display(g: impl Debug) { } //~ ERROR cannot find
   |                          ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
LL | use std::fmt::Debug;
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0405`.
