error[E0382]: use of moved value: `a`
  --> $DIR/dbg-macro-move-semantics.rs:9:18
   |
LL |     let _ = dbg!(a);
   |             ------- value moved here
LL |     let _ = dbg!(a); //~ ERROR use of moved value
   |                  ^ value used here after move
   |
   = note: move occurs because `a` has type `NoCopy`, which does not implement the `Copy` trait
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0382]: use of moved value: `a`
  --> $DIR/dbg-macro-move-semantics.rs:9:13
   |
LL |     let _ = dbg!(a);
   |             ------- value moved here
LL |     let _ = dbg!(a); //~ ERROR use of moved value
   |             ^^^^^^^ value used here after move
   |
   = note: move occurs because `a` has type `NoCopy`, which does not implement the `Copy` trait
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0382`.
