error[E0308]: mismatched types
  --> $DIR/nested-item-spans.rs:10:22
   |
LL |         let x: u32 = "x"; //~ ERROR: mismatched types
   |                      ^^^ expected u32, found reference
   |
   = note: expected type `u32`
              found type `&'static str`

error[E0308]: mismatched types
  --> $DIR/nested-item-spans.rs:19:22
   |
LL |         let x: u32 = "x"; //~ ERROR: mismatched types
   |                      ^^^ expected u32, found reference
   |
   = note: expected type `u32`
              found type `&'static str`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
