error[E0308]: mismatched types
   |
   = note: expected type `()`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/span-preservation.rs:13:20
   |
LL |     let x: usize = "hello";;;;; //~ ERROR mismatched types
   |                    ^^^^^^^ expected usize, found reference
   |
   = note: expected type `usize`
              found type `&'static str`

error[E0308]: mismatched types
  --> $DIR/span-preservation.rs:19:29
   |
LL |         Some(x) => { return x }, //~ ERROR mismatched types
   |                             ^ expected usize, found isize

error[E0308]: mismatched types
  --> $DIR/span-preservation.rs:35:22
   |
LL |     let x = Foo { a: 10isize }; //~ ERROR mismatched types
   |                      ^^^^^^^ expected usize, found isize

error[E0560]: struct `c::Foo` has no field named `b`
  --> $DIR/span-preservation.rs:36:26
   |
LL |     let y = Foo { a: 10, b: 10isize }; //~ ERROR has no field named `b`
   |                          ^ `c::Foo` does not have this field
   |
   = note: available fields are: `a`

error[E0308]: mismatched types
  --> $DIR/span-preservation.rs:49:5
   |
LL | extern "C" fn baz() {
   |                     - possibly return type missing here?
LL |     0 //~ ERROR mismatched types
   |     ^ expected (), found integer
   |
   = note: expected type `()`
              found type `{integer}`

error: aborting due to 6 previous errors

Some errors occurred: E0308, E0560.
For more information about an error, try `rustc --explain E0308`.
