error[E0004]: non-exhaustive patterns: `(B, _)`, `(C, _)`, `(D, _)` and 2 more not covered
  --> $DIR/issue-35609.rs:10:11
   |
LL |     match (A, ()) { //~ ERROR non-exhaustive
   |           ^^^^^^^ patterns `(B, _)`, `(C, _)`, `(D, _)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `(_, B)`, `(_, C)`, `(_, D)` and 2 more not covered
  --> $DIR/issue-35609.rs:14:11
   |
LL |     match (A, A) { //~ ERROR non-exhaustive
   |           ^^^^^^ patterns `(_, B)`, `(_, C)`, `(_, D)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered
  --> $DIR/issue-35609.rs:18:11
   |
LL |     match ((A, ()), ()) { //~ ERROR non-exhaustive
   |           ^^^^^^^^^^^^^ patterns `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered
  --> $DIR/issue-35609.rs:22:11
   |
LL |     match ((A, ()), A) { //~ ERROR non-exhaustive
   |           ^^^^^^^^^^^^ patterns `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered
  --> $DIR/issue-35609.rs:26:11
   |
LL |     match ((A, ()), ()) { //~ ERROR non-exhaustive
   |           ^^^^^^^^^^^^^ patterns `((B, _), _)`, `((C, _), _)`, `((D, _), _)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `S(B, _)`, `S(C, _)`, `S(D, _)` and 2 more not covered
  --> $DIR/issue-35609.rs:31:11
   |
LL |     match S(A, ()) { //~ ERROR non-exhaustive
   |           ^^^^^^^^ patterns `S(B, _)`, `S(C, _)`, `S(D, _)` and 2 more not covered

error[E0004]: non-exhaustive patterns: `Sd { x: B, .. }`, `Sd { x: C, .. }`, `Sd { x: D, .. }` and 2 more not covered
  --> $DIR/issue-35609.rs:35:11
   |
LL |     match (Sd { x: A, y: () }) { //~ ERROR non-exhaustive
   |           ^^^^^^^^^^^^^^^^^^^^ patterns `Sd { x: B, .. }`, `Sd { x: C, .. }`, `Sd { x: D, .. }` and 2 more not covered

error[E0004]: non-exhaustive patterns: `Some(B)`, `Some(C)`, `Some(D)` and 2 more not covered
  --> $DIR/issue-35609.rs:39:11
   |
LL |     match Some(A) { //~ ERROR non-exhaustive
   |           ^^^^^^^ patterns `Some(B)`, `Some(C)`, `Some(D)` and 2 more not covered

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0004`.
