error: macro expansion ignores token `,` and any following
  --> $DIR/macro-incomplete-parse.rs:7:9
   |
LL |         , //~ ERROR macro expansion ignores token `,`
   |         ^
...
LL | ignored_item!();
   | ---------------- caused by the macro expansion here
   |
   = note: the usage of `ignored_item!` is likely invalid in item context

error: expected one of `.`, `;`, `?`, `}`, or an operator, found `,`
  --> $DIR/macro-incomplete-parse.rs:12:14
   |
LL |     () => ( 1,  //~ ERROR expected one of `.`, `;`, `?`, `}`, or an operator, found `,`
   |              ^ expected one of `.`, `;`, `?`, `}`, or an operator here
...
LL |     ignored_expr!();
   |     ---------------- in this macro invocation

error: macro expansion ignores token `,` and any following
  --> $DIR/macro-incomplete-parse.rs:18:14
   |
LL |     () => ( 1, 2 ) //~ ERROR macro expansion ignores token `,`
   |              ^
...
LL |         ignored_pat!() => (),
   |         -------------- caused by the macro expansion here
   |
   = note: the usage of `ignored_pat!` is likely invalid in pattern context

error: aborting due to 3 previous errors

