error[E0119]: conflicting implementations of trait `std::ops::Deref` for type `&_`:
  --> $DIR/issue-28981.rs:5:1
   |
LL | impl<Foo> Deref for Foo { } //~ ERROR must be used
   | ^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T> std::ops::Deref for &T
             where T: ?Sized;

error[E0210]: type parameter `Foo` must be used as the type parameter for some local type (e.g., `MyStruct<Foo>`)
  --> $DIR/issue-28981.rs:5:1
   |
LL | impl<Foo> Deref for Foo { } //~ ERROR must be used
   | ^^^^^^^^^^^^^^^^^^^^^^^ type parameter `Foo` must be used as the type parameter for some local type
   |
   = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to 2 previous errors

Some errors occurred: E0119, E0210.
For more information about an error, try `rustc --explain E0119`.
