error[E0597]: `c` does not live long enough
  --> $DIR/adt-nullary-enums.rs:34:41
   |
LL |         SomeEnum::SomeVariant(Cell::new(&c)), //~ ERROR
   |                               ----------^^-
   |                               |         |
   |                               |         borrowed value does not live long enough
   |                               argument requires that `c` is borrowed for `'static`
...
LL | }
   | - `c` dropped here while still borrowed

error[E0597]: `c` does not live long enough
  --> $DIR/adt-nullary-enums.rs:42:41
   |
LL | fn annot_reference_named_lifetime<'a>(_d: &'a u32) {
   |                                   -- lifetime `'a` defined here
...
LL |         SomeEnum::SomeVariant(Cell::new(&c)), //~ ERROR
   |                               ----------^^-
   |                               |         |
   |                               |         borrowed value does not live long enough
   |                               argument requires that `c` is borrowed for `'a`
...
LL | }
   | - `c` dropped here while still borrowed

error[E0597]: `c` does not live long enough
  --> $DIR/adt-nullary-enums.rs:55:45
   |
LL | fn annot_reference_named_lifetime_in_closure<'a>(_: &'a u32) {
   |                                              -- lifetime `'a` defined here
...
LL |             SomeEnum::SomeVariant(Cell::new(&c)), //~ ERROR
   |                                   ----------^^-
   |                                   |         |
   |                                   |         borrowed value does not live long enough
   |                                   argument requires that `c` is borrowed for `'a`
...
LL |     };
   |     - `c` dropped here while still borrowed

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0597`.
