error: user substs: UserSubsts { substs: [u32], user_self_ty: None }
  --> $DIR/dump-fn-method.rs:26:13
   |
LL |     let x = foo::<u32>; //~ ERROR [u32]
   |             ^^^^^^^^^^

error: user substs: UserSubsts { substs: [^0, u32, ^1], user_self_ty: None }
  --> $DIR/dump-fn-method.rs:32:13
   |
LL |     let x = <_ as Bazoom<u32>>::method::<_>; //~ ERROR [^0, u32, ^1]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: user substs: UserSubsts { substs: [u8, u16, u32], user_self_ty: None }
  --> $DIR/dump-fn-method.rs:36:13
   |
LL |     let x = <u8 as Bazoom<u16>>::method::<u32>; //~ ERROR [u8, u16, u32]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: user substs: UserSubsts { substs: [^0, ^1, u32], user_self_ty: None }
  --> $DIR/dump-fn-method.rs:44:5
   |
LL |     y.method::<u32>(44, 66); //~ ERROR [^0, ^1, u32]
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

