error[E0609]: no field `d` on type `&A`
  --> $DIR/struct-pat-derived-error.rs:8:31
   |
LL |         let A { x, y } = self.d; //~ ERROR no field `d` on type `&A`
   |                               ^

error[E0026]: struct `A` does not have fields named `x`, `y`
  --> $DIR/struct-pat-derived-error.rs:8:17
   |
LL |         let A { x, y } = self.d; //~ ERROR no field `d` on type `&A`
   |                 ^  ^ struct `A` does not have these fields

error[E0027]: pattern does not mention fields `b`, `c`
  --> $DIR/struct-pat-derived-error.rs:8:13
   |
LL |         let A { x, y } = self.d; //~ ERROR no field `d` on type `&A`
   |             ^^^^^^^^^^ missing fields `b`, `c`

error: aborting due to 3 previous errors

Some errors occurred: E0026, E0027, E0609.
For more information about an error, try `rustc --explain E0026`.
