error[E0616]: field `a` of struct `inner::A` is private
  --> $DIR/struct-field-privacy.rs:23:5
   |
LL |     b.a; //~ ERROR: field `a` of struct `inner::A` is private
   |     ^^^

error[E0616]: field `b` of struct `inner::B` is private
  --> $DIR/struct-field-privacy.rs:26:5
   |
LL |     c.b; //~ ERROR: field `b` of struct `inner::B` is private
   |     ^^^

error[E0616]: field `a` of struct `xc::A` is private
  --> $DIR/struct-field-privacy.rs:28:5
   |
LL |     d.a; //~ ERROR: field `a` of struct `xc::A` is private
   |     ^^^

error[E0616]: field `b` of struct `xc::B` is private
  --> $DIR/struct-field-privacy.rs:32:5
   |
LL |     e.b; //~ ERROR: field `b` of struct `xc::B` is private
   |     ^^^

error[E0616]: field `1` of struct `inner::Z` is private
  --> $DIR/struct-field-privacy.rs:35:5
   |
LL |     z.1; //~ ERROR: field `1` of struct `inner::Z` is private
   |     ^^^

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0616`.
