error[E0603]: enum `Bar` is private
  --> $DIR/struct-variant-privacy-xc.rs:4:33
   |
LL | fn f(b: struct_variant_privacy::Bar) { //~ ERROR enum `Bar` is private
   |                                 ^^^

error[E0603]: enum `Bar` is private
  --> $DIR/struct-variant-privacy-xc.rs:6:33
   |
LL |         struct_variant_privacy::Bar::Baz { a: _a } => {} //~ ERROR enum `Bar` is private
   |                                 ^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0603`.
