error[E0658]: imports can only refer to extern crate names passed with `--extern` on stable channel (see issue #53130)
  --> $DIR/issue-54390.rs:8:5
   |
LL | use std::fmt;
   |     -------- not an extern crate passed with `--extern`
...
LL | use fmt::Write; //~ ERROR imports can only refer to extern crate names
   |     ^^^
   |
   = help: add #![feature(uniform_paths)] to the crate attributes to enable
note: this import refers to the module imported here
  --> $DIR/issue-54390.rs:5:5
   |
LL | use std::fmt;
   |     ^^^^^^^^

error: unused import: `fmt::Write`
  --> $DIR/issue-54390.rs:8:5
   |
LL | use fmt::Write; //~ ERROR imports can only refer to extern crate names
   |     ^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/issue-54390.rs:3:9
   |
LL | #![deny(unused)]
   |         ^^^^^^
   = note: #[deny(unused_imports)] implied by #[deny(unused)]

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0658`.
