error: cannot import a built-in macro
  --> $DIR/prelude-fail.rs:6:5
   |
LL | use env as env_imported; //~ ERROR cannot import a built-in macro
   |     ^^^^^^^^^^^^^^^^^^^

error[E0432]: unresolved import `rustfmt`
  --> $DIR/prelude-fail.rs:9:5
   |
LL | use rustfmt::skip as imported_rustfmt_skip; //~ ERROR unresolved import `rustfmt`
   |     ^^^^^^^ not a module `rustfmt`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
