error[E0178]: expected a path on the left-hand side of `+`, not `&'a Foo`
  --> $DIR/E0178.rs:4:8
   |
LL |     w: &'a Foo + Copy, //~ ERROR expected a path
   |        ^^^^^^^^^^^^^^ help: try adding parentheses: `&'a (Foo + Copy)`

error[E0178]: expected a path on the left-hand side of `+`, not `&'a Foo`
  --> $DIR/E0178.rs:5:8
   |
LL |     x: &'a Foo + 'a, //~ ERROR expected a path
   |        ^^^^^^^^^^^^ help: try adding parentheses: `&'a (Foo + 'a)`

error[E0178]: expected a path on the left-hand side of `+`, not `&'a mut Foo`
  --> $DIR/E0178.rs:6:8
   |
LL |     y: &'a mut Foo + 'a, //~ ERROR expected a path
   |        ^^^^^^^^^^^^^^^^ help: try adding parentheses: `&'a mut (Foo + 'a)`

error[E0178]: expected a path on the left-hand side of `+`, not `fn() -> Foo`
  --> $DIR/E0178.rs:7:8
   |
LL |     z: fn() -> Foo + 'a, //~ ERROR expected a path
   |        ^^^^^^^^^^^^^^^^ perhaps you forgot parentheses?

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0178`.
