error[E0596]: cannot borrow `f.v` as mutable, as `f` is not declared as mutable
  --> $DIR/issue-35937.rs:7:5
   |
LL |     let f = Foo { v: Vec::new() };
   |         - help: consider changing this to be mutable: `mut f`
LL |     f.v.push("cat".to_string()); //~ ERROR cannot borrow
   |     ^^^ cannot borrow as mutable

error[E0594]: cannot assign to `s.x`, as `s` is not declared as mutable
  --> $DIR/issue-35937.rs:16:5
   |
LL |     let s = S { x: 42 };
   |         - help: consider changing this to be mutable: `mut s`
LL |     s.x += 1; //~ ERROR cannot assign
   |     ^^^^^^^^ cannot assign

error[E0594]: cannot assign to `s.x`, as `s` is not declared as mutable
  --> $DIR/issue-35937.rs:20:5
   |
LL | fn bar(s: S) {
   |        - help: consider changing this to be mutable: `mut s`
LL |     s.x += 1; //~ ERROR cannot assign
   |     ^^^^^^^^ cannot assign

error: aborting due to 3 previous errors

Some errors occurred: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
