error[E0596]: cannot borrow immutable item `z.x` as mutable
  --> $DIR/issue-39544.rs:21:13
   |
LL |     let z = Z { x: X::Y };
   |         - help: consider changing this to be mutable: `mut z`
LL |     let _ = &mut z.x; //~ ERROR cannot borrow
   |             ^^^^^^^^ cannot borrow as mutable

error[E0596]: cannot borrow immutable item `self.x` as mutable
  --> $DIR/issue-39544.rs:26:17
   |
LL |     fn foo<'z>(&'z self) {
   |                -------- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `self.x` as mutable
  --> $DIR/issue-39544.rs:30:17
   |
LL |     fn foo1(&self, other: &Z) {
   |             ----- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `other.x` as mutable
  --> $DIR/issue-39544.rs:31:17
   |
LL |     fn foo1(&self, other: &Z) {
   |                           -- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
LL |         let _ = &mut other.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^ `other` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `self.x` as mutable
  --> $DIR/issue-39544.rs:35:17
   |
LL |     fn foo2<'a>(&'a self, other: &Z) {
   |                 -------- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `other.x` as mutable
  --> $DIR/issue-39544.rs:36:17
   |
LL |     fn foo2<'a>(&'a self, other: &Z) {
   |                                  -- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
LL |         let _ = &mut other.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^ `other` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `self.x` as mutable
  --> $DIR/issue-39544.rs:40:17
   |
LL |     fn foo3<'a>(self: &'a Self, other: &Z) {
   |                       -------- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `other.x` as mutable
  --> $DIR/issue-39544.rs:41:17
   |
LL |     fn foo3<'a>(self: &'a Self, other: &Z) {
   |                                        -- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut self.x; //~ ERROR cannot borrow
LL |         let _ = &mut other.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^ `other` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `other.x` as mutable
  --> $DIR/issue-39544.rs:45:17
   |
LL |     fn foo4(other: &Z) {
   |                    -- help: consider changing this to be a mutable reference: `&mut Z`
LL |         let _ = &mut other.x; //~ ERROR cannot borrow
   |                 ^^^^^^^^^^^^ `other` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0596]: cannot borrow immutable item `z.x` as mutable
  --> $DIR/issue-39544.rs:51:13
   |
LL | pub fn with_arg(z: Z, w: &Z) {
   |                 - help: consider changing this to be mutable: `mut z`
LL |     let _ = &mut z.x; //~ ERROR cannot borrow
   |             ^^^^^^^^ cannot borrow as mutable

error[E0596]: cannot borrow immutable item `w.x` as mutable
  --> $DIR/issue-39544.rs:52:13
   |
LL | pub fn with_arg(z: Z, w: &Z) {
   |                          -- help: consider changing this to be a mutable reference: `&mut Z`
LL |     let _ = &mut z.x; //~ ERROR cannot borrow
LL |     let _ = &mut w.x; //~ ERROR cannot borrow
   |             ^^^^^^^^ `w` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error[E0594]: cannot assign to `*x.0` which is behind a `&` reference
  --> $DIR/issue-39544.rs:58:5
   |
LL |     *x.0 = 1;
   |     ^^^^^^^^ cannot assign

error: aborting due to 12 previous errors

Some errors occurred: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
