error[E0423]: expected function, found enum `Option`
  --> $DIR/issue-43871-enum-instead-of-variant.rs:4:13
   |
LL |     let x = Option(1); //~ ERROR expected function, found enum
   |             ^^^^^^
   |
   = note: did you mean to use one of the following variants?
           - `std::prelude::v1::Option::None`
           - `std::prelude::v1::Option::Some`

error[E0532]: expected tuple struct/variant, found enum `Option`
  --> $DIR/issue-43871-enum-instead-of-variant.rs:6:12
   |
LL |     if let Option(_) = x { //~ ERROR expected tuple struct/variant, found enum
   |            ^^^^^^
   |
   = note: did you mean to use one of the following variants?
           - `std::prelude::v1::Option::None`
           - `std::prelude::v1::Option::Some`

error[E0532]: expected tuple struct/variant, found enum `Example`
  --> $DIR/issue-43871-enum-instead-of-variant.rs:12:12
   |
LL |     if let Example(_) = y { //~ ERROR expected tuple struct/variant, found enum
   |            ^^^^^^^
   |
   = note: did you mean to use one of the following variants?
           - `Example::Ex`
           - `Example::NotEx`

error: aborting due to 3 previous errors

Some errors occurred: E0423, E0532.
For more information about an error, try `rustc --explain E0423`.
