error[E0596]: cannot borrow field `f.v` of immutable binding as mutable
  --> $DIR/issue-35937.rs:7:5
   |
LL |     let f = Foo { v: Vec::new() };
   |         - help: make this binding mutable: `mut f`
LL |     f.v.push("cat".to_string()); //~ ERROR cannot borrow
   |     ^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `s.x` of immutable binding
  --> $DIR/issue-35937.rs:16:5
   |
LL |     let s = S { x: 42 };
   |         - help: make this binding mutable: `mut s`
LL |     s.x += 1; //~ ERROR cannot assign
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error[E0594]: cannot assign to field `s.x` of immutable binding
  --> $DIR/issue-35937.rs:20:5
   |
LL | fn bar(s: S) {
   |        - help: make this binding mutable: `mut s`
LL |     s.x += 1; //~ ERROR cannot assign
   |     ^^^^^^^^ cannot mutably borrow field of immutable binding

error: aborting due to 3 previous errors

Some errors occurred: E0594, E0596.
For more information about an error, try `rustc --explain E0594`.
