error[E0178]: expected a path on the left-hand side of `+`, not `&Copy`
  --> $DIR/trait-object-reference-without-parens-suggestion.rs:2:12
   |
LL |     let _: &Copy + 'static; //~ ERROR expected a path
   |            ^^^^^^^^^^^^^^^ help: try adding parentheses: `&(Copy + 'static)`

error[E0178]: expected a path on the left-hand side of `+`, not `&'static Copy`
  --> $DIR/trait-object-reference-without-parens-suggestion.rs:4:12
   |
LL |     let _: &'static Copy + 'static; //~ ERROR expected a path
   |            ^^^^^^^^^^^^^^^^^^^^^^^ help: try adding parentheses: `&'static (Copy + 'static)`

error[E0038]: the trait `std::marker::Copy` cannot be made into an object
  --> $DIR/trait-object-reference-without-parens-suggestion.rs:2:12
   |
LL |     let _: &Copy + 'static; //~ ERROR expected a path
   |            ^^^^^ the trait `std::marker::Copy` cannot be made into an object
   |
   = note: the trait cannot require that `Self : Sized`

error: aborting due to 3 previous errors

Some errors occurred: E0038, E0178.
For more information about an error, try `rustc --explain E0038`.
