error[E0004]: non-exhaustive patterns: `A` not covered
  --> $DIR/non-exhaustive-match.rs:8:11
   |
LL |     match x { T::B => { } } //~ ERROR non-exhaustive patterns: `A` not covered
   |           ^ pattern `A` not covered

error[E0004]: non-exhaustive patterns: `false` not covered
  --> $DIR/non-exhaustive-match.rs:9:11
   |
LL |     match true { //~ ERROR non-exhaustive patterns: `false` not covered
   |           ^^^^ pattern `false` not covered

error[E0004]: non-exhaustive patterns: `Some(_)` not covered
  --> $DIR/non-exhaustive-match.rs:12:11
   |
LL |     match Some(10) { //~ ERROR non-exhaustive patterns: `Some(_)` not covered
   |           ^^^^^^^^ pattern `Some(_)` not covered

error[E0004]: non-exhaustive patterns: `(_, _, -2147483648i32..=3i32)` and `(_, _, 5i32..=2147483647i32)` not covered
  --> $DIR/non-exhaustive-match.rs:15:11
   |
LL |     match (2, 3, 4) { //~ ERROR non-exhaustive patterns: `(_, _, -2147483648i32..=3i32)`
   |           ^^^^^^^^^ patterns `(_, _, -2147483648i32..=3i32)` and `(_, _, 5i32..=2147483647i32)` not covered

error[E0004]: non-exhaustive patterns: `(A, A)` not covered
  --> $DIR/non-exhaustive-match.rs:19:11
   |
LL |     match (T::A, T::A) { //~ ERROR non-exhaustive patterns: `(A, A)` not covered
   |           ^^^^^^^^^^^^ pattern `(A, A)` not covered

error[E0004]: non-exhaustive patterns: `B` not covered
  --> $DIR/non-exhaustive-match.rs:23:11
   |
LL |     match T::A { //~ ERROR non-exhaustive patterns: `B` not covered
   |           ^^^^ pattern `B` not covered

error[E0004]: non-exhaustive patterns: `[]` not covered
  --> $DIR/non-exhaustive-match.rs:34:11
   |
LL |     match *vec { //~ ERROR non-exhaustive patterns: `[]` not covered
   |           ^^^^ pattern `[]` not covered

error[E0004]: non-exhaustive patterns: `[_, _, _, _]` not covered
  --> $DIR/non-exhaustive-match.rs:47:11
   |
LL |     match *vec { //~ ERROR non-exhaustive patterns: `[_, _, _, _]` not covered
   |           ^^^^ pattern `[_, _, _, _]` not covered

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0004`.
