error[E0599]: no method named `closure` found for type `Obj<[closure@$DIR/issue-2392.rs:39:36: 39:41]>` in the current scope
  --> $DIR/issue-2392.rs:40:15
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `closure` not found for this
...
LL |     o_closure.closure(); //~ ERROR no method named `closure` found
   |               ^^^^^^^ field, not a method
   |
   = help: use `(o_closure.closure)(...)` if you meant to call the function stored in the `closure` field

error[E0599]: no method named `not_closure` found for type `Obj<[closure@$DIR/issue-2392.rs:39:36: 39:41]>` in the current scope
  --> $DIR/issue-2392.rs:42:15
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `not_closure` not found for this
...
LL |     o_closure.not_closure();
   |               ^^^^^^^^^^^ field, not a method
   |
   = help: did you mean to write `o_closure.not_closure` instead of `o_closure.not_closure(...)`?

error[E0599]: no method named `closure` found for type `Obj<fn() -> u32 {func}>` in the current scope
  --> $DIR/issue-2392.rs:46:12
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `closure` not found for this
...
LL |     o_func.closure(); //~ ERROR no method named `closure` found
   |            ^^^^^^^ field, not a method
   |
   = help: use `(o_func.closure)(...)` if you meant to call the function stored in the `closure` field

error[E0599]: no method named `boxed_closure` found for type `BoxedObj` in the current scope
  --> $DIR/issue-2392.rs:49:14
   |
LL | struct BoxedObj {
   | --------------- method `boxed_closure` not found for this
...
LL |     boxed_fn.boxed_closure();//~ ERROR no method named `boxed_closure` found
   |              ^^^^^^^^^^^^^ field, not a method
   |
   = help: use `(boxed_fn.boxed_closure)(...)` if you meant to call the function stored in the `boxed_closure` field

error[E0599]: no method named `boxed_closure` found for type `BoxedObj` in the current scope
  --> $DIR/issue-2392.rs:52:19
   |
LL | struct BoxedObj {
   | --------------- method `boxed_closure` not found for this
...
LL |     boxed_closure.boxed_closure();//~ ERROR no method named `boxed_closure` found
   |                   ^^^^^^^^^^^^^ field, not a method
   |
   = help: use `(boxed_closure.boxed_closure)(...)` if you meant to call the function stored in the `boxed_closure` field

error[E0599]: no method named `closure` found for type `Obj<fn() -> u32 {func}>` in the current scope
  --> $DIR/issue-2392.rs:57:12
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `closure` not found for this
...
LL |     w.wrap.closure();//~ ERROR no method named `closure` found
   |            ^^^^^^^ field, not a method
   |
   = help: use `(w.wrap.closure)(...)` if you meant to call the function stored in the `closure` field

error[E0599]: no method named `not_closure` found for type `Obj<fn() -> u32 {func}>` in the current scope
  --> $DIR/issue-2392.rs:59:12
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `not_closure` not found for this
...
LL |     w.wrap.not_closure();
   |            ^^^^^^^^^^^ field, not a method
   |
   = help: did you mean to write `w.wrap.not_closure` instead of `w.wrap.not_closure(...)`?

error[E0599]: no method named `closure` found for type `Obj<std::boxed::Box<(dyn std::boxed::FnBox<(), Output=u32> + 'static)>>` in the current scope
  --> $DIR/issue-2392.rs:62:24
   |
LL | struct Obj<F> where F: FnOnce() -> u32 {
   | -------------------------------------- method `closure` not found for this
...
LL |     check_expression().closure();//~ ERROR no method named `closure` found
   |                        ^^^^^^^ field, not a method
   |
   = help: use `(check_expression().closure)(...)` if you meant to call the function stored in the `closure` field

error[E0599]: no method named `f1` found for type `FuncContainer` in the current scope
  --> $DIR/issue-2392.rs:68:31
   |
LL | struct FuncContainer {
   | -------------------- method `f1` not found for this
...
LL |             (*self.container).f1(1); //~ ERROR no method named `f1` found
   |                               ^^ field, not a method
   |
   = help: use `((*self.container).f1)(...)` if you meant to call the function stored in the `f1` field

error[E0599]: no method named `f2` found for type `FuncContainer` in the current scope
  --> $DIR/issue-2392.rs:69:31
   |
LL | struct FuncContainer {
   | -------------------- method `f2` not found for this
...
LL |             (*self.container).f2(1); //~ ERROR no method named `f2` found
   |                               ^^ field, not a method
   |
   = help: use `((*self.container).f2)(...)` if you meant to call the function stored in the `f2` field

error[E0599]: no method named `f3` found for type `FuncContainer` in the current scope
  --> $DIR/issue-2392.rs:70:31
   |
LL | struct FuncContainer {
   | -------------------- method `f3` not found for this
...
LL |             (*self.container).f3(1); //~ ERROR no method named `f3` found
   |                               ^^ field, not a method
   |
   = help: use `((*self.container).f3)(...)` if you meant to call the function stored in the `f3` field

error: aborting due to 11 previous errors

For more information about this error, try `rustc --explain E0599`.
