error[E0596]: cannot borrow immutable local variable `y` as mutable
  --> $DIR/augmented-assignments.rs:21:5
   |
LL |     let y = Int(2);
   |         - help: make this binding mutable: `mut y`
...
LL |     y   //~ error: cannot borrow immutable local variable `y` as mutable
   |     ^ cannot borrow mutably

error[E0382]: use of moved value: `x`
  --> $DIR/augmented-assignments.rs:13:5
   |
LL |     x   //~ error: use of moved value: `x`
   |     ^ value used here after move
...
LL |     x;  //~ value moved here
   |     - value moved here
   |
   = note: move occurs because `x` has type `Int`, which does not implement the `Copy` trait

error: aborting due to 2 previous errors

Some errors occurred: E0382, E0596.
For more information about an error, try `rustc --explain E0382`.
