error[E0658]: The attribute `foo` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/custom_attribute.rs:3:3
   |
LL | #[foo] //~ ERROR The attribute `foo`
   |   ^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error[E0658]: The attribute `foo` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/custom_attribute.rs:5:7
   |
LL |     #[foo] //~ ERROR The attribute `foo`
   |       ^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error[E0658]: The attribute `foo` is currently unknown to the compiler and may have meaning added to it in the future (see issue #29642)
  --> $DIR/custom_attribute.rs:7:7
   |
LL |     #[foo] //~ ERROR The attribute `foo`
   |       ^^^
   |
   = help: add #![feature(custom_attribute)] to the crate attributes to enable

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0658`.
