error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:18:9
   |
LL |     foo(s); //~ ERROR mismatched types
   |         ^
   |         |
   |         expected struct `std::string::String`, found reference
   |         help: try using a conversion method: `s.to_string()`
   |
   = note: expected type `std::string::String`
              found type `&std::string::String`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:23:10
   |
LL |     foo3(u); //~ ERROR mismatched types
   |          ^
   |          |
   |          expected u32, found &u32
   |          help: consider dereferencing the borrow: `*u`
   |
   = note: expected type `u32`
              found type `&u32`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:30:9
   |
LL |     foo(&"aaa".to_owned()); //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^^^^^
   |         |
   |         expected struct `std::string::String`, found reference
   |         help: consider removing the borrow: `"aaa".to_owned()`
   |
   = note: expected type `std::string::String`
              found type `&std::string::String`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:31:9
   |
LL |     foo(&mut "aaa".to_owned()); //~ ERROR mismatched types
   |         ^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         expected struct `std::string::String`, found mutable reference
   |         help: consider removing the borrow: `"aaa".to_owned()`
   |
   = note: expected type `std::string::String`
              found type `&mut std::string::String`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:12:20
   |
LL |     ($x:expr) => { &$x } //~ ERROR mismatched types
   |                    ^^^ expected u32, found &{integer}
...
LL |     foo3(borrow!(0));
   |          ---------- in this macro invocation
   |
   = note: expected type `u32`
              found type `&{integer}`

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0308`.
