warning: empty trait list in `derive`
  --> $DIR/deriving-meta-empty-trait-list.rs:15:1
   |
LL | #[derive]   //~ WARNING empty trait list in `derive`
   | ^^^^^^^^^

warning: empty trait list in `derive`
  --> $DIR/deriving-meta-empty-trait-list.rs:18:1
   |
LL | #[derive()] //~ WARNING empty trait list in `derive`
   | ^^^^^^^^^^^

