error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:27:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 127
   |
   = note: explicitly set `OhNo = -128` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:36:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 255
   |
   = note: explicitly set `OhNo = 0` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:45:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 32767
   |
   = note: explicitly set `OhNo = -32768` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:54:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 65535
   |
   = note: explicitly set `OhNo = 0` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:63:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 2147483647
   |
   = note: explicitly set `OhNo = -2147483648` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:72:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 4294967295
   |
   = note: explicitly set `OhNo = 0` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:81:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 9223372036854775807
   |
   = note: explicitly set `OhNo = -9223372036854775808` if that is desired outcome

error[E0370]: enum discriminant overflowed
  --> $DIR/discrim-overflow-2.rs:90:9
   |
LL |         OhNo, //~ ERROR enum discriminant overflowed [E0370]
   |         ^^^^ overflowed on value after 18446744073709551615
   |
   = note: explicitly set `OhNo = 0` if that is desired outcome

error: aborting due to 8 previous errors

For more information about this error, try `rustc --explain E0370`.
