error[E0308]: mismatched types
  --> $DIR/diverging-tuple-parts-39485.rs:8:5
   |
LL | fn g() {
   |        - help: try adding a return type: `-> &_`
LL |     &panic!() //~ ERROR mismatched types
   |     ^^^^^^^^^ expected (), found reference
   |
   = note: expected type `()`
              found type `&_`

error[E0308]: mismatched types
  --> $DIR/diverging-tuple-parts-39485.rs:12:5
   |
LL | fn f() -> isize {
   |           ----- expected `isize` because of return type
LL |     (return 1, return 2) //~ ERROR mismatched types
   |     ^^^^^^^^^^^^^^^^^^^^ expected isize, found tuple
   |
   = note: expected type `isize`
              found type `(!, !)`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
