error[E0642]: patterns aren't allowed in methods without bodies
  --> $DIR/E0642.rs:5:12
   |
LL |     fn foo((x, y): (i32, i32)); //~ ERROR patterns aren't allowed in methods without bodies
   |            ^^^^^^
help: give this argument a name or use an underscore to ignore it
   |
LL |     fn foo(_: (i32, i32)); //~ ERROR patterns aren't allowed in methods without bodies
   |            ^

error[E0642]: patterns aren't allowed in methods without bodies
  --> $DIR/E0642.rs:7:12
   |
LL |     fn bar((x, y): (i32, i32)) {} //~ ERROR patterns aren't allowed in methods without bodies
   |            ^^^^^^
help: give this argument a name or use an underscore to ignore it
   |
LL |     fn bar(_: (i32, i32)) {} //~ ERROR patterns aren't allowed in methods without bodies
   |            ^

error[E0642]: patterns aren't allowed in methods without bodies
  --> $DIR/E0642.rs:9:15
   |
LL |     fn method(S { .. }: S) {} //~ ERROR patterns aren't allowed in methods without bodies
   |               ^^^^^^^^
help: give this argument a name or use an underscore to ignore it
   |
LL |     fn method(_: S) {} //~ ERROR patterns aren't allowed in methods without bodies
   |               ^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0642`.
