error[E0618]: expected function, found `Empty2`
  --> $DIR/empty-struct-unit-expr.rs:25:14
   |
LL | struct Empty2;
   | -------------- `Empty2` defined here
...
LL |     let e2 = Empty2(); //~ ERROR expected function, found `Empty2`
   |              ^^^^^^^^ not a function

error[E0618]: expected function, found enum variant `E::Empty4`
  --> $DIR/empty-struct-unit-expr.rs:26:14
   |
LL |     Empty4
   |     ------ `E::Empty4` defined here
...
LL |     let e4 = E::Empty4();
   |              ^^^^^^^^^^^ not a function
help: `E::Empty4` is a unit variant, you need to write it without the parenthesis
   |
LL |     let e4 = E::Empty4;
   |              ^^^^^^^^^

error[E0618]: expected function, found `empty_struct::XEmpty2`
  --> $DIR/empty-struct-unit-expr.rs:28:15
   |
LL |     let xe2 = XEmpty2(); //~ ERROR expected function, found `empty_struct::XEmpty2`
   |               ^^^^^^^^^ not a function

error[E0618]: expected function, found enum variant `XE::XEmpty4`
  --> $DIR/empty-struct-unit-expr.rs:29:15
   |
LL |     let xe4 = XE::XEmpty4();
   |               ^^^^^^^^^^^^^ not a function
help: `XE::XEmpty4` is a unit variant, you need to write it without the parenthesis
   |
LL |     let xe4 = XE::XEmpty4;
   |               ^^^^^^^^^^^

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0618`.
