error[E0309]: the parameter type `T` may not live long enough
  --> $DIR/feature-gate-infer_outlives_requirements.rs:15:5
   |
LL | struct Foo<'a, T> {
   |                - help: consider adding an explicit lifetime bound `T: 'a`...
LL |     bar: &'a [T] //~ ERROR the parameter type `T` may not live long enough [E0309]
   |     ^^^^^^^^^^^^
   |
note: ...so that the reference type `&'a [T]` does not outlive the data it points at
  --> $DIR/feature-gate-infer_outlives_requirements.rs:15:5
   |
LL |     bar: &'a [T] //~ ERROR the parameter type `T` may not live long enough [E0309]
   |     ^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0309`.
