error[E0308]: mismatched types
  --> $DIR/float-literal-inference-restrictions.rs:2:18
   |
LL |     let x: f32 = 1; //~ ERROR mismatched types
   |                  ^
   |                  |
   |                  expected f32, found integer
   |                  help: use a float literal: `1.0`
   |
   = note: expected type `f32`
              found type `{integer}`

error[E0308]: mismatched types
  --> $DIR/float-literal-inference-restrictions.rs:3:18
   |
LL |     let y: f32 = 1f64; //~ ERROR mismatched types
   |                  ^^^^ expected f32, found f64

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
