error[E0425]: cannot find function `fpriv` in this scope
  --> $DIR/glob-resolve1.rs:22:5
   |
LL |     fpriv(); //~ ERROR cannot find function `fpriv` in this scope
   |     ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::fpriv;
   |

error[E0425]: cannot find function `epriv` in this scope
  --> $DIR/glob-resolve1.rs:23:5
   |
LL |     epriv(); //~ ERROR cannot find function `epriv` in this scope
   |     ^^^^^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::epriv;
   |

error[E0423]: expected value, found enum `B`
  --> $DIR/glob-resolve1.rs:24:5
   |
LL |     B; //~ ERROR expected value, found enum `B`
   |     ^
   |
   = note: did you mean to use one of the following variants?
           - `B::B1`

error[E0425]: cannot find value `C` in this scope
  --> $DIR/glob-resolve1.rs:25:5
   |
LL |     C; //~ ERROR cannot find value `C` in this scope
   |     ^ not found in this scope
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::C;
   |

error[E0425]: cannot find function `import` in this scope
  --> $DIR/glob-resolve1.rs:26:5
   |
LL |     import(); //~ ERROR: cannot find function `import` in this scope
   |     ^^^^^^ not found in this scope

error[E0412]: cannot find type `A` in this scope
  --> $DIR/glob-resolve1.rs:28:11
   |
LL |     foo::<A>(); //~ ERROR: cannot find type `A` in this scope
   |           ^ did you mean `B`?
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::A;
   |

error[E0412]: cannot find type `C` in this scope
  --> $DIR/glob-resolve1.rs:29:11
   |
LL |     foo::<C>(); //~ ERROR: cannot find type `C` in this scope
   |           ^ did you mean `B`?
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::C;
   |

error[E0412]: cannot find type `D` in this scope
  --> $DIR/glob-resolve1.rs:30:11
   |
LL |     foo::<D>(); //~ ERROR: cannot find type `D` in this scope
   |           ^ did you mean `B`?
help: possible candidate is found in another module, you can import it into scope
   |
LL | use bar::D;
   |

error: aborting due to 8 previous errors

Some errors occurred: E0412, E0423, E0425.
For more information about an error, try `rustc --explain E0412`.
