error[E0432]: unresolved import `__test`
  --> $DIR/inaccessible-test-modules.rs:5:5
   |
LL | use __test as x; //~ ERROR unresolved import `__test`
   |     ^^^^^^^^^^^ no `__test` in the root. Did you mean to use `test`?

error[E0432]: unresolved import `__test_reexports`
  --> $DIR/inaccessible-test-modules.rs:6:5
   |
LL | use __test_reexports as y; //~ ERROR unresolved import `__test_reexports`
   |     ^^^^^^^^^^^^^^^^^^^^^ no `__test_reexports` in the root. Did you mean to use `__test_reexports`?

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0432`.
