error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:16:31
   |
LL |     println!("{}", a as usize < long_name); //~ ERROR `<` is interpreted as a start of generic
   |                    ---------- ^ --------- interpreted as generic arguments
   |                    |          |
   |                    |          not interpreted as comparison
   |                    help: try comparing the cast value: `(a as usize)`

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:17:33
   |
LL |     println!("{}{}", a as usize < long_name, long_name);
   |                      ---------- ^ -------------------- interpreted as generic arguments
   |                      |          |
   |                      |          not interpreted as comparison
   |                      help: try comparing the cast value: `(a as usize)`

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:19:31
   |
LL |     println!("{}", a as usize < 4); //~ ERROR `<` is interpreted as a start of generic
   |                    ---------- ^ - interpreted as generic arguments
   |                    |          |
   |                    |          not interpreted as comparison
   |                    help: try comparing the cast value: `(a as usize)`

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:21:31
   |
LL |     println!("{}{}", a: usize < long_name, long_name);
   |                      -------- ^ -------------------- interpreted as generic arguments
   |                      |        |
   |                      |        not interpreted as comparison
   |                      help: try comparing the cast value: `(a: usize)`

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:23:29
   |
LL |     println!("{}", a: usize < 4); //~ ERROR `<` is interpreted as a start of generic
   |                    -------- ^ - interpreted as generic arguments
   |                    |        |
   |                    |        not interpreted as comparison
   |                    help: try comparing the cast value: `(a: usize)`

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:28:20
   |
LL |                    < //~ ERROR `<` is interpreted as a start of generic
   |                    ^ not interpreted as comparison
LL |                    4);
   |                    - interpreted as generic arguments
help: try comparing the cast value
   |
LL |     println!("{}", (a
LL |                    as
LL |                    usize)
   |

error: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:37:20
   |
LL |                    < //~ ERROR `<` is interpreted as a start of generic
   |                    ^ not interpreted as comparison
LL |                    5);
   |                    - interpreted as generic arguments
help: try comparing the cast value
   |
LL |     println!("{}", (a
LL | 
LL | 
LL |                    as
LL | 
LL | 
 ...

error: `<` is interpreted as a start of generic arguments for `usize`, not a shift
  --> $DIR/issue-22644.rs:40:31
   |
LL |     println!("{}", a as usize << long_name); //~ ERROR `<` is interpreted as a start of generic
   |                    ---------- ^^ --------- interpreted as generic arguments
   |                    |          |
   |                    |          not interpreted as shift
   |                    help: try shifting the cast value: `(a as usize)`

error: expected type, found `4`
  --> $DIR/issue-22644.rs:42:28
   |
LL |     println!("{}", a: &mut 4); //~ ERROR expected type, found `4`
   |                            ^ expecting a type here because of type ascription

error: aborting due to 9 previous errors

