error: expected identifier, found keyword `true`
  --> $DIR/issue-44406.rs:18:10
   |
LL |     foo!(true); //~ ERROR expected type, found keyword
   |          ^^^^ expected identifier, found keyword

error: expected type, found keyword `true`
  --> $DIR/issue-44406.rs:18:10
   |
LL |         bar(baz: $rest)
   |                - help: try using a semicolon: `;`
...
LL |     foo!(true); //~ ERROR expected type, found keyword
   |          ^^^^ expecting a type here because of type ascription

error: aborting due to 2 previous errors

