error: missing condition for `if` statemement
  --> $DIR/issue-13483.rs:13:14
   |
LL |     } else if { //~ ERROR missing condition
   |              ^ expected if condition here

error: missing condition for `if` statemement
  --> $DIR/issue-13483.rs:20:14
   |
LL |     } else if { //~ ERROR missing condition
   |              ^ expected if condition here

error: aborting due to 2 previous errors

