error: unnecessary `unsafe` block
  --> $DIR/issue-45107-unnecessary-unsafe-in-closure.rs:17:13
   |
LL |     unsafe {
   |     ------ because it's nested under this `unsafe` block
LL |         let f = |v: &mut Vec<_>| {
LL |             unsafe { //~ ERROR unnecessary `unsafe`
   |             ^^^^^^ unnecessary `unsafe` block
   |
note: lint level defined here
  --> $DIR/issue-45107-unnecessary-unsafe-in-closure.rs:11:8
   |
LL | #[deny(unused_unsafe)]
   |        ^^^^^^^^^^^^^

error: unnecessary `unsafe` block
  --> $DIR/issue-45107-unnecessary-unsafe-in-closure.rs:19:38
   |
LL |     unsafe {
   |     ------ because it's nested under this `unsafe` block
...
LL |                 |w: &mut Vec<u32>| { unsafe { //~ ERROR unnecessary `unsafe`
   |                                      ^^^^^^ unnecessary `unsafe` block

error: unnecessary `unsafe` block
  --> $DIR/issue-45107-unnecessary-unsafe-in-closure.rs:23:34
   |
LL |     unsafe {
   |     ------ because it's nested under this `unsafe` block
...
LL |             |x: &mut Vec<u32>| { unsafe { //~ ERROR unnecessary `unsafe`
   |                                  ^^^^^^ unnecessary `unsafe` block

error: aborting due to 3 previous errors

