error: `macro_two` is already in scope
  --> $DIR/macro-shadowing.rs:22:5
   |
LL |     #[macro_use] //~ ERROR `macro_two` is already in scope
   |     ^^^^^^^^^^^^
...
LL | m1!();
   | ------ in this macro invocation
   |
   = note: macro-expanded `#[macro_use]`s may not shadow existing macros (see RFC 1560)

error: `foo` is already in scope
  --> $DIR/macro-shadowing.rs:20:5
   |
LL |     macro_rules! foo { () => {} } //~ ERROR `foo` is already in scope
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
LL | m1!();
   | ------ in this macro invocation
   |
   = note: macro-expanded `macro_rules!`s may not shadow existing macros (see RFC 1560)

error: aborting due to 2 previous errors

