error: incorrect close delimiter: `)`
  --> $DIR/parser-recovery-2.rs:8:5
   |
LL |     fn bar() {
   |              - un-closed delimiter
LL |         let x = foo(); //~ ERROR cannot find function `foo` in this scope
LL |     ) //~ ERROR incorrect close delimiter: `)`
   |     ^ incorrect close delimiter

error: unexpected token: `;`
  --> $DIR/parser-recovery-2.rs:12:15
   |
LL |     let x = y.;  //~ ERROR unexpected token
   |               ^

error[E0425]: cannot find function `foo` in this scope
  --> $DIR/parser-recovery-2.rs:7:17
   |
LL |         let x = foo(); //~ ERROR cannot find function `foo` in this scope
   |                 ^^^ not found in this scope

error[E0425]: cannot find value `y` in this scope
  --> $DIR/parser-recovery-2.rs:12:13
   |
LL |     let x = y.;  //~ ERROR unexpected token
   |             ^ not found in this scope

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0425`.
