error[E0004]: non-exhaustive patterns: `$ISIZE_MIN..=-6isize` and `21isize..=$ISIZE_MAX` not covered
  --> $DIR/precise_pointer_size_matching.rs:24:11
   |
LL |     match 0isize { //~ ERROR non-exhaustive patterns
   |           ^^^^^^ patterns `$ISIZE_MIN..=-6isize` and `21isize..=$ISIZE_MAX` not covered

error[E0004]: non-exhaustive patterns: `0usize` and `21usize..=$USIZE_MAX` not covered
  --> $DIR/precise_pointer_size_matching.rs:29:11
   |
LL |     match 0usize { //~ ERROR non-exhaustive patterns
   |           ^^^^^^ patterns `0usize` and `21usize..=$USIZE_MAX` not covered

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0004`.
