error[E0478]: lifetime bound not satisfied
  --> $DIR/static-lifetime.rs:13:20
   |
LL | impl<'a, A: Clone> Arbitrary for ::std::borrow::Cow<'a, A> {} //~ ERROR lifetime bound
   |                    ^^^^^^^^^
   |
note: lifetime parameter instantiated with the lifetime 'a as defined on the impl at 13:1
  --> $DIR/static-lifetime.rs:13:1
   |
LL | impl<'a, A: Clone> Arbitrary for ::std::borrow::Cow<'a, A> {} //~ ERROR lifetime bound
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: but lifetime parameter must outlive the static lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0478`.
