error[E0507]: cannot move out of borrowed content
  --> $DIR/std-uncopyable-atomics.rs:9:13
   |
LL |     let x = *&x; //~ ERROR: cannot move out of borrowed content
   |             ^^^
   |             |
   |             cannot move out of borrowed content
   |             help: consider using a reference instead: `&*&x`

error[E0507]: cannot move out of borrowed content
  --> $DIR/std-uncopyable-atomics.rs:11:13
   |
LL |     let x = *&x; //~ ERROR: cannot move out of borrowed content
   |             ^^^
   |             |
   |             cannot move out of borrowed content
   |             help: consider using a reference instead: `&*&x`

error[E0507]: cannot move out of borrowed content
  --> $DIR/std-uncopyable-atomics.rs:13:13
   |
LL |     let x = *&x; //~ ERROR: cannot move out of borrowed content
   |             ^^^
   |             |
   |             cannot move out of borrowed content
   |             help: consider using a reference instead: `&*&x`

error[E0507]: cannot move out of borrowed content
  --> $DIR/std-uncopyable-atomics.rs:15:13
   |
LL |     let x = *&x; //~ ERROR: cannot move out of borrowed content
   |             ^^^
   |             |
   |             cannot move out of borrowed content
   |             help: consider using a reference instead: `&*&x`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0507`.
