error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:4:5
   |
LL |     trace_macros!(); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:5:5
   |
LL |     trace_macros!(1); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:6:5
   |
LL |     trace_macros!(ident); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:7:5
   |
LL |     trace_macros!(for); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:8:5
   |
LL |     trace_macros!(true,); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^^^^^^

error: trace_macros! accepts only `true` or `false`
  --> $DIR/trace_macros-format.rs:9:5
   |
LL |     trace_macros!(false 1); //~ ERROR trace_macros! accepts only `true` or `false`
   |     ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 6 previous errors

