error[E0382]: borrow of moved value: `x`
  --> $DIR/unop-move-semantics.rs:8:5
   |
LL |     !x;
   |      - value moved here
LL | 
LL |     x.clone();  //~ ERROR: use of moved value
   |     ^ value borrowed here after move
   |
   = note: move occurs because `x` has type `T`, which does not implement the `Copy` trait

error[E0505]: cannot move out of `x` because it is borrowed
  --> $DIR/unop-move-semantics.rs:15:6
   |
LL |     let m = &x;
   |             -- borrow of `x` occurs here
...
LL |     !x;  //~ ERROR: cannot move out of `x` because it is borrowed
   |      ^ move out of `x` occurs here
...
LL |     use_mut(n); use_imm(m);
   |                         - borrow later used here

error[E0505]: cannot move out of `y` because it is borrowed
  --> $DIR/unop-move-semantics.rs:17:6
   |
LL |     let n = &mut y;
   |             ------ borrow of `y` occurs here
...
LL |     !y;  //~ ERROR: cannot move out of `y` because it is borrowed
   |      ^ move out of `y` occurs here
LL |     use_mut(n); use_imm(m);
   |             - borrow later used here

error[E0507]: cannot move out of borrowed content
  --> $DIR/unop-move-semantics.rs:24:6
   |
LL |     !*m;  //~ ERROR: cannot move out of borrowed content
   |      ^^ cannot move out of borrowed content

error[E0507]: cannot move out of borrowed content
  --> $DIR/unop-move-semantics.rs:26:6
   |
LL |     !*n;  //~ ERROR: cannot move out of borrowed content
   |      ^^ cannot move out of borrowed content

error: aborting due to 5 previous errors

Some errors occurred: E0382, E0505, E0507.
For more information about an error, try `rustc --explain E0382`.
