error: doc comment not used by rustdoc
  --> $DIR/useless_comment.rs:4:5
   |
LL |     /// a //~ ERROR doc comment not used by rustdoc
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: lint level defined here
  --> $DIR/useless_comment.rs:1:9
   |
LL | #![deny(unused_doc_comments)]
   |         ^^^^^^^^^^^^^^^^^^^

error: doc comment not used by rustdoc
  --> $DIR/useless_comment.rs:7:5
   |
LL |     /// b //~ doc comment not used by rustdoc
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: doc comment not used by rustdoc
  --> $DIR/useless_comment.rs:9:9
   |
LL |         /// c //~ ERROR doc comment not used by rustdoc
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: doc comment not used by rustdoc
  --> $DIR/useless_comment.rs:14:5
   |
LL |     /// foo //~ ERROR doc comment not used by rustdoc
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

