error: field is never used: `c`
  --> $DIR/union-fields-1.rs:6:5
   |
LL |     c: u8, //~ ERROR field is never used
   |     ^^^^^
   |
note: lint level defined here
  --> $DIR/union-fields-1.rs:1:9
   |
LL | #![deny(dead_code)]
   |         ^^^^^^^^^

error: field is never used: `a`
  --> $DIR/union-fields-1.rs:9:5
   |
LL |     a: u8, //~ ERROR field is never used
   |     ^^^^^

error: field is never used: `a`
  --> $DIR/union-fields-1.rs:13:20
   |
LL | union NoDropLike { a: u8 } //~ ERROR field is never used
   |                    ^^^^^

error: field is never used: `c`
  --> $DIR/union-fields-1.rs:18:5
   |
LL |     c: u8, //~ ERROR field is never used
   |     ^^^^^

error: aborting due to 4 previous errors

