error: union expressions should have exactly one field
  --> $DIR/union-fields-2.rs:17:13
   |
LL |     let u = U {}; //~ ERROR union expressions should have exactly one field
   |             ^

error: union expressions should have exactly one field
  --> $DIR/union-fields-2.rs:19:13
   |
LL |     let u = U { a: 0, b: 1 }; //~ ERROR union expressions should have exactly one field
   |             ^

error[E0560]: union `U` has no field named `c`
  --> $DIR/union-fields-2.rs:20:29
   |
LL |     let u = U { a: 0, b: 1, c: 2 }; //~ ERROR union expressions should have exactly one field
   |                             ^ `U` does not have this field
   |
   = note: available fields are: `a`, `b`

error: union expressions should have exactly one field
  --> $DIR/union-fields-2.rs:20:13
   |
LL |     let u = U { a: 0, b: 1, c: 2 }; //~ ERROR union expressions should have exactly one field
   |             ^

error: union expressions should have exactly one field
  --> $DIR/union-fields-2.rs:22:13
   |
LL |     let u = U { ..u }; //~ ERROR union expressions should have exactly one field
   |             ^

error[E0436]: functional record update syntax requires a struct
  --> $DIR/union-fields-2.rs:22:19
   |
LL |     let u = U { ..u }; //~ ERROR union expressions should have exactly one field
   |                   ^

error: union patterns should have exactly one field
  --> $DIR/union-fields-2.rs:25:9
   |
LL |     let U {} = u; //~ ERROR union patterns should have exactly one field
   |         ^^^^

error: union patterns should have exactly one field
  --> $DIR/union-fields-2.rs:27:9
   |
LL |     let U { a, b } = u; //~ ERROR union patterns should have exactly one field
   |         ^^^^^^^^^^

error[E0026]: union `U` does not have a field named `c`
  --> $DIR/union-fields-2.rs:28:19
   |
LL |     let U { a, b, c } = u; //~ ERROR union patterns should have exactly one field
   |                   ^ union `U` does not have this field

error: union patterns should have exactly one field
  --> $DIR/union-fields-2.rs:28:9
   |
LL |     let U { a, b, c } = u; //~ ERROR union patterns should have exactly one field
   |         ^^^^^^^^^^^^^

error: union patterns should have exactly one field
  --> $DIR/union-fields-2.rs:30:9
   |
LL |     let U { .. } = u; //~ ERROR union patterns should have exactly one field
   |         ^^^^^^^^

error: `..` cannot be used in union patterns
  --> $DIR/union-fields-2.rs:30:9
   |
LL |     let U { .. } = u; //~ ERROR union patterns should have exactly one field
   |         ^^^^^^^^

error: `..` cannot be used in union patterns
  --> $DIR/union-fields-2.rs:32:9
   |
LL |     let U { a, .. } = u; //~ ERROR `..` cannot be used in union patterns
   |         ^^^^^^^^^^^

error: aborting due to 13 previous errors

Some errors occurred: E0026, E0436, E0560.
For more information about an error, try `rustc --explain E0026`.
