warning: not reporting region error due to nll
  --> $DIR/dyn-trait.rs:32:16
   |
LL |     static_val(x); //~ ERROR cannot infer
   |                ^

error: free region `'a` does not outlive free region `'static`
  --> $DIR/dyn-trait.rs:32:5
   |
LL |     static_val(x); //~ ERROR cannot infer
   |     ^^^^^^^^^^^^^

error: aborting due to previous error

