error[E0310]: the associated type `<T as Iter>::Item` may not live long enough
  --> $DIR/regions-close-associated-type-into-object.rs:15:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `<T as Iter>::Item: 'static`...
note: ...so that the type `<T as Iter>::Item` will meet its required lifetime bounds
  --> $DIR/regions-close-associated-type-into-object.rs:15:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^

error[E0310]: the associated type `<T as Iter>::Item` may not live long enough
  --> $DIR/regions-close-associated-type-into-object.rs:22:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `<T as Iter>::Item: 'static`...
note: ...so that the type `std::boxed::Box<<T as Iter>::Item>` will meet its required lifetime bounds
  --> $DIR/regions-close-associated-type-into-object.rs:22:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^

error[E0309]: the associated type `<T as Iter>::Item` may not live long enough
  --> $DIR/regions-close-associated-type-into-object.rs:28:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `<T as Iter>::Item: 'a`...
note: ...so that the type `<T as Iter>::Item` will meet its required lifetime bounds
  --> $DIR/regions-close-associated-type-into-object.rs:28:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^

error[E0309]: the associated type `<T as Iter>::Item` may not live long enough
  --> $DIR/regions-close-associated-type-into-object.rs:35:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^
   |
   = help: consider adding an explicit lifetime bound `<T as Iter>::Item: 'a`...
note: ...so that the type `std::boxed::Box<<T as Iter>::Item>` will meet its required lifetime bounds
  --> $DIR/regions-close-associated-type-into-object.rs:35:5
   |
LL |     Box::new(item) //~ ERROR associated type `<T as Iter>::Item` may not live long enough
   |     ^^^^^^^^^^^^^^

error: aborting due to 4 previous errors

Some errors occurred: E0309, E0310.
For more information about an error, try `rustc --explain E0309`.
