error[E0308]: mismatched types
  --> $DIR/mismatched-types.rs:2:20
   |
LL |     let b: &[u8] = include_str!("file.txt");    //~ ERROR mismatched types
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^ expected slice, found str
   |
   = note: expected type `&[u8]`
              found type `&'static str`

error[E0308]: mismatched types
  --> $DIR/mismatched-types.rs:3:19
   |
LL |     let s: &str = include_bytes!("file.txt");   //~ ERROR mismatched types
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected str, found array of 0 elements
   |
   = note: expected type `&str`
              found type `&'static [u8; 0]`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
