error[E0425]: cannot find value `does_not_exist` in this scope
  --> $DIR/match-vec-mismatch.rs:28:11
   |
LL |     match does_not_exist { //~ ERROR cannot find value `does_not_exist` in this scope
   |           ^^^^^^^^^^^^^^ not found in this scope

error[E0529]: expected an array or slice, found `std::string::String`
  --> $DIR/match-vec-mismatch.rs:5:9
   |
LL |         ['f', 'o', ..] => {}
   |         ^^^^^^^^^^^^^^ pattern cannot match with input type `std::string::String`

error[E0527]: pattern requires 1 elements but array has 3
  --> $DIR/match-vec-mismatch.rs:20:9
   |
LL |         [0] => {}, //~ ERROR pattern requires
   |         ^^^ expected 3 elements

error[E0528]: pattern requires at least 4 elements but array has 3
  --> $DIR/match-vec-mismatch.rs:25:9
   |
LL |         [0, 1, 2, 3, x..] => {} //~ ERROR pattern requires
   |         ^^^^^^^^^^^^^^^^^ pattern cannot match array of 3 elements

error[E0282]: type annotations needed
  --> $DIR/match-vec-mismatch.rs:36:9
   |
LL |         [] => {}  //~ ERROR type annotations needed
   |         ^^ cannot infer type
   |
   = note: type must be known at this point

error: aborting due to 5 previous errors

Some errors occurred: E0282, E0425, E0527, E0528, E0529.
For more information about an error, try `rustc --explain E0282`.
